<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Samuele Soraggi">

<title>Advanced single cell analysis - Single Cell Analysis Tutorial</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<script type="application/ld+json">
  {
    "@context": "https://schema.org/",
    "@type": "TrainingMaterial",
    "@id": "https://hds-sandbox.github.io/AdvancedSingleCell",
    "dct:conformsTo": {
      "http://purl.org/dc/terms/conformsTo": {
        "@id": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
        "@type": "CreativeWork"
      }
    },
    "name": "Advanced Single Cell Tutorials",
    "description": "Tutorials with both complete downstream analysis and guides for specific tasks, such as gene networks, cell-cell communication, clustering techniques, integration and comparison, data visualization.",
    "keywords": [
      "single cell",
      "scRNA-seq",
      "gene networks",
      "single cell clustering",
      "tutorial",
      "bioinformatics",
      "downstream analysis",
      "single cell omics",
      "data integration",
      "single cell analysis"
    ],
    "about": [
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/data_3112",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "data_3112",
        "name": "Gene expression matrix",
        "url": "https://edamontology.github.io/edam-browser/#http://edamontology.org/data_3112"
      },
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/topic_0769",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "topic_0769",
        "name": "Workflows",
        "url": "https://edamontology.github.io/edam-browser/#http://edamontology.org/topic_0769"
      },
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/topic_3345",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "topic_3345",
        "name": "Data identity and mapping",
        "url": "https://edamontology.github.io/edam-browser/#http://edamontology.org/topic_3345"
      },
      {
        "@type": "DefinedTerm",
        "@id": "http://edamontology.org/data_3670",
        "inDefinedTermSet": "http://edamontology.org",
        "termCode": "data_3670",
        "name": "Online course",
        "url": "https://edamontology.github.io/edam-browser/#http://edamontology.org/data_3670"
      }
    ],
    "abstract": "Tutorials with both complete downstream analysis and guides for specific tasks, such as gene networks, cell-cell communication, clustering techniques, integration and comparison, data visualization.",
    "audience": [
      {
        "@type": "Audience",
        "audienceType": "Students"
      },
      {
        "@type": "Audience",
        "audienceType": "PhD students"
      },
      {
        "@type": "Audience",
        "audienceType": "Postdoctoral researchers"
      }
      {
        "@type": "Audience",
        "audienceType": "Bioinformaticians"
      },
    ],
    "competencyRequired": [
      "Basic python programming",
      "Basic R programming",
      "Machine learning basics",
      "Basic cell biology knowledge"
    ],
    "inLanguage": ["en-US"],
    "learningResourceType": ["tutorials"],
    "license": "https://spdx.org/licenses/MIT.html",
    "url": "https://hds-sandbox.github.io/AdvancedSingleCell"
  }
</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../../img/logo.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">Advanced single cell analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../index.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../../Modules/Infected_Plant_Gene_Analysis/Notebook/notebook.source.html" rel="" target="" aria-current="page">
 <span class="menu-text">Notebook test</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/hds-sandbox/AdvancedSingleCell" rel="" target=""><i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#a-rather-very-short-biological-background" id="toc-a-rather-very-short-biological-background" class="nav-link active" data-scroll-target="#a-rather-very-short-biological-background"><span class="header-section-number">0.1</span> A (rather very) short biological background</a></li>
  <li><a href="#umi-based-single-cell-data-from-microdroplets" id="toc-umi-based-single-cell-data-from-microdroplets" class="nav-link" data-scroll-target="#umi-based-single-cell-data-from-microdroplets"><span class="header-section-number">1</span> UMI-based single cell data from microdroplets</a>
  <ul class="collapse">
  <li><a href="#the-raw-data-in-practice" id="toc-the-raw-data-in-practice" class="nav-link" data-scroll-target="#the-raw-data-in-practice"><span class="header-section-number">1.1</span> The raw data in practice</a></li>
  <li><a href="#alignment-and-expression-matrix" id="toc-alignment-and-expression-matrix" class="nav-link" data-scroll-target="#alignment-and-expression-matrix"><span class="header-section-number">1.2</span> Alignment and expression matrix</a></li>
  </ul></li>
  <li><a href="#sec-preprocessing" id="toc-sec-preprocessing" class="nav-link" data-scroll-target="#sec-preprocessing"><span class="header-section-number">2</span> Preprocessing</a>
  <ul class="collapse">
  <li><a href="#download-data" id="toc-download-data" class="nav-link" data-scroll-target="#download-data"><span class="header-section-number">2.1</span> Download data</a></li>
  <li><a href="#import-data" id="toc-import-data" class="nav-link" data-scroll-target="#import-data"><span class="header-section-number">2.2</span> Import data</a></li>
  <li><a href="#create-a-single-cell-object-in-seurat" id="toc-create-a-single-cell-object-in-seurat" class="nav-link" data-scroll-target="#create-a-single-cell-object-in-seurat"><span class="header-section-number">2.3</span> Create a single cell object in Seurat</a>
  <ul class="collapse">
  <li><a href="#content-of-a-seurat-object" id="toc-content-of-a-seurat-object" class="nav-link" data-scroll-target="#content-of-a-seurat-object"><span class="header-section-number">2.3.1</span> Content of a Seurat Object</a></li>
  </ul></li>
  <li><a href="#finding-filtering-criteria" id="toc-finding-filtering-criteria" class="nav-link" data-scroll-target="#finding-filtering-criteria"><span class="header-section-number">2.4</span> Finding filtering criteria</a>
  <ul class="collapse">
  <li><a href="#quality-measure-distributions" id="toc-quality-measure-distributions" class="nav-link" data-scroll-target="#quality-measure-distributions"><span class="header-section-number">2.4.1</span> Quality measure distributions</a></li>
  <li><a href="#filtering-with-the-chosen-criteria" id="toc-filtering-with-the-chosen-criteria" class="nav-link" data-scroll-target="#filtering-with-the-chosen-criteria"><span class="header-section-number">2.4.2</span> Filtering with the chosen criteria</a></li>
  </ul></li>
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization"><span class="header-section-number">2.5</span> Normalization</a>
  <ul class="collapse">
  <li><a href="#finding-technical-sources-of-variation" id="toc-finding-technical-sources-of-variation" class="nav-link" data-scroll-target="#finding-technical-sources-of-variation"><span class="header-section-number">2.5.1</span> Finding technical sources of variation</a></li>
  <li><a href="#executing-normalization" id="toc-executing-normalization" class="nav-link" data-scroll-target="#executing-normalization"><span class="header-section-number">2.5.2</span> Executing normalization</a></li>
  <li><a href="#visualizing-the-result" id="toc-visualizing-the-result" class="nav-link" data-scroll-target="#visualizing-the-result"><span class="header-section-number">2.5.3</span> Visualizing the result</a></li>
  </ul></li>
  <li><a href="#removing-doublets" id="toc-removing-doublets" class="nav-link" data-scroll-target="#removing-doublets"><span class="header-section-number">2.6</span> Removing doublets</a></li>
  </ul></li>
  <li><a href="#sec-integration" id="toc-sec-integration" class="nav-link" data-scroll-target="#sec-integration"><span class="header-section-number">3</span> Integration</a>
  <ul class="collapse">
  <li><a href="#clustering-and-cell-type-assignment" id="toc-clustering-and-cell-type-assignment" class="nav-link" data-scroll-target="#clustering-and-cell-type-assignment"><span class="header-section-number">3.1</span> Clustering and cell type assignment</a>
  <ul class="collapse">
  <li><a href="#cluster-assignment-from-visualized-marker-scores" id="toc-cluster-assignment-from-visualized-marker-scores" class="nav-link" data-scroll-target="#cluster-assignment-from-visualized-marker-scores"><span class="header-section-number">3.1.1</span> Cluster assignment from visualized marker scores</a></li>
  <li><a href="#cluster-assignment-from-an-annotated-dataset" id="toc-cluster-assignment-from-an-annotated-dataset" class="nav-link" data-scroll-target="#cluster-assignment-from-an-annotated-dataset"><span class="header-section-number">3.1.2</span> Cluster assignment from an annotated dataset</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#sec-geneanalysis" id="toc-sec-geneanalysis" class="nav-link" data-scroll-target="#sec-geneanalysis"><span class="header-section-number">4</span> Gene Expression Analysis</a>
  <ul class="collapse">
  <li><a href="#differential-gene-expression-dge" id="toc-differential-gene-expression-dge" class="nav-link" data-scroll-target="#differential-gene-expression-dge"><span class="header-section-number">4.1</span> Differential Gene Expression (DGE)</a></li>
  <li><a href="#sec-networks" id="toc-sec-networks" class="nav-link" data-scroll-target="#sec-networks"><span class="header-section-number">4.2</span> Coexpression analysis</a>
  <ul class="collapse">
  <li><a href="#construct-metacells" id="toc-construct-metacells" class="nav-link" data-scroll-target="#construct-metacells"><span class="header-section-number">4.2.1</span> Construct metacells</a></li>
  <li><a href="#select-soft-power-threshold" id="toc-select-soft-power-threshold" class="nav-link" data-scroll-target="#select-soft-power-threshold"><span class="header-section-number">4.2.2</span> Select soft-power threshold</a></li>
  </ul></li>
  <li><a href="#construction-of-co-expression-network" id="toc-construction-of-co-expression-network" class="nav-link" data-scroll-target="#construction-of-co-expression-network"><span class="header-section-number">4.3</span> Construction of co-expression network</a>
  <ul class="collapse">
  <li><a href="#module-eigengenes-mes" id="toc-module-eigengenes-mes" class="nav-link" data-scroll-target="#module-eigengenes-mes"><span class="header-section-number">4.3.1</span> Module Eigengenes (MEs)</a></li>
  <li><a href="#compute-module-connectivity" id="toc-compute-module-connectivity" class="nav-link" data-scroll-target="#compute-module-connectivity"><span class="header-section-number">4.3.2</span> Compute module connectivity</a></li>
  <li><a href="#getting-the-module-assignment-table" id="toc-getting-the-module-assignment-table" class="nav-link" data-scroll-target="#getting-the-module-assignment-table"><span class="header-section-number">4.3.3</span> Getting the module assignment table</a></li>
  </ul></li>
  <li><a href="#differential-module-expression-dme-analysis" id="toc-differential-module-expression-dme-analysis" class="nav-link" data-scroll-target="#differential-module-expression-dme-analysis"><span class="header-section-number">4.4</span> Differential module Expression (DME) analysis</a>
  <ul class="collapse">
  <li><a href="#one-versus-all-dme-analysis" id="toc-one-versus-all-dme-analysis" class="nav-link" data-scroll-target="#one-versus-all-dme-analysis"><span class="header-section-number">4.4.1</span> One-versus-all DME analysis</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="notebook.source.out.ipynb" download="notebook.source.out.ipynb"><i class="bi bi-journal-code"></i>Jupyter</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Single Cell Analysis Tutorial</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Samuele Soraggi <a href="https://orcid.org/0000-0002-1159-5535" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<p>This tutorial will give you the extensive basic commands and explanations for the single cell analysis of your own dataset.</p>
<ul>
<li><p>The <strong>first part of the tutorial</strong> (<a href="#sec-preprocessing">Section&nbsp;2</a>) is focused on <strong>preprocessing</strong> the data, which means primarily filtering and normalizing it.</p></li>
<li><p>The <strong>second part of the tutorial</strong> (<a href="#sec-integration">Section&nbsp;3</a>) is focused on <strong>integrating</strong> all sixteen datasets produced from the lab sessions (you will perform this integration analysis in groups), identifying cell types and find a population of cells expressing the HAR1 gene to analyze different conditions of mutant VS wild type Lotus japonicus.</p></li>
<li><p>The <strong>third part of the tutorial</strong> (<a href="#sec-geneanalysis">Section&nbsp;4</a>) applies tipycal gene analysis to detect genes conserved and differentially expressed between conditions</p></li>
<li><p>The <strong>fourth part of the tutorial</strong> (<a href="#sec-networks">Section&nbsp;4.2</a>) pivots around the study of groups of genes co-expressed in the data and in specific clusters and conditions</p></li>
</ul>
<p>The first two parts follow the phylosophy of the best practices explained in <span class="citation" data-cites="luecken_current_2019">Luecken and Theis (<a href="#ref-luecken_current_2019" role="doc-biblioref">2019</a>)</span> and <span class="citation" data-cites="heumos_best_2023">Heumos et al. (<a href="#ref-heumos_best_2023" role="doc-biblioref">2023</a>)</span>. The third part applies standard statistical tests on the average gene expressions in subsets of the data. The last part is based pulling cells transcripts together with different granularities to improve the statistical power of calculations based on their gene expression (as in <span class="citation" data-cites="morabito_hdwgcna_2023">Morabito et al. (<a href="#ref-morabito_hdwgcna_2023" role="doc-biblioref">2023</a>)</span>).</p>
<p>The tutorial is based on four samples of Lotus Japonicus (two rhizobia-infected and two wild types) from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10630511/pdf/41467_2023_Article_42911.pdf"><span class="citation" data-cites="frank_single-cell_2023">Frank et al. (<span>2023</span>)</span></a>. The last section follows some of the <a href="https://smorabit.github.io/hdWGCNA/">tutorials from hdWGCNA</a>.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled" title="Learning outcomes">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning outcomes
</div>
</div>
<div class="callout-body-container callout-body">
<p>At the end of this tutorial <strong>you will be able to use <code>R</code> to</strong></p>
<ul>
<li><strong>Filter</strong> your data selecting specific criteria</li>
<li><strong>Preprocess</strong> your data for advanced analysis</li>
<li><strong>Merge and integrate</strong> datasets and perform <strong>cross-data analysis</strong></li>
<li><strong>Identify potential cell types</strong> by markers and exploiting other datasets</li>
<li>Perform and elaborate <strong>differential and conserved</strong> gene expression</li>
<li>Infer <strong>gene modules</strong> from your data and isolate significant ones related to a cell type</li>
<li><strong>Visualize gene modules</strong> and extract their <strong>gene ontology</strong> to draw biological conclusions</li>
</ul>
</div>
</div>
<section id="a-rather-very-short-biological-background" class="level2" data-number="0.1">
<h2 data-number="0.1" class="anchored" data-anchor-id="a-rather-very-short-biological-background"><span class="header-section-number">0.1</span> A (rather very) short biological background</h2>
<p>Lotus Japonicus is a legume characterized by the legume-rhizobium symbiotic interaction (rhizobia are soil microorganisms that can interact with leguminous plants to form root nodules within which conditions are favourable for bacterial nitrogen fixation. Legumes allow the development of very large rhizobial populations in the vicinity of their roots). <a href="#fig-nitrogenfixing">Figure&nbsp;1</a> and text below it from <span class="citation" data-cites="wang_genetic_2018">Wang, Liu, and Zhu (<a href="#ref-wang_genetic_2018" role="doc-biblioref">2018</a>)</span>.</p>
<div id="fig-nitrogenfixing" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/nitrogenfixing.jpg" class="img-fluid figure-img" width="800"></p>
<figcaption class="figure-caption">Figure&nbsp;1: Symbiosis signaling and plant immunity involved in recognition specificity in the legume-rhizobial interactions (indicated by the red stars). <strong>A</strong> The process of infection and nodule development. A mature indeterminate nodule contains a meristem zone (I), an infection zone (II), an interzone (IZ), a nitrogen fixing zone (III), and a senescent zone (IV). <strong>B</strong> The host secretes flavonoids to induce the expression of bacterial nodulation (nod) gene through the activation of NodD proteins. The enzymes encoded by the nod genes lead to the synthesis of Nod factors (NF) that are recognized by host Nod factor receptors (NFRs). Recognition specificity occurs both between Flavonoids and NodDs and between NF and NFRs. <strong>C</strong> In addition to NF signaling, bacteria also produce extracellular polysaccharides (EPS) and type III effectors to facilitate their infection in compatible interactions, but these molecules may also induce immune responses causing resistance to infection in incompatible interactions. <strong>D</strong> Certain legumes such as Medicago encode antimicrobial nodule-specific cysteine-rich (NCR) peptides to drive their bacterial partners to terminal differentiation that is required for nitrogen fixation. However, some rhizobial strains cannot survive the antibacterial activity of certain peptide isoforms, leading to formation of nodules defective in nitrogen fixation.</figcaption>
</figure>
</div>
<p>Rhizobial invasion of legumes is primarily mediated by a plant-made tubular invagination called an infection thread (IT). Research has shown that various genes are involved in some of the processes of the legume-rhizobia interaction. <a href="#fig-ljinfection">Figure&nbsp;2</a> and text below it from <span class="citation" data-cites="szczyglowski_nodule_1998">Szczyglowski et al. (<a href="#ref-szczyglowski_nodule_1998" role="doc-biblioref">1998</a>)</span>.</p>
<div id="fig-ljinfection" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/LJinfection.png" class="img-fluid figure-img" width="800"></p>
<figcaption class="figure-caption">Figure&nbsp;2: Primary infection of Lotus japonicus roots inoculated with Mesorhizobium loti strain NZP2235. <strong>A</strong>, Brightfield micrograph of root hair curlings resembling shepherd’s-crook structures (arrows). <strong>B–C</strong>, Phase contrast micrographs of the intracellular infection threads within curled root hairs. Arrows point to infection threads with papilla-like structures on their outer surface. Bar, 70 µm <strong>A</strong>, 25 µm <strong>B</strong>, and 20 µm <strong>C</strong>.</figcaption>
</figure>
</div>
<ul>
<li><strong>RINRK1</strong> (Rhizobial Infection Receptor-like Kinase1), that is induced by Nod factors (NFs) and is involved in IT formation but not nodule organogenesis. A paralog, RINRK2, plays a relatively minor role in infection. RINRK1 is required for full induction of early infection genes, including Nodule Inception (NIN), encoding an essential nodulation transcription factor. See <span class="citation" data-cites="li_atypical_2019">Li et al. (<a href="#ref-li_atypical_2019" role="doc-biblioref">2019</a>)</span>.</li>
<li><strong>HAR1</strong> mediates nitrate inhibition and autoregulation of nodulation. Autoregulation of nodulation involves root-to-shoot-to-root long-distance communication, and HAR1 functions in shoots. HAR1 is critical for the inhibition of nodulation at 10 mM nitrate. The nitrate-induced CLE-RS2 glycopeptide binds directly to the HAR1 receptor, this result suggests that CLE-RS2/HAR1 long-distance signaling plays an important role in the both nitrate inhibition and the autoregulation of nodulation. See <span class="citation" data-cites="okamoto_shoot_2015">Okamoto and Kawaguchi (<a href="#ref-okamoto_shoot_2015" role="doc-biblioref">2015</a>)</span>.</li>
<li><strong>SYMRKL1</strong>, encodes a protein with an ectodomain predicted to be nearly identical to that of SYMRK and is required for normal infection thread formation. See <span class="citation" data-cites="frank_single-cell_2023">Frank et al. (<a href="#ref-frank_single-cell_2023" role="doc-biblioref">2023</a>)</span>.</li>
</ul>
</section>
<section id="umi-based-single-cell-data-from-microdroplets" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> UMI-based single cell data from microdroplets</h1>
<p>The dataset is based on a <strong>microdroplet-based method from 10X chromium</strong>. We remember that a microdroplet single cell sequencing protocol works as follow:</p>
<ul>
<li>each cell is isolated together with a barcode bead in a gel/oil droplet</li>
</ul>
<div id="fig-beads" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/dropletisolation.jpg" class="img-fluid figure-img" width="600"></p>
<figcaption class="figure-caption">Figure&nbsp;3: Isolation of cells and beads into microdroplets.</figcaption>
</figure>
</div>
<ul>
<li>each transcript in the cell is captured via the bead and assigned a cell barcode and a transcript unique molecular identifier (UMI)</li>
<li>3’ reverse transcription of mRNA into cDNA is then performed in preparation to the PCR amplification</li>
<li>the cDNA is amplified through PCR cycles</li>
</ul>
<div id="fig-steps" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/10X.png" class="img-fluid figure-img" width="600"></p>
<figcaption class="figure-caption">Figure&nbsp;4: steps for the microdroplet-based single cell RNA sequencing after isolation.</figcaption>
</figure>
</div>
<section id="the-raw-data-in-practice" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="the-raw-data-in-practice"><span class="header-section-number">1.1</span> The raw data in practice</h2>
<p>Let’s look at a specific read and its UMI and cell barcode. The data is organized in paired-end reads (written on <code>fastq</code> files), where the first <code>fastq</code> file contains reads in the following format</p>
<pre><code>@SRR8363305.1 1 length=26
NTGAAGTGTTAAGACAAGCGTGAACT
+SRR8363305.1 1 length=26
#AAFFJJJJJJJJJJJJJJJJFJJJJ</code></pre>
<p>Here, the first 16 characters <code>NTGAAGTGTTAAGACA</code> represent the cell barcode, while the last 10 characters <code>AGCGTGAACT</code> are the transcript UMI tag. The last line represents the quality scores of the 26 characters of barcode+UMI.</p>
<p>The associated second <code>fastq</code> file contains reads of 98nt as the following</p>
<pre><code>@SRR8363305.1 1 length=98
NCTAAAGATCACACTAAGGCAACTCATGGAGGGGTCTTCAAAGA
    CCTTGCAAGAAGTACTAACTATGGAGTATCGGCTAAGTCAANCN
    TGTATGAGAT
+SRR8363305.1 1 length=98
#A&lt;77AFJJFAAAJJJ7-7-&lt;7FJ-7----&lt;77--7FAAA--
    &lt;JFFF-7--7&lt;&lt;-F77---FF---7-7A-777777A-&lt;
    -7---#-#A-7-7--7--</code></pre>
<p>The 98nt-long string of characters in the second line is a partial sequence of the cDNA transcript. Specifically, the 10X chromium protocol used for sequencing the data is biased towards the 3’ end, because the sequencing is oriented from the 3’ to the 5’ end of the transcripts. The last line contains the corresponding quality scores.</p>
</section>
<section id="alignment-and-expression-matrix" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="alignment-and-expression-matrix"><span class="header-section-number">1.2</span> Alignment and expression matrix</h2>
<p>The data is aligned with <code>cellranger</code>, a completely automatized <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">pipeline implemented by 10X</a> for 10X-genomics data.</p>
<p>Apart from the data, the output contains an interactive document reporting the quality of the data and a small preliminary UMAP plot and clustering. In this report it is especially instructive to look at the <strong>knee plot</strong>.</p>
<p>The knee plot is created by plotting the number of unique molecular identifiers (UMIs) or reads against the number of cells sequenced, sorted in descending order. The UMIs or reads are a measure of the amount of RNA captured for each cell, and thus a measure of the quality of the data. The plot typically shows a <strong>steep slope at the beginning, followed by a plateau, and then a gradual decrese into a second slope and a final plateau</strong>.</p>
<ul>
<li>The steep slope represents the initial cells that are of <strong>high quality</strong> and have the highest number of UMIs or reads.</li>
<li>The first plateau represents the cells that have <strong>lower quality data</strong>, and the gradual decrease represents the addition of droplets with even lower quality data.</li>
<li>Usually, beyond the first slope, you have droplets that are <strong>either empty or of so poor quality</strong>, that they are not worth keeping for analysis.</li>
<li>The height of the last plateau gives you an idea of the <strong>presence of ambient RNA</strong> inside droplets. If the last plateau is located high up, then the corresponding amount of UMIs consist of background ambient RNA which likely pollutes all cells in your data.</li>
</ul>
<p>Below, the knee plot from the <code>control 1</code> sample used in this analysis. You can see that around 10,000 cells with above ~1000 UMIs seems to be coinsidered of decent quality by <code>cellranger</code> (the part of line coloured in blue). Note that the last plateau is located at a very low amount of UMIs, meaning there is not really any relevant contamination from ambient RNA.</p>
<div id="fig-knee" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/knee.png" class="img-fluid figure-img" width="600"></p>
<figcaption class="figure-caption">Figure&nbsp;5: Knee plot of the Control 1 sample of the tutorial. Note the lower plateau of ambient RNA</figcaption>
</figure>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled" title="Exercise">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>In this same folder you have the document <code>web_summary.html</code> that shows you the quality report of an example dataset produced using the <code>10X cellranger</code> pipeline for alignment. <strong>Take some time to look at it and explore what it contains</strong> (Click on <code>Trust HTML</code> on the top menu if the html remains blank after opening it). Get an idea of how many UMIs there are in cells of decent quality. We will work more on filtering out cells based on their quality in this tutorial.</p>
<p>You can read a <a href="https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-cell-ranger-web-summary-files-for-single-cell-gene-expression-assays">technical note</a> on interpreting cellranger reports.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Something more about knee plots">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Something more about knee plots
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The background RNA (sequenced together with the transcript coming from the cell of interest) makes up the <em>ambient plateau</em>: the same background RNA is contained in empty droplets. If your dataset has extremely few UMI counts in empty droplets, then there is not much background RNA present - This is the best situation in which you can find yourself. See Exhibit A in <a href="#fig-bender">Figure&nbsp;6</a>.</p>
<p>If you have a dataset where you can identify an <em>empty droplet plateau</em> by eye (Exhibit B in <a href="#fig-bender">Figure&nbsp;6</a>), and these empty droplets have 50 or 100 or several hundred counts, then it can be advisable to use a specific software to remove the background transcripts (e.g.&nbsp;<code>CellBender</code> (<span class="citation" data-cites="fleming_unsupervised_2023">Fleming et al. (<a href="#ref-fleming_unsupervised_2023" role="doc-biblioref">2023</a>)</span>), <code>SoupX</code> (<span class="citation" data-cites="young_soupx_2020">Young and Behjati (<a href="#ref-young_soupx_2020" role="doc-biblioref">2020</a>)</span>)).</p>
<p>If you have a dataset with so much background RNA that you cannot identify the <em>empty droplet plateau</em> yourself by eye (Exhibit C in <a href="#fig-bender">Figure&nbsp;6</a>), then any software to remove background transcripts will also likely have a difficult time. Such the algorithms might be worth a try, but you should <strong>strongly consider re-running the experiment, as the knee plot points to a real QC failure</strong></p>
<div id="fig-bender" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/UMI_curve_tropes.png" class="img-fluid figure-img" width="600"></p>
<figcaption class="figure-caption">Figure&nbsp;6: Various cases of knee plot you can encounter from sequenced data. Figure from <a href="https://cellbender.readthedocs.io/">the webpage of Cellbender.</a></figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="sec-preprocessing" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Preprocessing</h1>
<div class="callout callout-style-default callout-note callout-titled" title="Learning outcome">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning outcome
</div>
</div>
<div class="callout-body-container callout-body">
<p>We will answer to the following questions:</p>
<ul>
<li>How can I <strong>import single-cell data</strong> into R?</li>
<li>How are different types of data/information (e.g.&nbsp;cell information, gene information, etc.) <strong>stored and manipulated</strong>?</li>
<li>How can I obtain basic <strong>summary metrics</strong> for cells and genes and <strong>filter the data</strong> accordingly?</li>
<li>How can I <strong>visually explore</strong> these metrics?</li>
</ul>
</div>
</div>
<p>We start by loading all the packages necessary for the analysis and setup a few things</p>
<div class="cell" data-tags="[]" data-execution_count="8">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratDisk); </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat);</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DoubletFinder);</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel);</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(multtest);</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(metap);</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr);</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr);</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr);</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble);</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2);</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MAST);</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(WGCNA);</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(hdWGCNA);</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork);</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(doFuture);})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"../Scripts/script.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="10">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.seed=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="11">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>(<span class="st">"multicore"</span>, <span class="at">workers =</span> <span class="dv">8</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">8</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="download-data" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="download-data"><span class="header-section-number">2.1</span> Download data</h2>
<p>We check if the data exists, otherwise a script will download the missing data files in the appropriate folder, which should be <code>../Data</code>.</p>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">downloadData</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Data folder exists. Check for files and eventually downloads them. Please wait.

Done!
</code></pre>
</div>
</div>
</section>
<section id="import-data" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="import-data"><span class="header-section-number">2.2</span> Import data</h2>
<p>We import the data reading the matrix files aligned by 10X. Those are usually contained in a folder with a name of the type <code>aligned_dataset/outs/filtered_bc_matrix</code>, that 10X Cellranger creates automatically after the alignment. You need to use such a folder when your own data is aligned and you need to import it. In this tutorial, the aligned data is in the folder <code>../Data/control_1</code> used below. The command for reading the data is simply <code>Read10X</code>.</p>
<div class="cell" data-tags="[]" data-execution_count="16">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>Control1 <span class="ot">&lt;-</span> <span class="fu">Read10X</span>(<span class="st">"../Data/control_1/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled" title="Preprocessing multiple datasets">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Preprocessing multiple datasets
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that we are loading only one dataset (<code>control_1</code>, one of the two control replicates). Another control dataset, and two infected datasets, have already been preprocessed and will be used later - so <strong>we will now focus on the preprocessing of a single dataset</strong>. In general, when you have multiple datasets, you must preprocess them one at a time before integrating them together.</p>
</div>
</div>
<p>What we obtain in the command above is an expression matrix. Look at the first 10 rows and columns of the matrix (whose rows represent genes and columns droplets/cells) - the dots are zeros (they are not stored in the data, which has a compressed format called <code>dgCMatrix</code>), and <strong>are the majority of the expression values obtained in scRNA data!</strong></p>
<div class="cell" data-tags="[]" data-execution_count="17">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>Control1[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>  [[ suppressing 10 column names ‘AAACCCAAGGGCAGTT-1’, ‘AAACCCAAGTCAGCGA-1’, ‘AAACCCACACTAACCA-1’ ... ]]
</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>10 x 10 sparse Matrix of class "dgCMatrix"
                                         
LotjaGi0g1v0000100    . . . . . . . . . .
LotjaGi0g1v0000200    . . . . . . . . . .
LotjaGi0g1v0000300    . . . . . . . . . .
LotjaGi0g1v0000400    . . . . . . 1 1 . .
LotjaGi0g1v0000500    . . . . . . . . . .
LotjaGi0g1v0000600    . . . . . . . . . .
LotjaGi0g1v0000700    . . . . . . . . . .
LotjaGi0g1v0000800    . . . . 1 . . . . .
LotjaGi0g1v0000900_LC . . . . . . . . . .
LotjaGi0g1v0001000_LC . . . . . . . . . .</code></pre>
</div>
</div>
<p>What is the percentage of zeros in this matrix? You can see it for yourself below - it is a lot, but quite surprisingly we can get a lot of information from the data!</p>
<div class="cell" data-tags="[]" data-execution_count="18">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Number of zeros: "</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>zeros <span class="ot">&lt;-</span>  <span class="fu">sum</span>(Control1<span class="sc">==</span><span class="dv">0</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( zeros )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of zeros: 307060673</code></pre>
</div>
</div>
<div class="cell" data-tags="[]" data-execution_count="19">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Number of expression entries: "</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>total <span class="ot">&lt;-</span> <span class="fu">dim</span>(Control1)[<span class="dv">1</span>] <span class="sc">*</span> <span class="fu">dim</span>(Control1)[<span class="dv">2</span>]</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( total )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of expression entries: 329798055</code></pre>
</div>
</div>
<div class="cell" data-tags="[]" data-execution_count="20">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Percentage of zeros: "</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( zeros <span class="sc">/</span> total <span class="sc">*</span> <span class="dv">100</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Percentage of zeros: 93.10567</code></pre>
</div>
</div>
</section>
<section id="create-a-single-cell-object-in-seurat" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="create-a-single-cell-object-in-seurat"><span class="header-section-number">2.3</span> Create a single cell object in Seurat</h2>
<p>We use our count matrix to create a Seurat object. A Seurat object allows you to <strong>store the count matrix</strong> and future modifications of it (for example its normalized version), together with <strong>information regarding cells and genes</strong> (such as clusters of cell types) and their <strong>projections</strong> (such as PCA and tSNE). We will go through these elements, but first we create the object with <code>CreateSeuratObject</code>:</p>
<div class="cell" data-tags="[]" data-execution_count="52">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> Control1, </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">project =</span> <span class="st">"Control1_seurat"</span>, </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">min.cells =</span> <span class="dv">3</span>, </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">min.features =</span> <span class="dv">200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Feature names cannot have underscores ('_'), replacing with dashes ('-')”
Warning message:
“Feature names cannot have underscores ('_'), replacing with dashes ('-')”</code></pre>
</div>
</div>
<p>The arguments of the command are * <code>counts</code>: the count matrix * <code>project</code>: a project name * <code>min.cells</code>: a minimum requirement for genes, in our case saying they must be expressed in at least 3 cells. If not, they are filtered out already now when creating the object. * <code>min.features</code>: a minimum requirement for cells. Cells having less than 200 expressed genes are removed from the beginning from the data.</p>
<p>Values for the minimum requirements chosen above are standard checks when running the analysis. Droplets not satisfying those requirements are of extremely bad quality and not worth carrying on during the analysis (remember the knee plot).</p>
<p>How many genes and cells have been filtered out?</p>
<div class="cell" data-tags="[]" data-execution_count="53">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Starting Genes and Cells:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( <span class="fu">dim</span>(Control1) )</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Filtered Genes and Cells:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( <span class="fu">dim</span>(Control1) <span class="sc">-</span> <span class="fu">dim</span>(Control1_seurat) )</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Remaining Genes and Cells:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( <span class="fu">dim</span>(Control1_seurat) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting Genes and Cells:
30585 10783
Filtered Genes and Cells:
6747 11
Remaining Genes and Cells:
23838 10772</code></pre>
</div>
</div>
<p>We want to use this data later in the analysis with other <code>Control</code> and <code>Infected</code> datasets. Therefore we add a <code>Condition</code> to the metadata table, and for this dataset we establish that each cell is <code>Control</code>.</p>
<div class="cell" data-tags="[]" data-execution_count="54">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(<span class="at">object =</span> Control1_seurat, </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">metadata =</span> <span class="st">"Control"</span>, </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">col.name =</span> <span class="st">"Condition"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="content-of-a-seurat-object" class="level3" data-number="2.3.1">
<h3 data-number="2.3.1" class="anchored" data-anchor-id="content-of-a-seurat-object"><span class="header-section-number">2.3.1</span> Content of a Seurat Object</h3>
<p>What is contained in the Seurat object? We can use the command <code>str</code> to list the various <em>slots</em> of the object.</p>
<div class="cell" data-tags="[]" data-execution_count="55">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(Control1_seurat, <span class="at">max.level =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Formal class 'Seurat' [package "SeuratObject"] with 13 slots
  ..@ assays      :List of 1
  ..@ meta.data   :'data.frame':    10772 obs. of  4 variables:
  ..@ active.assay: chr "RNA"
  ..@ active.ident: Factor w/ 1 level "Control1_seurat": 1 1 1 1 1 1 1 1 1 1 ...
  .. ..- attr(*, "names")= chr [1:10772] "AAACCCAAGGGCAGTT-1" "AAACCCAAGTCAGCGA-1" "AAACCCACACTAACCA-1" "AAACCCACATGATCTG-1" ...
  ..@ graphs      : list()
  ..@ neighbors   : list()
  ..@ reductions  : list()
  ..@ images      : list()
  ..@ project.name: chr "Control1_seurat"
  ..@ misc        : list()
  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1
  ..@ commands    : list()
  ..@ tools       : list()</code></pre>
</div>
</div>
<p>The first slot is called <code>assays</code>, and it contains all the count matrices we have collected during our analysis when, for example, normalizing data or doing other transformations of it. Right now we only have the <code>RNA</code> assay with the raw counts:</p>
<div class="cell" data-tags="[]" data-execution_count="56">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat<span class="sc">@</span>assays</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>$RNA
Assay data with 23838 features for 10772 cells
First 10 features:
 LotjaGi0g1v0000100, LotjaGi0g1v0000200, LotjaGi0g1v0000300,
LotjaGi0g1v0000400, LotjaGi0g1v0000500, LotjaGi0g1v0000700,
LotjaGi0g1v0000800, LotjaGi0g1v0001100, LotjaGi0g1v0001200,
LotjaGi0g1v0001400 </code></pre>
</div>
</div>
<p>You can always select which matrix is currently in use for the analysis by assigning it to <code>DefaultAssay()</code>. The default assay is often changed automatically by Seurat, for example the normalized assay is used as default after normalization is performed.</p>
<div class="cell" data-tags="[]" data-execution_count="57">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(<span class="at">object =</span> Control1_seurat) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="58">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Your default assay is "</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">DefaultAssay</span>(<span class="at">object =</span> Control1_seurat))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Your default assay is RNA</code></pre>
</div>
</div>
<p>The second slot is the one that contains the metadata for each cell. It is easily visualized as a table (the command <code>head</code> shows only the first 6 rows of the table):</p>
<div class="cell" data-tags="[]" data-execution_count="59">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>( Control1_seurat<span class="sc">@</span>meta.data )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 4</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">orig.ident</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">Condition</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">AAACCCAAGGGCAGTT-1</td>
<td>Control1_seurat</td>
<td>3567</td>
<td>1919</td>
<td>Control</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">AAACCCAAGTCAGCGA-1</td>
<td>Control1_seurat</td>
<td>7015</td>
<td>2751</td>
<td>Control</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">AAACCCACACTAACCA-1</td>
<td>Control1_seurat</td>
<td>1484</td>
<td>828</td>
<td>Control</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">AAACCCACATGATCTG-1</td>
<td>Control1_seurat</td>
<td>20942</td>
<td>4711</td>
<td>Control</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">AAACCCAGTAGCTTGT-1</td>
<td>Control1_seurat</td>
<td>29105</td>
<td>5157</td>
<td>Control</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">AAACCCAGTCTCTCAC-1</td>
<td>Control1_seurat</td>
<td>6115</td>
<td>2124</td>
<td>Control</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The table contains a name for the dataset (<code>orig.ident</code>, useful to distinguish multiple datasets merged together), how many RNA transcripts are contained in each cell (<code>nCount_RNA</code>), the number of expressed genes in each cell (<code>nFeature_RNA</code>), and the <code>Condition</code> (added by us previously). More metadata can be added along the analysis, and some is added automatically by Seurat when running specific commands.</p>
<p>The <code>assays</code> and <code>meta.data</code> slots are the most relevant and useful to know - the other ones are mostly for internal use by Seurat and we do not go into detail with those.</p>
</section>
</section>
<section id="finding-filtering-criteria" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="finding-filtering-criteria"><span class="header-section-number">2.4</span> Finding filtering criteria</h2>
<p>We want to look in depth at which droplets do not contain good quality data, so that we can filter them out. The standard approach - which works quite well - is to study the <strong>distribution of various quality measures and remove doublets</strong> (droplets containing more than one cell) which can confound the analysis results. We will look at some plots and decide some threshold, then we will apply them at the end after looking at all the histograms.</p>
<section id="quality-measure-distributions" class="level3" data-number="2.4.1">
<h3 data-number="2.4.1" class="anchored" data-anchor-id="quality-measure-distributions"><span class="header-section-number">2.4.1</span> Quality measure distributions</h3>
<p>A first step is to calculate the percentage of mitochondrial and chloroplastic genes. A high percentage indicates the presence of spilled material from broken cells. We use the command <code>PercentageFeatureSet</code> and provide the pattern of the gene ID which corresponds to mitochondrial and ribosomal genes. The percentages are saved into the metadata simply by using the double squared brackets <code>[[</code>.</p>
<div class="cell" data-tags="[]" data-execution_count="60">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(Control1_seurat, </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>                                                                 <span class="at">pattern =</span> <span class="st">"LotjaGiM1v"</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>Control1_seurat[[<span class="st">"percent.chloroplast"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(Control1_seurat, </span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>                                                                          <span class="at">pattern =</span> <span class="st">"LotjaGiC1v"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can see the new metadata is now added for each cell</p>
<div class="cell" data-tags="[]" data-execution_count="61">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>( Control1_seurat<span class="sc">@</span>meta.data )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 6</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">orig.ident</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">Condition</th>
<th data-quarto-table-cell-role="th" scope="col">percent.mt</th>
<th data-quarto-table-cell-role="th" scope="col">percent.chloroplast</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">AAACCCAAGGGCAGTT-1</td>
<td>Control1_seurat</td>
<td>3567</td>
<td>1919</td>
<td>Control</td>
<td>4.6818054</td>
<td>0.02803476</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">AAACCCAAGTCAGCGA-1</td>
<td>Control1_seurat</td>
<td>7015</td>
<td>2751</td>
<td>Control</td>
<td>0.7127584</td>
<td>0.04276550</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">AAACCCACACTAACCA-1</td>
<td>Control1_seurat</td>
<td>1484</td>
<td>828</td>
<td>Control</td>
<td>6.6037736</td>
<td>0.06738544</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">AAACCCACATGATCTG-1</td>
<td>Control1_seurat</td>
<td>20942</td>
<td>4711</td>
<td>Control</td>
<td>0.4775093</td>
<td>0.12415242</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">AAACCCAGTAGCTTGT-1</td>
<td>Control1_seurat</td>
<td>29105</td>
<td>5157</td>
<td>Control</td>
<td>0.2954819</td>
<td>0.05497337</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">AAACCCAGTCTCTCAC-1</td>
<td>Control1_seurat</td>
<td>6115</td>
<td>2124</td>
<td>Control</td>
<td>1.6026165</td>
<td>0.01635323</td>
</tr>
</tbody>
</table>
</div>
</div>
<section id="number-of-transcripts-per-cell" class="level4" data-number="2.4.1.1">
<h4 data-number="2.4.1.1" class="anchored" data-anchor-id="number-of-transcripts-per-cell"><span class="header-section-number">2.4.1.1</span> Number of transcripts per cell</h4>
<p>We plot a histogram of the number of transcripts per cell in <a href="#fig-rnacount">Figure&nbsp;7</a> below. On the right, we zoom into the histogram. We want to <strong>filter out the cells with the lowest number of transcripts</strong> - often there is a peak we can identify with a group of low-quality cells. Here we can choose to remove cells with less than ~700 transcripts (some people prefere to do a lighter filtering, and would for example set a threshold to a lower value). We remove also <strong>cells with too many transcripts</strong> that might contain some weird transcripts - which is also helpful for normalization because it removes some outlying values. For those we can set a limit to 30000, where there is a very thin tail in the histogram.</p>
<div class="cell" data-tags="[]" data-execution_count="62">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">14</span>, <span class="at">repr.plot.height=</span><span class="dv">5</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(Control1_seurat<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x=</span>nCount_RNA)) <span class="sc">+</span> </span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_histogram</span>(<span class="at">fill=</span><span class="st">"#69b3a2"</span>, <span class="at">color=</span><span class="st">"#e9ecef"</span>, <span class="at">alpha=</span><span class="fl">0.9</span>)</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>plot2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(Control1_seurat<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x=</span>nCount_RNA)) <span class="sc">+</span> </span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_histogram</span>(<span class="at">fill=</span><span class="st">"#69b3a2"</span>, <span class="at">color=</span><span class="st">"#e9ecef"</span>, <span class="at">alpha=</span><span class="fl">0.9</span>) <span class="sc">+</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>     <span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">2000</span>)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>plot1 <span class="sc">+</span> plot2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
Warning message:
“Removed 7668 rows containing non-finite values (`stat_bin()`).”
Warning message:
“Removed 2 rows containing missing values (`geom_bar()`).”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-rnacount" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-rnacount-output-2.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;7: Histogram of transcripts per cell (left) and a zoom onto the histogram (right)</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="number-of-detected-genes-per-cell" class="level4" data-number="2.4.1.2">
<h4 data-number="2.4.1.2" class="anchored" data-anchor-id="number-of-detected-genes-per-cell"><span class="header-section-number">2.4.1.2</span> Number of detected genes per cell</h4>
<p>Here we work similarly to filter out cells based on how many genes are detected (<a href="#fig-genecounts">Figure&nbsp;8</a>). The right-side plot is a zoom into the histogram. It seems easy to set the thresholds at ~400 and ~7000 detected genes.</p>
<div class="cell" data-tags="[]" data-execution_count="63">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">14</span>, <span class="at">repr.plot.height=</span><span class="dv">5</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(Control1_seurat<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x=</span>nFeature_RNA)) <span class="sc">+</span> </span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_histogram</span>(<span class="at">fill=</span><span class="st">"#69b3a2"</span>, <span class="at">color=</span><span class="st">"#e9ecef"</span>, <span class="at">alpha=</span><span class="fl">0.9</span>)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>plot2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(Control1_seurat<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x=</span>nFeature_RNA)) <span class="sc">+</span> </span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_histogram</span>(<span class="at">fill=</span><span class="st">"#69b3a2"</span>, <span class="at">color=</span><span class="st">"#e9ecef"</span>, <span class="at">alpha=</span><span class="fl">0.9</span>) <span class="sc">+</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>     <span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">1000</span>)</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>plot1 <span class="sc">+</span> plot2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
Warning message:
“Removed 7793 rows containing non-finite values (`stat_bin()`).”
Warning message:
“Removed 2 rows containing missing values (`geom_bar()`).”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-genecounts" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-genecounts-output-2.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;8: Histogram of detected counts per cell (left) and a zoom onto the histogram (right)</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="mitochondrial-and-chloroplast-percentages" class="level4" data-number="2.4.1.3">
<h4 data-number="2.4.1.3" class="anchored" data-anchor-id="mitochondrial-and-chloroplast-percentages"><span class="header-section-number">2.4.1.3</span> Mitochondrial and Chloroplast percentages</h4>
<p>The percentages of mitochondrial and chloroplastic transcripts tells us the data is of good quality, since most cells have low values of those (<a href="#fig-mt">Figure&nbsp;9</a>). Thresholds are usually set between 5% and 20% in single cell data analysis. In the paper, thresholds were for example set at 20%.</p>
<div class="cell" data-tags="[]" data-execution_count="64">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">14</span>, <span class="at">repr.plot.height=</span><span class="dv">5</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(Control1_seurat<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x=</span>percent.mt)) <span class="sc">+</span> </span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_histogram</span>(<span class="at">fill=</span><span class="st">"#69b3a2"</span>, <span class="at">color=</span><span class="st">"#e9ecef"</span>, <span class="at">alpha=</span><span class="fl">0.9</span>)</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>plot2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(Control1_seurat<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x=</span>percent.chloroplast)) <span class="sc">+</span> </span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_histogram</span>(<span class="at">fill=</span><span class="st">"#69b3a2"</span>, <span class="at">color=</span><span class="st">"#e9ecef"</span>, <span class="at">alpha=</span><span class="fl">0.9</span>)</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>plot1 <span class="sc">+</span> plot2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-mt" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-mt-output-2.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;9: Histogram of mitochondrial (left) and chloroplastic (right) percentage of transcripts in each cell</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="counts-features-relationship" class="level4" data-number="2.4.1.4">
<h4 data-number="2.4.1.4" class="anchored" data-anchor-id="counts-features-relationship"><span class="header-section-number">2.4.1.4</span> Counts-Features relationship</h4>
<p>In <a href="#fig-gentra">Figure&nbsp;10</a> below, we look at the plot of the number of transcripts per cell vs the number of detected genes per cell. Usually, those two measure grow simultaneously. At lower counts the relationship is quite linear, then becomes a curve, typically bending in favour of the number of transcripts per cell. You can see below that each dot (representing a droplet) is coloured by percentage of mitochondria. Droplets with a high percentage of mitochondrial genes also have very low amount of transcripts and detected genes, <strong>confirming that high mitochondrial content is a measure of low quality</strong>.</p>
<div class="cell" data-tags="[]" data-execution_count="65">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">14</span>, <span class="at">repr.plot.height=</span><span class="dv">5</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> Control1_seurat<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(percent.mt)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>( meta, <span class="fu">aes</span>(<span class="at">x=</span>nCount_RNA, <span class="at">y=</span>nFeature_RNA, <span class="at">colour=</span>percent.mt)) <span class="sc">+</span> </span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>         <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.75</span>, <span class="at">size=</span><span class="dv">5</span>)<span class="sc">+</span></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>         <span class="fu">geom_smooth</span>(<span class="at">se=</span><span class="cn">TRUE</span>, <span class="at">method=</span><span class="st">"loess"</span>)</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>plot1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'
Warning message:
“The following aesthetics were dropped during statistical transformation: colour
ℹ This can happen when ggplot fails to infer the correct grouping structure in
  the data.
ℹ Did you forget to specify a `group` aesthetic or to convert a numerical
  variable into a factor?”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-gentra" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-gentra-output-2.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;10: Histogram of the relationship between detected genes and transcripts per cell, coloured by mitochondrial content.</figcaption>
</figure>
</div>
</div>
</div>
<p>In a similar way the chloroplastic genes confirm the pattern of low quality droplets.</p>
<div class="cell" data-tags="[]" data-execution_count="66">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">14</span>, <span class="at">repr.plot.height=</span><span class="dv">5</span>)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> Control1_seurat<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(percent.chloroplast)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>( meta, <span class="fu">aes</span>(<span class="at">x=</span>nCount_RNA, <span class="at">y=</span>nFeature_RNA, <span class="at">colour=</span>percent.chloroplast)) <span class="sc">+</span> </span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>         <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.75</span>, <span class="at">size=</span><span class="dv">5</span>)<span class="sc">+</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>         <span class="fu">geom_smooth</span>(<span class="at">se=</span><span class="cn">TRUE</span>, <span class="at">method=</span><span class="st">"loess"</span>)</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>plot1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'
Warning message:
“The following aesthetics were dropped during statistical transformation: colour
ℹ This can happen when ggplot fails to infer the correct grouping structure in
  the data.
ℹ Did you forget to specify a `group` aesthetic or to convert a numerical
  variable into a factor?”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-gentrach" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-gentrach-output-2.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;11: Histogram of the relationship between detected genes and transcripts per cell, coloured by chloroplastic content.</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="filtering-with-the-chosen-criteria" class="level3" data-number="2.4.2">
<h3 data-number="2.4.2" class="anchored" data-anchor-id="filtering-with-the-chosen-criteria"><span class="header-section-number">2.4.2</span> Filtering with the chosen criteria</h3>
<p>Here we use the command <code>subset</code> and impose the criteria we chose above looking at the histograms. We set each criteria for keeping cells of good quality using the names of the features in metadata. We print those names to remember them.</p>
<div class="cell" data-tags="[]" data-execution_count="67">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Meta data names:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( <span class="fu">names</span>(Control1_seurat<span class="sc">@</span>meta.data), <span class="at">sep=</span><span class="st">'; '</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Meta data names:
orig.ident; nCount_RNA; nFeature_RNA; Condition; percent.mt; percent.chloroplast</code></pre>
</div>
</div>
<p>The filtered object is called <code>Control1_seurat_filt</code></p>
<div class="cell" data-tags="[]" data-execution_count="68">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat_filt <span class="ot">&lt;-</span> <span class="fu">subset</span>(<span class="at">x =</span> Control1_seurat, </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">subset =</span> nCount_RNA <span class="sc">&gt;</span> <span class="dv">700</span> <span class="sc">&amp;</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>                                                 nCount_RNA <span class="sc">&lt;</span> <span class="dv">35000</span> <span class="sc">&amp;</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>                                                 nFeature_RNA <span class="sc">&gt;</span> <span class="dv">400</span> <span class="sc">&amp;</span> </span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>                                                 nFeature_RNA <span class="sc">&lt;</span> <span class="dv">7000</span> <span class="sc">&amp;</span> </span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>                                                 percent.mt <span class="sc">&lt;</span> <span class="dv">5</span> <span class="sc">&amp;</span> </span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>                                                 percent.chloroplast <span class="sc">&lt;</span> <span class="dv">5</span>)</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Filtered Genes and Cells: "</span>)</span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( <span class="fu">dim</span>(Control1_seurat) <span class="sc">-</span> <span class="fu">dim</span>(Control1_seurat_filt) )</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Remaining Genes and Cells: "</span>)</span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( <span class="fu">dim</span>(Control1_seurat_filt) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Filtered Genes and Cells: 0 2762
Remaining Genes and Cells: 23838 8010</code></pre>
</div>
</div>
<p>Now the transcripts vs genes can be seen in <a href="#fig-filtrelationship">Figure&nbsp;12</a>. The relationship is much more linear than previously after the removal of extreme values for transcripts and detected genes.</p>
<div class="cell" data-tags="[]" data-execution_count="69">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">14</span>, <span class="at">repr.plot.height=</span><span class="dv">5</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> Control1_seurat_filt<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(percent.mt)</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>( meta, <span class="fu">aes</span>(<span class="at">x=</span>nCount_RNA, <span class="at">y=</span>nFeature_RNA, <span class="at">colour=</span>percent.mt)) <span class="sc">+</span> </span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>         <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.75</span>, <span class="at">size=</span><span class="dv">5</span>)<span class="sc">+</span></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>         <span class="fu">geom_smooth</span>(<span class="at">se=</span><span class="cn">TRUE</span>, <span class="at">method=</span><span class="st">"loess"</span>)</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>plot1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'
Warning message:
“The following aesthetics were dropped during statistical transformation: colour
ℹ This can happen when ggplot fails to infer the correct grouping structure in
  the data.
ℹ Did you forget to specify a `group` aesthetic or to convert a numerical
  variable into a factor?”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-filtrelationship" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-filtrelationship-output-2.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12: Count vs features after filtering with the chosen criteria</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="normalization" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="normalization"><span class="header-section-number">2.5</span> Normalization</h2>
<p>scRNA-seq data is <strong>affected by highly variable RNA quantities and qualities</strong> across different cells. Furthermore, it is often subject to <strong>batch effects, sequencing depth differences, and other technical biases</strong> that can confound downstream analyses.</p>
<p>Normalization methods are used to <strong>adjust for these technical variations so that true biological differences between cells can be accurately identified</strong>.</p>
<p>Some commonly used normalization methods in scRNA-seq data include the following:</p>
<ul>
<li><strong>Total count normalization</strong>: Normalizing the read counts to the total number of transcripts in each sample</li>
<li><strong>TPM (transcripts per million)</strong> normalization: Normalizing the read counts to the total number of transcripts in each sample, scaled to a million</li>
<li><strong>Library size normalization:</strong> Normalizing the read counts to the total number of reads or transcripts in each sample, adjusted for sequencing depth</li>
</ul>
<p>All the above suffer from distorting some gene expressions, especially if the data varies a lot in term of sequencing depth. A new and more advanced method, at the moment the state-of-the-art, is <code>SCTransform</code> (<span class="citation" data-cites="hafemeister_normalization_2019">Hafemeister and Satija (<a href="#ref-hafemeister_normalization_2019" role="doc-biblioref">2019</a>)</span>), a software package that can <strong>correct for technical sources of variation and remove batch effects</strong>.</p>
<section id="finding-technical-sources-of-variation" class="level3" data-number="2.5.1">
<h3 data-number="2.5.1" class="anchored" data-anchor-id="finding-technical-sources-of-variation"><span class="header-section-number">2.5.1</span> Finding technical sources of variation</h3>
<p>Before normalizing we want to check for technical sources of variation in the data. One of those is the total number of transcripts: two similar cells might be sequenced at different depth. This influences of course normalization. The influence of the number of transcripts per cell is however always removed by <code>SCtransform</code>.</p>
<p>We want to look into other possible sources of variation. Those are usually quantities we calculate for each cell, for example the percentage of mitochondrial and chloroplastic genes.</p>
<p>To see if those quantities actually influence our data a lot, we check how much is their highest correlation with the first 10 components of the PCA of the dataset. In short, <strong>we see if any technical variation is such that it explains much of the variability of the data, covering possibly biological signal</strong>.</p>
<p>We now use the function <code>plotCorrelations</code> to plot the highest correlation of three quantities with the PCA: number of transcripts, percent of mitochondrial genes and percent of chloroplastic genes. You will see in <a href="#fig-corr">Figure&nbsp;13</a> how <strong>there is little correlation for the two percentages</strong>, for which we do not need to worry about, while <strong>there is correlation with the total number of transcripts per cell</strong> (this is always expected and, as mentioned before, is removed automatically by the normalization process). We created the function <code>plotCorrelations</code> specifically for the course, together with a few others, mostly for plotting or handling tables. You can find them in the file <code>script.R</code>.</p>
<div id="fig-corr" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotCorrelations</span>( <span class="at">object=</span>Control1_seurat, <span class="at">measures=</span><span class="fu">c</span>(<span class="st">'nCount_RNA'</span>, <span class="st">'percent.mt'</span>, <span class="st">'percent.chloroplast'</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'
`geom_smooth()` using formula = 'y ~ x'
`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-corr-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-corr-output-2.png" class="img-fluid figure-img" data-ref-parent="fig-corr"></p>
<figcaption class="figure-caption">(a) Relationships of maximal correlation between cell quantities and principal components</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div id="fig-corr-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-corr-output-3.png" class="img-fluid figure-img" data-ref-parent="fig-corr"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div id="fig-corr-3" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-corr-output-4.png" class="img-fluid figure-img" data-ref-parent="fig-corr"></p>
<figcaption class="figure-caption">(c)</figcaption>
</figure>
</div>
</div>
<figcaption class="figure-caption">Figure&nbsp;13: <strong>?(caption)</strong></figcaption>
</figure>
</div>
</section>
<section id="executing-normalization" class="level3" data-number="2.5.2">
<h3 data-number="2.5.2" class="anchored" data-anchor-id="executing-normalization"><span class="header-section-number">2.5.2</span> Executing normalization</h3>
<p>We run <code>SCtransform</code> normalization below. Here you can choose to subsample some cells to do the normalization (<code>ncells</code> option): this is <strong>useful to avoid ending up waiting for a long time</strong>. A few thousands cells is enough.</p>
<p>You can also choose how many genes to consider for normalization (<code>variable.features.n</code> option): in this case it is best to <strong>use the genes that vary the most their expression across cells</strong>. We look at a histogram (<a href="#fig-hvg">Figure&nbsp;14</a>) of the variance of each gene to choose a threshold to identify highly-variable genes.</p>
<div class="cell" data-execution_count="71">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>variance_genes <span class="ot">&lt;-</span> <span class="fu">apply</span>( <span class="fu">as.matrix</span>(Control1_seurat_filt[[<span class="st">'RNA'</span>]]<span class="sc">@</span>counts), <span class="dv">1</span>, var)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">14</span>, <span class="at">repr.plot.height=</span><span class="dv">5</span>)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">data.frame</span>(variance_genes), <span class="fu">aes</span>(variance_genes)) <span class="sc">+</span> </span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_histogram</span>(<span class="at">fill=</span><span class="st">"#69b3a2"</span>, <span class="at">color=</span><span class="st">"#e9ecef"</span>, <span class="at">alpha=</span><span class="fl">0.9</span>) <span class="sc">+</span> <span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>plot1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message in asMethod(object):
“sparse-&gt;dense coercion: allocating vector of size 1.4 GiB”
`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
Warning message:
“Removed 3289 rows containing non-finite values (`stat_bin()`).”
Warning message:
“Removed 2 rows containing missing values (`geom_bar()`).”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-hvg" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-hvg-output-2.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;14: Histogram of genes variance. We choose the threshold 0.1 to identify highly variable genes.</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-execution_count="72">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>hvighly_var_genes <span class="ot">&lt;-</span> variance_genes <span class="sc">&gt;</span> .<span class="dv">25</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"The total number of highly variable genes selected is: "</span>)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">sum</span>( hvighly_var_genes ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The total number of highly variable genes selected is: 6652</code></pre>
</div>
</div>
<div class="cell" data-tags="[]" data-execution_count="73">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat_norm <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(Control1_seurat_filt, </span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>                                             <span class="at">return.only.var.genes =</span> <span class="cn">FALSE</span>, </span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>                                             <span class="at">ncells =</span> <span class="dv">3000</span>, </span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>                                             <span class="at">variable.features.n =</span> <span class="fu">sum</span>( hvighly_var_genes ),</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>                                             <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Normalized data is now in the object <code>Control1_seurat_norm</code>, in a new assay called <code>SCT</code>. This assay is now the default used for data analysis: you can verify it very easily below:</p>
<div class="cell" data-execution_count="74">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Your default assay is: "</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">DefaultAssay</span>(<span class="at">object =</span> Control1_seurat_norm))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Your default assay is: SCT</code></pre>
</div>
</div>
</section>
<section id="visualizing-the-result" class="level3" data-number="2.5.3">
<h3 data-number="2.5.3" class="anchored" data-anchor-id="visualizing-the-result"><span class="header-section-number">2.5.3</span> Visualizing the result</h3>
<p>Now we plot the UMAP plot of the data to have a first impression of how the data is structured (presence of clusters, how many, etc.). First of all, we create a PCA plot, which tells us how many PCA components are of relevance with the elbow plot of <a href="#fig-elbow">Figure&nbsp;15</a>. In the elbow plot, we see the variability of each component in descending order. Note how, after a few rapidly descending components, there is an elbow. We schoose a threshold just after the elbow (for example at 15), which means those components will be used to calculate some other things of relevance in the data, such as distance between cells and the UMAP projection of <a href="#fig-umap">Figure&nbsp;16</a>: specific commands using PCA allow to choose the components, and we will set 10 with the option <code>dims = 1:15</code>.</p>
<div class="cell" data-tags="[]" data-execution_count="75">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat_norm <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(Control1_seurat_norm,</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>                                                     <span class="at">nfeatures =</span> <span class="fu">sum</span>( hvighly_var_genes ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="76">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat_norm <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(<span class="at">object =</span> Control1_seurat_norm, </span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">seed.use =</span> <span class="dv">123</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="77">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(Control1_seurat_norm, <span class="at">ndims =</span> <span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-elbow" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-elbow-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;15: Elbow plot of the first 30 principal components calculated from the data</figcaption>
</figure>
</div>
</div>
</div>
<p>We calculate the projection using the UMAP algorithm (<span class="citation" data-cites="mcinnes_umap_2018">McInnes et al. (<a href="#ref-mcinnes_umap_2018" role="doc-biblioref">2018</a>)</span>, <span class="citation" data-cites="becht_dimensionality_2019">Becht et al. (<a href="#ref-becht_dimensionality_2019" role="doc-biblioref">2019</a>)</span>). The parameters <code>a</code> and <code>b</code> will change how stretched and scattered the data looks like. When you do your own UMAP projection, you can avoid setting <code>a</code> and <code>b</code>, and those will be chosen automatically by the command.</p>
<div class="cell" data-tags="[]" data-execution_count="78">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat_norm <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(<span class="at">object =</span> Control1_seurat_norm, </span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">a =</span> .<span class="dv">8</span>, <span class="at">b=</span><span class="dv">1</span>,</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, </span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">verbose =</span> <span class="cn">FALSE</span>, </span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">seed.use =</span> <span class="dv">123</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'

Also defined by ‘BiocGenerics’

Found more than one class "dist" in cache; using the first, from namespace 'spam'

Also defined by ‘BiocGenerics’
</code></pre>
</div>
</div>
<p>In <a href="#fig-umap">Figure&nbsp;16</a> we can see the resulting projection. The result looks pretty neat and structured (we can clearly see there are various clusters).</p>
<div class="cell" data-tags="[]" data-execution_count="79">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">10</span>, <span class="at">repr.plot.height=</span><span class="dv">8</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">UMAPPlot</span>(<span class="at">object =</span> Control1_seurat_norm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-umap" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-umap-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;16: UMAP projection of the data</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="removing-doublets" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="removing-doublets"><span class="header-section-number">2.6</span> Removing doublets</h2>
<p>Doublets removal is part of filtering, but it needs normalized data to work. This is why we do it after using <code>SCtransform</code>.</p>
<p>Doublets (and the very rare multiplets) refer to droplets that <strong>contain the transcriptional profiles of two or more distinct cells</strong>. Doublets can occur during the cell dissociation process or when two or more cells are captured in the same droplet during the library preparation step.</p>
<p>It is quite obvious that a doublet transcriptional profile can confound downstream analyses, such as cell clustering and differential gene expression analysis. Most doublet detectors, like <code>DoubletFinder</code> (<span class="citation" data-cites="mcginnis_doubletfinder_2019">McGinnis, Murrow, and Gartner (<a href="#ref-mcginnis_doubletfinder_2019" role="doc-biblioref">2019</a>)</span>) which we will use, <strong>simulates doublets and then finds cells in the data which are similar to the simulated doublets</strong>. Most such packages need an idea of the number/proportion of expected doublets in the dataset. <strong>As indicated from the Chromium user guide, expected doublet rates are about as follows:</strong></p>
<div id="fig-doubletrates" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="./images/doublet_rates.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;17: Table of expected doublet rates based on the number of cells.</figcaption>
</figure>
</div>
<p>The data we are using contained about 10000 cells per sample (as in the knee plot at the beginning), hence we can assume that it originates from around 18000 loaded cells and should have a doublet rate at about 7.6%.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Doublet prediction, like the rest of the filtering, <strong>should be run on each sample separately</strong>.</p>
</div>
</div>
<p>Here, we apply <code>DoubletFinder</code> to predict doublet cells. Most parameters are quite standard, we mostly need to choose <code>nExp</code> (expected number of doublets), <code>PCs</code> (number of principal components to use), <code>sct</code> (use the normalized data). The last three option are not part of the package, but have been added by creating a slightly modified version (<a href="https://github.com/SamueleSoraggi/DoubletFinder">here</a>) - they allow to use multiple cores and a subset of cells for calculations for a considerable speedup. However, the code takes some time to run, so be patient. There will be a lot of printout as well, but don’t worry.</p>
<div class="cell" data-tags="[]" data-execution_count="80">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>nExp <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">ncol</span>(Control1_seurat_norm) <span class="sc">*</span> <span class="fl">0.076</span>)  <span class="co"># expected doublet rate</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="81">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat_norm <span class="ot">&lt;-</span> <span class="fu">doubletFinder_v3</span>(Control1_seurat_norm,</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">pN =</span> <span class="fl">0.25</span>, <span class="co">#proportion of doublets to simulate)</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">pK =</span> <span class="fl">0.09</span>, </span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">nExp =</span> nExp, </span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">PCs =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, </span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">sct=</span><span class="cn">TRUE</span>, </span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">workers=</span><span class="dv">8</span>, </span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">future.globals.maxSize =</span> <span class="dv">8</span><span class="sc">*</span><span class="dv">1024</span><span class="sc">^</span><span class="dv">13</span>,</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">seurat.ncells=</span><span class="dv">3000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: fields

Loading required package: spam

Spam version 2.8-0 (2022-01-05) is loaded.
Type 'help( Spam)' or 'demo( spam)' for a short introduction 
and overview of this package.
Help for individual functions is also obtained by adding the
suffix '.spam' to the function name, e.g. 'help( chol.spam)'.


Attaching package: ‘spam’


The following object is masked from ‘package:stats4’:

    mle


The following objects are masked from ‘package:base’:

    backsolve, forwardsolve


Loading required package: viridisLite


Try help(fields) to get started.

Loading required package: KernSmooth

KernSmooth 2.23 loaded
Copyright M. P. Wand 1997-2009

Loading required package: sctransform

Calculating cell attributes from input UMI matrix: log_umi

Variance stabilizing transformation of count matrix of size 23388 by 10680

Model formula is y ~ log_umi

Get Negative Binomial regression parameters per gene

Using 2000 genes, 3000 cells

Found 151 outliers - those will be ignored in fitting/regularization step


Second step: Get residuals using fitted parameters for 23388 genes

Computing corrected count matrix for 23388 genes

Calculating gene attributes

Wall clock passed: Time difference of 1.02049 mins

Determine variable features

Place corrected count matrix in counts slot

Centering data matrix

Set default assay to SCT

PC_ 1 
Positive:  LotjaGi2g1v0360900, LotjaGi5g1v0359500, LotjaGi6g1v0155900, LotjaGi6g1v0155800, LotjaGi3g1v0321700, LotjaGi3g1v0414900, LotjaGi3g1v0030500, LotjaGi5g1v0211100, LotjaGi3g1v0506700, LotjaGi4g1v0109600 
       LotjaGi3g1v0009600, LotjaGi1g1v0539300, LotjaGi3g1v0450900, LotjaGi2g1v0269100, LotjaGi3g1v0010900, LotjaGi3g1v0530000, LotjaGi3g1v0373700, LotjaGi4g1v0137700, LotjaGi3g1v0380900, LotjaGi4g1v0309700 
       LotjaGi1g1v0014300, LotjaGi5g1v0359400, LotjaGi6g1v0071000, LotjaGi3g1v0012400, LotjaGi3g1v0162300, LotjaGi3g1v0554100, LotjaGi1g1v0516900, LotjaGi2g1v0316800, LotjaGi2g1v0285600, LotjaGi2g1v0358600 
Negative:  LotjaGi6g1v0254300, LotjaGi3g1v0068000, LotjaGi6g1v0286800-LC, LotjaGi1g1v0080000, LotjaGi5g1v0005800, LotjaGi5g1v0269800-LC, LotjaGi5g1v0293100-LC, LotjaGi3g1v0222100, LotjaGi1g1v0006200, LotjaGi6g1v0254700 
       LotjaGi3g1v0358300, LotjaGi3g1v0329100, LotjaGi3g1v0445300, LotjaGi1g1v0646500-LC, LotjaGi2g1v0157900, LotjaGi3g1v0505900, LotjaGi1g1v0022100, LotjaGi4g1v0076500, LotjaGi4g1v0293000-LC, LotjaGi1g1v0558200 
       LotjaGi1g1v0577100, LotjaGi3g1v0395900-LC, LotjaGi5g1v0031100, LotjaGi5g1v0288600, LotjaGi3g1v0097800, LotjaGi1g1v0261700, LotjaGi4g1v0207600, LotjaGi4g1v0313900, LotjaGi2g1v0303000, LotjaGi1g1v0515200 
PC_ 2 
Positive:  LotjaGi3g1v0358300, LotjaGi6g1v0254300, LotjaGi1g1v0646500-LC, LotjaGi3g1v0038800, LotjaGi6g1v0253800, LotjaGi6g1v0255000, LotjaGi1g1v0006200, LotjaGi6g1v0254700, LotjaGi5g1v0089300, LotjaGi6g1v0022500 
       LotjaGi3g1v0329100, LotjaGi6g1v0043900, LotjaGi1g1v0261700, LotjaGi4g1v0313900, LotjaGi2g1v0303000, LotjaGi5g1v0005800, LotjaGi1g1v0277900, LotjaGi6g1v0022600, LotjaGi4g1v0325700, LotjaGi3g1v0222100 
       LotjaGi1g1v0690000, LotjaGi6g1v0155800, LotjaGi4g1v0293000-LC, LotjaGi2g1v0360900, LotjaGi1g1v0555200, LotjaGi4g1v0284700, LotjaGi3g1v0046800, LotjaGi1g1v0336600, LotjaGi2g1v0239200, LotjaGi2g1v0388100 
Negative:  LotjaGi5g1v0248500, LotjaGi1g1v0405300, LotjaGi1g1v0074900, LotjaGi1g1v0683300, LotjaGi2g1v0160200, LotjaGi4g1v0018700-LC, LotjaGi5g1v0163900, LotjaGi6g1v0315500, LotjaGi6g1v0078500, LotjaGi2g1v0221100 
       LotjaGi2g1v0019900, LotjaGi4g1v0217400, LotjaGi5g1v0248600, LotjaGi4g1v0256800, LotjaGi2g1v0002800-LC, LotjaGi3g1v0204100, LotjaGi2g1v0426500, LotjaGi4g1v0297800, LotjaGi6g1v0246700, LotjaGi1g1v0109100 
       LotjaGi1g1v0109000, LotjaGi3g1v0493400-LC, LotjaGi5g1v0159400, LotjaGi3g1v0086100-LC, LotjaGi5g1v0120700, LotjaGi1g1v0430800-LC, LotjaGi6g1v0292800, LotjaGi1g1v0593900, LotjaGi5g1v0249500, LotjaGi6g1v0216300 
PC_ 3 
Positive:  LotjaGi3g1v0445300, LotjaGi3g1v0505900, LotjaGi1g1v0594900, LotjaGi1g1v0502700, LotjaGi4g1v0275500, LotjaGi1g1v0723600-LC, LotjaGi6g1v0151500, LotjaGi4g1v0208100, LotjaGi5g1v0269800-LC, LotjaGi2g1v0163300 
       LotjaGi6g1v0028000-LC, LotjaGi4g1v0207600, LotjaGi3g1v0115600, LotjaGi1g1v0475000-LC, LotjaGi2g1v0163500, LotjaGi4g1v0207900, LotjaGi1g1v0022100, LotjaGi1g1v0348000, LotjaGi2g1v0176500-LC, LotjaGi2g1v0406200 
       LotjaGi5g1v0266100, LotjaGi2g1v0402200, LotjaGi3g1v0359600, LotjaGi3g1v0506700, LotjaGi6g1v0286800-LC, LotjaGi5g1v0211100, LotjaGi3g1v0414900, LotjaGi4g1v0014800, LotjaGi5g1v0296300-LC, LotjaGi1g1v0659300 
Negative:  LotjaGi1g1v0405300, LotjaGi5g1v0248500, LotjaGi4g1v0018700-LC, LotjaGi1g1v0683300, LotjaGi1g1v0074900, LotjaGi5g1v0163900, LotjaGi3g1v0218300, LotjaGi2g1v0221100, LotjaGi2g1v0019900, LotjaGi6g1v0078500 
       LotjaGi4g1v0217400, LotjaGi4g1v0256800, LotjaGi2g1v0160200, LotjaGi3g1v0204100, LotjaGi3g1v0068000, LotjaGi5g1v0248600, LotjaGi6g1v0253800, LotjaGi3g1v0038800, LotjaGi1g1v0109100, LotjaGi3g1v0493400-LC 
       LotjaGi6g1v0315500, LotjaGi3g1v0086100-LC, LotjaGi3g1v0358300, LotjaGi6g1v0246700, LotjaGi2g1v0002800-LC, LotjaGi5g1v0249500, LotjaGi6g1v0043900, LotjaGi2g1v0426500, LotjaGi2g1v0429600, LotjaGi6g1v0255000 
PC_ 4 
Positive:  LotjaGi5g1v0005800, LotjaGi1g1v0558200, LotjaGi5g1v0288600, LotjaGi3g1v0174100, LotjaGi3g1v0222100, LotjaGi3g1v0068000, LotjaGi2g1v0386600, LotjaGi5g1v0166000-LC, LotjaGi6g1v0286800-LC, LotjaGi3g1v0162600 
       LotjaGi4g1v0121800, LotjaGi3g1v0395900-LC, LotjaGi2g1v0126700, LotjaGi5g1v0099800, LotjaGi4g1v0293000-LC, LotjaGi4g1v0431800, LotjaGi3g1v0178400, LotjaGi1g1v0636800, LotjaGi4g1v0076500, LotjaGi6g1v0012100 
       LotjaGi2g1v0368200, LotjaGi1g1v0393600, LotjaGi1g1v0114400, LotjaGi4g1v0064700, LotjaGi5g1v0003500, LotjaGi1g1v0340500, LotjaGi3g1v0112700, LotjaGi3g1v0191200, LotjaGi6g1v0069200, LotjaGi3g1v0192300 
Negative:  LotjaGi5g1v0269800-LC, LotjaGi5g1v0120700, LotjaGi3g1v0420400, LotjaGi2g1v0157900, LotjaGi1g1v0377600, LotjaGi6g1v0253800, LotjaGi6g1v0254700, LotjaGi5g1v0293100-LC, LotjaGi1g1v0613100, LotjaGi1g1v0646500-LC 
       LotjaGi1g1v0022100, LotjaGi2g1v0303000, LotjaGi3g1v0055400, LotjaGi2g1v0176500-LC, LotjaGi5g1v0119900, LotjaGi1g1v0515200, LotjaGi2g1v0163500, LotjaGi3g1v0445300, LotjaGi3g1v0420800, LotjaGi3g1v0046800 
       LotjaGi3g1v0115400, LotjaGi3g1v0505900, LotjaGi1g1v0723600-LC, LotjaGi1g1v0690000, LotjaGi1g1v0475000-LC, LotjaGi4g1v0207600, LotjaGi3g1v0329100, LotjaGi1g1v0277900, LotjaGi4g1v0060000, LotjaGi3g1v0420000 
PC_ 5 
Positive:  LotjaGi3g1v0328800, LotjaGi3g1v0222100, LotjaGi4g1v0417500, LotjaGi3g1v0328900, LotjaGi2g1v0450600, LotjaGi3g1v0201800, LotjaGi4g1v0293000-LC, LotjaGi3g1v0378600, LotjaGi5g1v0166000-LC, LotjaGi1g1v0353400 
       LotjaGi3g1v0395900-LC, LotjaGi6g1v0069200, LotjaGi6g1v0155800, LotjaGi6g1v0286800-LC, LotjaGi3g1v0546100, LotjaGi3g1v0445300, LotjaGi6g1v0326300, LotjaGi3g1v0162300, LotjaGi5g1v0288600, LotjaGi5g1v0031100 
       LotjaGi1g1v0601700, LotjaGi3g1v0174100, LotjaGi1g1v0594900, LotjaGi5g1v0266100, LotjaGi3g1v0505900, LotjaGi4g1v0284700, LotjaGi1g1v0393600, LotjaGi6g1v0043900, LotjaGi4g1v0269400, LotjaGi1g1v0795300 
Negative:  LotjaGi4g1v0121800, LotjaGi3g1v0068000, LotjaGi5g1v0099800, LotjaGi3g1v0178400, LotjaGi2g1v0368200, LotjaGi6g1v0012100, LotjaGi1g1v0114400, LotjaGi2g1v0452500, LotjaGi1g1v0240900-LC, LotjaGi5g1v0003500 
       LotjaGi3g1v0192300, LotjaGi2g1v0301500, LotjaGi4g1v0300800-LC, LotjaGi3g1v0162600, LotjaGi5g1v0120700, LotjaGi5g1v0102500, LotjaGi4g1v0298100, LotjaGi5g1v0100000, LotjaGi3g1v0506700, LotjaGi2g1v0345700 
       LotjaGi1g1v0137600, LotjaGi3g1v0001800, LotjaGi1g1v0760600, LotjaGi5g1v0094100, LotjaGi1g1v0221300, LotjaGi4g1v0431800, LotjaGi4g1v0455200, LotjaGi6g1v0232200, LotjaGi4g1v0064700, LotjaGi1g1v0300600 
</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Creating 2670 artificial doublets..."
[1] "Creating Seurat object..."
[1] "Running SCTransform..."
  |======================================================================| 100%
  |======================================================================| 100%
  |======================================================================| 100%
[1] "Running PCA..."
[1] "Calculating PC distance matrix..."
[1] "Computing pANN..."
[1] "Classifying doublets.."</code></pre>
</div>
</div>
<p>We visualize the UMAP plot and which cells are estimated doublets in <a href="#fig-doublets">Figure&nbsp;18</a>. Fortunately, there are only a few potential doublets.</p>
<div class="cell" data-tags="[]" data-execution_count="82">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">10</span>, <span class="at">repr.plot.height=</span><span class="dv">10</span>)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>DF.name <span class="ot">=</span> <span class="fu">colnames</span>(Control1_seurat_norm<span class="sc">@</span>meta.data)[<span class="fu">grepl</span>(<span class="st">"DF.classification"</span>, <span class="fu">colnames</span>(Control1_seurat_norm<span class="sc">@</span>meta.data))]</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(Control1_seurat_norm, <span class="at">group.by =</span> DF.name, <span class="at">pt.size =</span> <span class="dv">2</span>, </span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> DF.name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-doublets" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-doublets-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;18: UMAP projection of the data coloured by doublet or singlet label</figcaption>
</figure>
</div>
</div>
</div>
<p>Sometimes doublets have more detected genes than a single cell. In our case, some of the droplets have higher number of genes than the average (the red violin is large also above 3000 detected genes), so there is aclear sign of the presence of some doublets. Of course, as with any filtering, we might remove some actual cells. To be more effective in our filtering, we can select doublets with more than 2000 detected genes when we filter.</p>
<div class="cell" data-tags="[]" data-execution_count="83">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(Control1_seurat_norm, <span class="at">features =</span> <span class="st">"nFeature_RNA"</span>, <span class="at">group.by =</span> DF.name, <span class="at">pt.size =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="notebook.source_files/figure-html/cell-43-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Here we keep only singlets:</p>
<div class="cell" data-tags="[]" data-execution_count="84">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat_norm <span class="ot">=</span> Control1_seurat_norm[, (Control1_seurat_norm<span class="sc">@</span>meta.data[, DF.name] <span class="sc">==</span> <span class="st">"Singlet"</span>)<span class="sc">&amp;</span>(Control1_seurat_norm<span class="sc">@</span>meta.data<span class="sc">$</span>nFeature_RNA<span class="sc">&gt;</span><span class="dv">2000</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We save our data after all the filtering work!</p>
<div class="cell" data-tags="[]" data-execution_count="85">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SaveH5Seurat</span>(<span class="at">object =</span> Control1_seurat_norm, </span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">filename =</span> <span class="st">"control1.normalized.h5Seurat"</span>, </span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">overwrite =</span> <span class="cn">TRUE</span>,</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Overwriting previous file control1.normalized.h5Seurat”
Creating h5Seurat file for version 3.1.5.9900
</code></pre>
</div>
</div>
</section>
</section>
<section id="sec-integration" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Integration</h1>
<p>Integration of scRNA-seq data is useful to combine datasets from different experimental conditions (in our case the Control vs Infected) and sequencing runs, to gain a broader understanding of cellular processes. Integration is challenging due to technical variations and biological differences between the datasets (where we want to remove the formers to study correctly the latters).</p>
<p>Before integrating scRNA-seq datasets, we have applyed <strong>quality control and normalization to each sample</strong> to ensure consistency and accuracy of the data. Integration can happen using various methods (<span class="citation" data-cites="adossa_computational_2021">Adossa et al. (<a href="#ref-adossa_computational_2021" role="doc-biblioref">2021</a>)</span>). Seurat uses <strong>canonical correlation analysis (CCA)</strong> (<span class="citation" data-cites="stuart_comprehensive_2019">Stuart et al. (<a href="#ref-stuart_comprehensive_2019" role="doc-biblioref">2019</a>)</span>, <span class="citation" data-cites="xinming_seurat_2022">Xinming (<a href="#ref-xinming_seurat_2022" role="doc-biblioref">2022</a>)</span>) to integrate scRNA-seq datasets from different experimental conditions. CCA identifies shared variation between two datasets while accounting for technical differences, such as batch effects.</p>
<p>The shared covariance patterns <strong>can represent biological signals that are common across the datasets</strong>, such as cell types or signaling pathways.</p>
<p>We load another control and two infected datasets. Those have been previously preprocessed, so you will not need to. Remember again: each dataset must be preprocessed separately before integration.</p>
<div class="cell" data-tags="[]" data-execution_count="86">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>Control1_seurat_norm <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">"control1.normalized.h5Seurat"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Validating h5Seurat file
</code></pre>
</div>
</div>
<div class="cell" data-tags="[]" data-execution_count="87">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>Control2_seurat_norm <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">"../Data/control2.normalized.h5Seurat"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>Infected1_seurat_norm <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">"../Data/infected1.normalized.h5Seurat"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>Infected2_seurat_norm <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">"../Data/infected2.normalized.h5Seurat"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Validating h5Seurat file

Validating h5Seurat file

Validating h5Seurat file
</code></pre>
</div>
</div>
<p>To integrate the datasets, we need to start creating a list with all datasets.</p>
<div class="cell" data-tags="[]" data-execution_count="88">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>Gifu.list <span class="ot">&lt;-</span> <span class="fu">list</span>(Control1_seurat_norm, </span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>                  Control2_seurat_norm, </span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>                  Infected1_seurat_norm, </span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>                  Infected2_seurat_norm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then start by normalizing each dataset of the list with <code>SCtransform</code>.</p>
<div class="cell" data-tags="[]" data-execution_count="89">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>Gifu.list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="at">X =</span> Gifu.list, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">"Normalizing</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(x, <span class="at">ncells=</span><span class="dv">3000</span>, <span class="at">variable.features.n =</span> <span class="dv">2000</span>, <span class="at">return.only.var.genes =</span> <span class="cn">FALSE</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing


Normalizing


Normalizing


Normalizing

</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Running <code>SCTransform</code> as above does not contain any variable to regress out. By definition, the normalization will remove the technical effect of the number of transcripts per cell only. This is important because the amount of transcripts greatly vary in each cell, and it might seem a sign of biological variation, rather than a sign of varying capture efficiency of the mRNA transcripts. If you want to remove other sources of technical variation, for example chloroplastic and mitochondrial transcripts percentage, then you can use this command</p>
<pre><code>x &lt;- SCTransform(x, vars.to.regress = c("percent.mt",  "percent.chloroplast"),
                 variable.features.n = 10000,
                 return.only.var.genes = FALSE,
                 verbose = TRUE)</code></pre>
<p>You can also include differences due to biological variation which you want to remove, to highlight the effect of other biological processes. Find the <a href="https://satijalab.org/seurat/reference/sctransform">manual of SCTransform</a> to understand all possible options of the command. <a href="https://hbctraining.github.io/scRNA-seq/lessons/06_SC_SCT_and_integration.html">This other tutorial</a> has a good application of <code>SCTransform</code> which you can read.</p>
</div>
</div>
<p>Now we apply the CCA (Canonical Correlation Analysis) to put datasets together according to their similarities, while removing differences. The number of genes to use during integration is expressed below as <code>nfeatures</code>. The default choice is 2000 as written below.</p>
<div class="cell" data-tags="[]" data-execution_count="90">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>Gifu.features <span class="ot">&lt;-</span> <span class="fu">SelectIntegrationFeatures</span>(<span class="at">object.list =</span> Gifu.list, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>Gifu.list <span class="ot">&lt;-</span> <span class="fu">PrepSCTIntegration</span>(<span class="at">object.list =</span> Gifu.list, <span class="at">anchor.features =</span> Gifu.features)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="91">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>Gifu.anchors <span class="ot">&lt;-</span> <span class="fu">FindIntegrationAnchors</span>(<span class="at">object.list =</span> Gifu.list, <span class="at">normalization.method =</span> <span class="st">"SCT"</span>, </span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">anchor.features =</span> Gifu.features, <span class="at">reference =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>seurat.integrated <span class="ot">&lt;-</span> <span class="fu">IntegrateData</span>(<span class="at">anchorset =</span> Gifu.anchors, <span class="at">normalization.method =</span> <span class="st">"SCT"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message in CheckDuplicateCellNames(object.list = object.list):
“Some cell names are duplicated across objects provided. Renaming to enforce unique cell names.”
Finding anchors between all query and reference datasets

Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Running CCA

Merging objects

Finding neighborhoods

Finding anchors

    Found 8720 anchors

Filtering anchors

    Retained 8558 anchors

Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Running CCA

Merging objects

Finding neighborhoods

Finding anchors

    Found 8786 anchors

Filtering anchors

    Retained 8086 anchors

Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Running CCA

Merging objects

Finding neighborhoods

Finding anchors

    Found 10173 anchors

Filtering anchors

    Retained 9700 anchors

Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Running CCA

Merging objects

Finding neighborhoods

Finding anchors

    Found 8786 anchors

Filtering anchors

    Retained 8086 anchors

Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Running CCA

Merging objects

Finding neighborhoods

Finding anchors

    Found 10173 anchors

Filtering anchors

    Retained 9700 anchors

Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Building integrated reference

Merging dataset 1 into 2

Extracting anchors for merged samples

Finding integration vectors

Finding integration vector weights

Integrating data

Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”

Integrating dataset 3 with reference dataset

Finding integration vectors

Finding integration vector weights

Integrating data

Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to "ignore".”

Integrating dataset 4 with reference dataset

Finding integration vectors

Finding integration vector weights

Integrating data

Warning message:
“Layer counts isn't present in the assay object; returning NULL”
Warning message:
“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-2’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to "ignore".”
Warning message:
“Layer counts isn't present in the assay object; returning NULL”</code></pre>
</div>
</div>
<p>Now the default assay used for analysis has changed into <code>integrated</code>:</p>
<div class="cell" data-tags="[]" data-execution_count="92">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"The default assay of the data is now called: "</span>)</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">DefaultAssay</span>(seurat.integrated))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The default assay of the data is now called: integrated</code></pre>
</div>
</div>
<p>We need to recalculate PCA and UMAP to look at all datasets integrated together. We choose again 10 principal components from <a href="#fig-elbowint">Figure&nbsp;19</a>. The newUMAP is in <a href="#fig-umapint">Figure&nbsp;20</a>.</p>
<div class="cell" data-tags="[]" data-execution_count="93">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>seurat.integrated <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(<span class="at">object =</span> seurat.integrated, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="94">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(seurat.integrated, <span class="at">ndims =</span> <span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-elbowint" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-elbowint-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;19: Elbow plot of integrated datasets</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-tags="[]" data-execution_count="95">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>seurat.integrated <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(<span class="at">object =</span> seurat.integrated, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">k.param =</span> <span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph

Computing SNN
</code></pre>
</div>
</div>
<div class="cell" data-scrolled="true" data-tags="[]" data-execution_count="96">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>seurat.integrated <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(<span class="at">object =</span> seurat.integrated, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">a=</span>.<span class="dv">8</span>, <span class="at">b=</span>.<span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'

Also defined by ‘BiocGenerics’

13:34:01 Read 20572 rows and found 20 numeric columns

13:34:01 Using Annoy for neighbor search, n_neighbors = 30

Found more than one class "dist" in cache; using the first, from namespace 'spam'

Also defined by ‘BiocGenerics’

13:34:01 Building Annoy index with metric = cosine, n_trees = 50

0%   10   20   30   40   50   60   70   80   90   100%

[----|----|----|----|----|----|----|----|----|----|

*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
*
|

13:34:03 Writing NN index file to temp file /tmp/RtmpnH6zJf/file4392b2ad167

13:34:03 Searching Annoy index using 8 threads, search_k = 3000

13:34:04 Annoy recall = 100%

13:34:05 Commencing smooth kNN distance calibration using 8 threads
 with target n_neighbors = 30

13:34:07 Initializing from normalized Laplacian + noise (using irlba)

13:34:18 Commencing optimization for 200 epochs, with 757466 positive edges

13:34:28 Optimization finished
</code></pre>
</div>
</div>
<div class="cell" data-execution_count="97">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>seurat.integrated <span class="ot">&lt;-</span> <span class="fu">SetIdent</span>(seurat.integrated, <span class="at">value =</span> <span class="st">"orig.ident"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="98">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">10</span>, <span class="at">repr.plot.height=</span><span class="dv">8</span>)</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(<span class="at">object =</span> seurat.integrated, <span class="at">reduction =</span> <span class="st">"umap"</span>,  <span class="at">label =</span> T, <span class="at">repel =</span> <span class="cn">TRUE</span>, <span class="at">pt.size =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-umapint" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-umapint-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;20: UMAP of the integrated datasets.</figcaption>
</figure>
</div>
</div>
</div>
<p>We save our integrated data</p>
<div class="cell" data-tags="[]" data-execution_count="101">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SaveH5Seurat</span>(<span class="at">object =</span> seurat.integrated, <span class="at">filename =</span> <span class="st">"seurat.integrated.h5Seurat"</span>, <span class="at">overwrite=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Overwriting previous file seurat.integrated.h5Seurat”
Creating h5Seurat file for version 3.1.5.9900

Adding counts for SCT

Adding data for SCT

Adding scale.data for SCT

No variable features found for SCT

No feature-level metadata found for SCT

Writing out SCTModel.list for SCT

Adding counts for RNA

Adding data for RNA

No variable features found for RNA

No feature-level metadata found for RNA

Adding data for integrated

Adding scale.data for integrated

Adding variable features for integrated

No feature-level metadata found for integrated

Writing out SCTModel.list for integrated

Adding cell embeddings for pca

Adding loadings for pca

No projected loadings for pca

Adding standard deviations for pca

No JackStraw data for pca

Adding cell embeddings for umap

No loadings for umap

No projected loadings for umap

No standard deviations for umap

No JackStraw data for umap
</code></pre>
</div>
</div>
<section id="clustering-and-cell-type-assignment" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="clustering-and-cell-type-assignment"><span class="header-section-number">3.1</span> Clustering and cell type assignment</h2>
<p>We perform clustering on the data using the leiden algorithm (blondel_fast_2008, <span class="citation" data-cites="traag_louvain_2019">Traag, Waltman, and Van Eck (<a href="#ref-traag_louvain_2019" role="doc-biblioref">2019</a>)</span>). Then, we look at a typical strategy of <strong>naming clusters by visualizing known markers</strong>. Since this is very subjective and biased, we then resort to naming cell types <strong>using a reference annotated dataset</strong>. An overview of cell type assignment procedures can be found at <span class="citation" data-cites="cheng_review_2023">Cheng et al. (<a href="#ref-cheng_review_2023" role="doc-biblioref">2023</a>)</span>.</p>
<div class="cell" data-execution_count="102">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>seurat.integrated <span class="ot">&lt;-</span>  <span class="fu">LoadH5Seurat</span>(<span class="st">"seurat.integrated.h5Seurat"</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Validating h5Seurat file

Warning message:
“Adding a command log without an assay associated with it”</code></pre>
</div>
</div>
<p>Clustering function <code>FindClusters</code>. The resolution is used to change the number of clusters detected. We do not need many, so we set on to 0.5. Usual values range between 0.1 and 1.</p>
<div class="cell" data-tags="[]" data-execution_count="103">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>seurat.integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(<span class="at">object =</span> seurat.integrated, </span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">resolution =</span> .<span class="dv">5</span>,</span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">random.seed =</span> <span class="dv">123</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 20572
Number of edges: 1140129

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9378
Number of communities: 20
Elapsed time: 2 seconds</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“UNRELIABLE VALUE: One of the ‘future.apply’ iterations (‘future_lapply-1’) unexpectedly generated random numbers without declaring so. There is a risk that those random numbers are not statistically sound and the overall results might be invalid. To fix this, specify 'future.seed=TRUE'. This ensures that proper, parallel-safe random numbers are produced via the L'Ecuyer-CMRG method. To disable this check, use 'future.seed = NULL', or set option 'future.rng.onMisuse' to "ignore".”</code></pre>
</div>
</div>
<p>The clusters are saved in the meta data table as <code>integrated_snn_res.0.25</code>. <strong>Note that the name changes with the resolution</strong>. Also observe how much metadata we have: many columns come from tools we have applied, such as doubletfinder (DF) and nearest neighbor distances (snn).</p>
<div class="cell" data-tags="[]" data-execution_count="104">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>( seurat.integrated<span class="sc">@</span>meta.data )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 16</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">nCount_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_RNA</th>
<th data-quarto-table-cell-role="th" scope="col">nCount_SCT</th>
<th data-quarto-table-cell-role="th" scope="col">nFeature_SCT</th>
<th data-quarto-table-cell-role="th" scope="col">orig.ident</th>
<th data-quarto-table-cell-role="th" scope="col">Condition</th>
<th data-quarto-table-cell-role="th" scope="col">percent.mt</th>
<th data-quarto-table-cell-role="th" scope="col">percent.chloroplast</th>
<th data-quarto-table-cell-role="th" scope="col">pANN_0.25_0.09_609</th>
<th data-quarto-table-cell-role="th" scope="col">DF.classifications_0.25_0.09_609</th>
<th data-quarto-table-cell-role="th" scope="col">pANN_0.25_0.09_329</th>
<th data-quarto-table-cell-role="th" scope="col">DF.classifications_0.25_0.09_329</th>
<th data-quarto-table-cell-role="th" scope="col">pANN_0.25_0.09_309</th>
<th data-quarto-table-cell-role="th" scope="col">DF.classifications_0.25_0.09_309</th>
<th data-quarto-table-cell-role="th" scope="col">integrated_snn_res.0.5</th>
<th data-quarto-table-cell-role="th" scope="col">seurat_clusters</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">AAACCCACATGATCTG-1_1</td>
<td>20942</td>
<td>4711</td>
<td>11291</td>
<td>4192</td>
<td>Control1_seurat</td>
<td>Control</td>
<td>0.4775093</td>
<td>0.12415242</td>
<td>0.2299688</td>
<td>Singlet</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>8</td>
<td>8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">AAACCCAGTAGCTTGT-1_1</td>
<td>29105</td>
<td>5157</td>
<td>10486</td>
<td>3382</td>
<td>Control1_seurat</td>
<td>Control</td>
<td>0.2954819</td>
<td>0.05497337</td>
<td>0.2695109</td>
<td>Singlet</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>10</td>
<td>10</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">AAACCCAGTCTCTCAC-1_1</td>
<td>6115</td>
<td>2124</td>
<td>10086</td>
<td>2163</td>
<td>Control1_seurat</td>
<td>Control</td>
<td>1.6026165</td>
<td>0.01635323</td>
<td>0.1904266</td>
<td>Singlet</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">AAACCCATCACCTTGC-1_1</td>
<td>7410</td>
<td>2988</td>
<td>10003</td>
<td>2985</td>
<td>Control1_seurat</td>
<td>Control</td>
<td>0.5263158</td>
<td>0.09446694</td>
<td>0.2466181</td>
<td>Singlet</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>18</td>
<td>18</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">AAACGAAAGTCCTGTA-1_1</td>
<td>5616</td>
<td>2034</td>
<td>9866</td>
<td>2097</td>
<td>Control1_seurat</td>
<td>Control</td>
<td>0.7122507</td>
<td>0.10683761</td>
<td>0.2382934</td>
<td>Singlet</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>17</td>
<td>17</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">AAACGAAAGTGTGTTC-1_1</td>
<td>12580</td>
<td>3329</td>
<td>11180</td>
<td>3328</td>
<td>Control1_seurat</td>
<td>Control</td>
<td>0.1510334</td>
<td>0.06359300</td>
<td>0.2559834</td>
<td>Singlet</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>NA</td>
<td>18</td>
<td>18</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We can plot the clusters in the UMAP plot</p>
<div class="cell" data-tags="[]" data-execution_count="105">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">10</span>, <span class="at">repr.plot.height=</span><span class="dv">8</span>)</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(<span class="at">object =</span> seurat.integrated, <span class="at">reduction =</span> <span class="st">"umap"</span>,  <span class="at">label =</span> T, <span class="at">repel =</span> <span class="cn">TRUE</span>, <span class="at">pt.size =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-umapleiden" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-umapleiden-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;21: UMAP of the integrated datasets with clusters (still unassigned to cell types).</figcaption>
</figure>
</div>
</div>
</div>
<section id="cluster-assignment-from-visualized-marker-scores" class="level3" data-number="3.1.1">
<h3 data-number="3.1.1" class="anchored" data-anchor-id="cluster-assignment-from-visualized-marker-scores"><span class="header-section-number">3.1.1</span> Cluster assignment from visualized marker scores</h3>
<p>Here, we look at how to assign names based on known markers. In this procedure, biological knowledge of the cell types is needed. Below, there is a list of known markers for each cell type, extracted from the supplementary data of <span class="citation" data-cites="frank_single-cell_2023">Frank et al. (<a href="#ref-frank_single-cell_2023" role="doc-biblioref">2023</a>)</span>.</p>
<div class="cell" data-execution_count="106">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>features_list <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Cortex_scoring'</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi1g1v0006200"</span>,</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"LotjaGi1g1v0022100"</span>,</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"LotjaGi1g1v0261700"</span>,</span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"LotjaGi1g1v0348000"</span>,</span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"LotjaGi2g1v0303000"</span>,</span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"LotjaGi3g1v0505900"</span>),</span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Epidermis_scoring'</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi1g1v0080000"</span>,</span>
<span id="cb104-9"><a href="#cb104-9" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"LotjaGi1g1v0377600"</span>,</span>
<span id="cb104-10"><a href="#cb104-10" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"LotjaGi1g1v0613100"</span>,</span>
<span id="cb104-11"><a href="#cb104-11" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"LotjaGi3g1v0070500"</span>),</span>
<span id="cb104-12"><a href="#cb104-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Endodermis_scoring'</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi1g1v0114400"</span>,</span>
<span id="cb104-13"><a href="#cb104-13" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"LotjaGi1g1v0221300"</span>,</span>
<span id="cb104-14"><a href="#cb104-14" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"LotjaGi1g1v0240900-LC"</span>,</span>
<span id="cb104-15"><a href="#cb104-15" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"LotjaGi1g1v0707500"</span>), </span>
<span id="cb104-16"><a href="#cb104-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">'RootCap_scoring'</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi1g1v0020900"</span>,</span>
<span id="cb104-17"><a href="#cb104-17" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi1g1v0039700-LC"</span>,</span>
<span id="cb104-18"><a href="#cb104-18" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi1g1v0040300"</span>,</span>
<span id="cb104-19"><a href="#cb104-19" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi1g1v0147500"</span>),  </span>
<span id="cb104-20"><a href="#cb104-20" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Meristem_scoring'</span><span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi4g1v0300900"</span>,</span>
<span id="cb104-21"><a href="#cb104-21" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"LotjaGi6g1v0056500"</span>,</span>
<span id="cb104-22"><a href="#cb104-22" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"LotjaGi1g1v0594200"</span>),</span>
<span id="cb104-23"><a href="#cb104-23" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Phloem_scoring'</span><span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi1g1v0028800"</span>,</span>
<span id="cb104-24"><a href="#cb104-24" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi1g1v0085900"</span>,</span>
<span id="cb104-25"><a href="#cb104-25" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi1g1v0119300"</span>,</span>
<span id="cb104-26"><a href="#cb104-26" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi1g1v0149100"</span>),</span>
<span id="cb104-27"><a href="#cb104-27" aria-hidden="true" tabindex="-1"></a>    <span class="st">'QuiescentCenter_scoring'</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi1g1v0004300"</span>,</span>
<span id="cb104-28"><a href="#cb104-28" aria-hidden="true" tabindex="-1"></a>                           <span class="st">"LotjaGi1g1v0021400"</span>,</span>
<span id="cb104-29"><a href="#cb104-29" aria-hidden="true" tabindex="-1"></a>                           <span class="st">"LotjaGi1g1v0052700"</span>,</span>
<span id="cb104-30"><a href="#cb104-30" aria-hidden="true" tabindex="-1"></a>                           <span class="st">"LotjaGi1g1v0084000"</span>),</span>
<span id="cb104-31"><a href="#cb104-31" aria-hidden="true" tabindex="-1"></a>    <span class="st">'RootHair_scoring'</span><span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi1g1v0014300"</span>,</span>
<span id="cb104-32"><a href="#cb104-32" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi1g1v0109000"</span>,</span>
<span id="cb104-33"><a href="#cb104-33" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi1g1v0109100"</span>,</span>
<span id="cb104-34"><a href="#cb104-34" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LLotjaGi1g1v0143900"</span>),  </span>
<span id="cb104-35"><a href="#cb104-35" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Pericycle_scoring'</span><span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi3g1v0222100"</span>,</span>
<span id="cb104-36"><a href="#cb104-36" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi3g1v0395900-LC"</span>,</span>
<span id="cb104-37"><a href="#cb104-37" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi5g1v0166000-LC"</span>,</span>
<span id="cb104-38"><a href="#cb104-38" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi3g1v0395500-LC"</span>,</span>
<span id="cb104-39"><a href="#cb104-39" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi1g1v0783700-LC"</span>,</span>
<span id="cb104-40"><a href="#cb104-40" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi2g1v0333200"</span>,</span>
<span id="cb104-41"><a href="#cb104-41" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"LotjaGi4g1v0293000-LC"</span>),     </span>
<span id="cb104-42"><a href="#cb104-42" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Stele_scoring'</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi2g1v0126700"</span>,</span>
<span id="cb104-43"><a href="#cb104-43" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi1g1v0558200"</span>,</span>
<span id="cb104-44"><a href="#cb104-44" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi4g1v0215500"</span>,</span>
<span id="cb104-45"><a href="#cb104-45" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi3g1v0174100"</span>,</span>
<span id="cb104-46"><a href="#cb104-46" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi5g1v0288600"</span>,</span>
<span id="cb104-47"><a href="#cb104-47" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi3g1v0129700"</span>),</span>
<span id="cb104-48"><a href="#cb104-48" aria-hidden="true" tabindex="-1"></a>    <span class="st">'Xylem_scoring'</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LotjaGi1g1v0623100"</span>,</span>
<span id="cb104-49"><a href="#cb104-49" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi1g1v0569300"</span>,</span>
<span id="cb104-50"><a href="#cb104-50" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi1g1v0443000"</span>,</span>
<span id="cb104-51"><a href="#cb104-51" aria-hidden="true" tabindex="-1"></a>                <span class="st">"LotjaGi1g1v0428800"</span>)</span>
<span id="cb104-52"><a href="#cb104-52" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, we need a function calculating the scores for each cell type. This is the average expression of the markers in the list, from which we remove the average expression of some control genes, which are supposed not to be specific for the cell type of interest. The cells matching the desired type should retain a high score.</p>
<div class="cell" data-execution_count="107">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">AddModuleScore</span>(</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> seurat.integrated,</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> features_list,</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">ctrl =</span> <span class="dv">5</span>,</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">name =</span> <span class="st">'LJ_scores'</span></span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“The following features are not present in the object: LotjaGi1g1v0085900, not searching for symbol synonyms”
Warning message:
“The following features are not present in the object: LLotjaGi1g1v0143900, not searching for symbol synonyms”
Warning message:
“The following features are not present in the object: LotjaGi3g1v0129700, not searching for symbol synonyms”</code></pre>
</div>
</div>
<p>We also apply a function (from <code>script.R</code>) to rename the scores in the metadata. Their names are not intuitive by default, they are all called with the name chosen above and a number after it:</p>
<div class="cell" data-execution_count="108">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(seurat.clustered<span class="sc">@</span>meta.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>'nCount_RNA'</li><li>'nFeature_RNA'</li><li>'nCount_SCT'</li><li>'nFeature_SCT'</li><li>'orig.ident'</li><li>'Condition'</li><li>'percent.mt'</li><li>'percent.chloroplast'</li><li>'pANN_0.25_0.09_609'</li><li>'DF.classifications_0.25_0.09_609'</li><li>'pANN_0.25_0.09_329'</li><li>'DF.classifications_0.25_0.09_329'</li><li>'pANN_0.25_0.09_309'</li><li>'DF.classifications_0.25_0.09_309'</li><li>'integrated_snn_res.0.5'</li><li>'seurat_clusters'</li><li>'LJ_scores1'</li><li>'LJ_scores2'</li><li>'LJ_scores3'</li><li>'LJ_scores4'</li><li>'LJ_scores5'</li><li>'LJ_scores6'</li><li>'LJ_scores7'</li><li>'LJ_scores8'</li><li>'LJ_scores9'</li><li>'LJ_scores10'</li><li>'LJ_scores11'</li></ol>
</div>
</div>
<div class="cell" data-execution_count="109">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">renameScores</span>(<span class="at">markers_list =</span> features_list, <span class="at">seurat_data =</span> seurat.clustered)      </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scores renamed FROM




TO



</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>LJ_scores1
LJ_scores2
LJ_scores3
LJ_scores4
LJ_scores5
LJ_scores6
LJ_scores7
LJ_scores8
LJ_scores9
LJ_scores10
LJ_scores11
Cortex_scoring
Epidermis_scoring
Endodermis_scoring
RootCap_scoring
Meristem_scoring
Phloem_scoring
QuiescentCenter_scoring
RootHair_scoring
Pericycle_scoring
Stele_scoring
Xylem_scoring</code></pre>
</div>
</div>
<p>Now we run the function <code>plotScoresUMAP</code> (from the file <code>script.R</code>). In <a href="#fig-scores">Figure&nbsp;22</a> we can see that some clusters are easy to classify (phloem and xylem), but many others are not. This is mainly due to the fact that the change of many cell types is a continuum, and this manual annotation is very subjective.</p>
<div class="cell" data-execution_count="110">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotScoresUMAP</span>(<span class="at">markers_list =</span> features_list, <span class="at">seurat_data =</span> seurat.clustered)                                  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-scores" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-scores-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;22: Scores of the list of markers for the Lotus Japonicus clusters. Those should help identifying cell types by plotting the average of markers subtracted the average of other random genes in the data.</figcaption>
</figure>
</div>
</div>
</div>
<p>One easy solution is to use the highest scoring of each cluster to assign the cluster name. Below, in the function <code>clusterNames</code>, for each cluster, we sum the scores of each cell type, and the highest value decides the cluster name.</p>
<div class="cell" data-execution_count="111">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="co">#use the clustering above</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat.clustered) <span class="ot">&lt;-</span> <span class="st">'integrated_snn_res.0.5'</span> </span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a><span class="co">#assign names</span></span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>seurat.clustered<span class="sc">@</span>meta.data[<span class="st">"Cell_types"</span>] <span class="ot">&lt;-</span> <span class="fu">clusterNames</span>(seurat.clustered, features_list)</span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a><span class="co">#use the new names as clustering labels</span></span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat.clustered) <span class="ot">&lt;-</span> <span class="st">'Cell_types'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Cluster assignment started

--- QuiescentCenter assigned to 8

--- RootHair assigned to 10

--- RootCap assigned to 0

--- Meristem assigned to 18

--- Phloem assigned to 17

--- Pericycle assigned to 4

--- Stele assigned to 13

--- Endodermis assigned to 2

--- Epidermis assigned to 12

--- Cortex assigned to 5

--- Epidermis assigned to 7

--- Epidermis assigned to 6

--- Pericycle assigned to 3

--- Xylem assigned to 19

--- Cortex assigned to 1

--- RootCap assigned to 16

--- RootHair assigned to 11

--- Cortex assigned to 9

--- Epidermis assigned to 14

--- Endodermis assigned to 15

Cluster assignment finished
</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Manual assignment">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Manual assignment
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you wanted to manually assign cell types, then you could use the command <code>RenameIdents</code>, for example</p>
<pre><code>Idents(seurat.clustered) &lt;- 'integrated_snn_res.0.5'

seurat.clustered &lt;- RenameIdents(object = seurat.clustered,
                               "2"="Cortex", "5"="Cortex", "11"="Cortex",
                               "6"="Epidermis", "12"="Epidermis", "7"="Epidermis",
                               "15"="Endodermis",  
                               "0"="Root_Cap", "13"="Root_Cap",
                               "16"="Meristem",  
                               "17"="Phloem",
                               "10"="Root_Hair", "9"="Root_Hair",
                               "1"="PericycleStele", 
                               "3"="Pericycle", "19"="Pericycle",
                               "20"="Xylem")</code></pre>
<p>(the names do not match correctly the numbers of our clustering, they are just for the sake of the example)</p>
</div>
</div>
<p>It seems from <a href="#fig-umapbyeye">Figure&nbsp;23</a> that we have quite consistent clustering across the various cell types.</p>
<div class="cell" data-tags="[]" data-execution_count="112">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">10</span>, <span class="at">repr.plot.height=</span><span class="dv">10</span>)</span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(<span class="at">object =</span> seurat.clustered, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>, <span class="at">label=</span>T, <span class="at">pt.size =</span> <span class="dv">2</span>, <span class="at">label.size =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-umapbyeye" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-umapbyeye-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;23: UMAP of the integrated datasets with clusters assigned by looking at markers scores and using a function to assign score-based names. Note that there are a few cells here and there that are not labelled correctly, which often happens as a few datapoints are noisy.</figcaption>
</figure>
</div>
</div>
</div>
<section id="optional-assigning-a-mixed-cluster" class="level4" data-number="3.1.1.1">
<h4 data-number="3.1.1.1" class="anchored" data-anchor-id="optional-assigning-a-mixed-cluster"><span class="header-section-number">3.1.1.1</span> Optional: assigning a mixed cluster</h4>
<p>Consider the marker plots for pericycle and Stele cell types in <a href="#fig-scores">Figure&nbsp;22</a>. Here you can see overlap of the markers, which is not unnormal, since biological processes often transition gradually and eventually share some markers. We can try to separate the two cell types more precisely by assigning the cell type to each single data point, by comparing its score for Pericycle and Stele, instead of renaming each cluster as a whole.</p>
<p>The code below runs such comparison for each cell in the pericycle and stele clusters.</p>
<div class="cell" data-scrolled="true" data-execution_count="118">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>peri <span class="ot">&lt;-</span> seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Pericycle_scoring[ seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Cell_types <span class="sc">==</span> <span class="st">'Pericycle'</span> <span class="sc">|</span> seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Cell_types <span class="sc">==</span> <span class="st">'Stele'</span>]</span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>stel <span class="ot">&lt;-</span> seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Stele_scoring[ seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Cell_types <span class="sc">==</span> <span class="st">'Pericycle'</span> <span class="sc">|</span> seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Cell_types <span class="sc">==</span> <span class="st">'Stele'</span> ]</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>peri_stele <span class="ot">&lt;-</span> peri<span class="sc">&gt;=</span>stel</span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>finalcl <span class="ot">=</span> <span class="fu">c</span>()</span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> peri_stele)</span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>    finalcl <span class="ot">=</span> <span class="fu">c</span>(finalcl, <span class="fu">ifelse</span>(i, <span class="st">"Pericycle"</span>, <span class="st">"Stele"</span>))</span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a>celltypes <span class="ot">&lt;-</span> seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Cell_types</span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a>celltypes[ seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Cell_types <span class="sc">==</span> <span class="st">'Pericycle'</span> <span class="sc">|</span> seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Cell_types <span class="sc">==</span> <span class="st">'Stele'</span> ] <span class="ot">&lt;-</span> finalcl</span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a>seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Cell_types <span class="ot">&lt;-</span> celltypes</span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat.clustered) <span class="ot">&lt;-</span> <span class="st">'Cell_types'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="119">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">10</span>, <span class="at">repr.plot.height=</span><span class="dv">10</span>)</span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(<span class="at">object =</span> seurat.clustered, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>, <span class="at">label=</span>T, <span class="at">pt.size =</span> <span class="dv">2</span>, <span class="at">label.size =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-umapsubcluster" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-umapsubcluster-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;24: UMAP of the integrated datasets with clusters assigned by refining Pericycle and Stele.</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="cluster-assignment-from-an-annotated-dataset" class="level3" data-number="3.1.2">
<h3 data-number="3.1.2" class="anchored" data-anchor-id="cluster-assignment-from-an-annotated-dataset"><span class="header-section-number">3.1.2</span> Cluster assignment from an annotated dataset</h3>
<p>We now use the annotated data from <span class="citation" data-cites="frank_single-cell_2023">Frank et al. (<a href="#ref-frank_single-cell_2023" role="doc-biblioref">2023</a>)</span> (which is the same we are using in the tutorial) to transfer data labels to our own processed data. More about label transfer can be read at <span class="citation" data-cites="stuart_comprehensive_2019">Stuart et al. (<a href="#ref-stuart_comprehensive_2019" role="doc-biblioref">2019</a>)</span>. We load the data from the paper and define reference and query data.</p>
<div class="cell" data-execution_count="120">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>seurat.reference <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../Data/data_lavinia.RDS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="121">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>seurat.query <span class="ot">&lt;-</span> seurat.clustered</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We have to define the data integration between query and reference before we can transfer the cluster names. For the algorithm to work, we need to use the “RNA” assay, which contains raw expression values.</p>
<div class="cell" data-execution_count="122">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat.query) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="123">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>lotusjaponicus.anchors <span class="ot">&lt;-</span> <span class="fu">FindTransferAnchors</span>(<span class="at">reference =</span> seurat.reference, </span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">features =</span> <span class="fu">intersect</span>( <span class="fu">rownames</span>(seurat.query), </span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>                                                   <span class="fu">rownames</span>( seurat.reference[[<span class="st">'SCT'</span>]]<span class="sc">@</span>scale.data ) ),</span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">query =</span> seurat.query, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, </span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">reference.reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">reference.assay=</span><span class="st">'RNA'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Projecting cell embeddings

Finding neighborhoods

Finding anchors

    Found 26902 anchors

Filtering anchors

    Retained 13016 anchors
</code></pre>
</div>
</div>
<p>Calculating the integration of the labels from the reference takes time. So we save the calculated anchors for the integration. If you need to rerun the code, skip the command above and instead load the data with <code>readRDS</code> below.</p>
<div class="cell" data-execution_count="124">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(lotusjaponicus.anchors, <span class="at">file =</span> <span class="st">"anchors.RDS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="125">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>lotusjaponicus.anchors <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"anchors.RDS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now it is finally time to transfer the labels and add them to the metadata. The column in the metadata is called by default <code>predicted.id</code>.</p>
<div class="cell" data-execution_count="126">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(<span class="at">anchorset =</span> lotusjaponicus.anchors, </span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">refdata =</span> <span class="fu">Idents</span>(seurat.reference), </span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vectors

Finding integration vector weights

Predicting cell labels
</code></pre>
</div>
</div>
<div class="cell" data-execution_count="127">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(seurat.clustered, <span class="at">metadata =</span> predictions[<span class="st">'predicted.id'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Just as a reminder of what is in the metadata, we can quickly look at the column names. Those are ordered by when we added things along the analysis. If you read the names, you can recognize part of the analysis steps until now.</p>
<div class="cell" data-execution_count="128">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>( seurat.clustered<span class="sc">@</span>meta.data )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>'nCount_RNA'</li><li>'nFeature_RNA'</li><li>'nCount_SCT'</li><li>'nFeature_SCT'</li><li>'orig.ident'</li><li>'Condition'</li><li>'percent.mt'</li><li>'percent.chloroplast'</li><li>'pANN_0.25_0.09_609'</li><li>'DF.classifications_0.25_0.09_609'</li><li>'pANN_0.25_0.09_329'</li><li>'DF.classifications_0.25_0.09_329'</li><li>'pANN_0.25_0.09_309'</li><li>'DF.classifications_0.25_0.09_309'</li><li>'integrated_snn_res.0.5'</li><li>'seurat_clusters'</li><li>'Cortex_scoring'</li><li>'Epidermis_scoring'</li><li>'Endodermis_scoring'</li><li>'RootCap_scoring'</li><li>'Meristem_scoring'</li><li>'Phloem_scoring'</li><li>'QuiescentCenter_scoring'</li><li>'RootHair_scoring'</li><li>'Pericycle_scoring'</li><li>'Stele_scoring'</li><li>'Xylem_scoring'</li><li>'Cell_types'</li><li>'predicted.id'</li></ol>
</div>
</div>
<p>Here we define as clustering for the data and the plots, the one transfered just before. We then have a look at <a href="#fig-transfer">Figure&nbsp;25</a> to observe that the labels look fine.</p>
<div class="cell" data-execution_count="129">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat.clustered) <span class="ot">&lt;-</span> <span class="st">'predicted.id'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="130">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">10</span>, <span class="at">repr.plot.height=</span><span class="dv">10</span>)</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(<span class="at">object =</span> seurat.clustered, </span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb130-4"><a href="#cb130-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">repel =</span> <span class="cn">TRUE</span>, <span class="at">label=</span>T, </span>
<span id="cb130-5"><a href="#cb130-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">pt.size =</span> <span class="fl">0.5</span>, <span class="at">label.size =</span> <span class="dv">10</span>, )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-transfer" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-transfer-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;25: UMAP of the integrated datasets with cell types named through label transfer.</figcaption>
</figure>
</div>
</div>
</div>
<p>We save the data</p>
<div class="cell" data-execution_count="131">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SaveH5Seurat</span>(<span class="at">object =</span> seurat.clustered, </span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">filename =</span> <span class="st">"seurat.clustered.h5Seurat"</span>, </span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">overwrite =</span> <span class="cn">TRUE</span>,</span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Overwriting previous file seurat.clustered.h5Seurat”
Creating h5Seurat file for version 3.1.5.9900
</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="sec-geneanalysis" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Gene Expression Analysis</h1>
<p>In this section we explore the gene expression through</p>
<ul>
<li>determining <strong>differentially expressed genes</strong> for the infected condition against the control. Differentially expressed genes are significantly more epressed in one of the two groups used for the comparison. Usually the wild type is used as query for the comparison, such that differentially expressed genes are referred to the perturbated condition (infection, knock-out, illness, …)</li>
<li>studying <strong>coexpression modules</strong> (a module is a group of gene similarly expressed across cells in the data) to find if
<ul>
<li>any of them contains the gene of interest,</li>
<li>they are significantly more expressed in specific cell groups (<strong>Differential module expression</strong>)</li>
<li>if there are specific known functions associated to some modules</li>
</ul></li>
</ul>
<p>We will also use gene ontology terms for understanding the function of groups of genes.</p>
<section id="differential-gene-expression-dge" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="differential-gene-expression-dge"><span class="header-section-number">4.1</span> Differential Gene Expression (DGE)</h2>
<p>Here we test each cluster to see which are significantly more expressed genes in the infected samples compared to the wild-type samples. We also see if we find the gene RINRK1 as being significant. Again, the resulting genes can be useful to be integrated with the GO terms as we did before.</p>
<p>We first have a quick look to see how much the RINRK1 gene is expressed in the data. We use the <code>RNA</code> assay to plot the true expression values. The UMAP plot shows few cells expressing the genes, meaning its average expression is going to be very low, so it is likely we will not find the gene to be differentially expressed anywhere.</p>
<div class="cell" data-tags="[]" data-execution_count="132">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a>RINRK1.id <span class="ot">&lt;-</span> <span class="st">'LotjaGi1g1v0182900'</span></span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat.clustered) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat.clustered,</span>
<span id="cb133-6"><a href="#cb133-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb133-7"><a href="#cb133-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(RINRK1.id), </span>
<span id="cb133-8"><a href="#cb133-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb133-9"><a href="#cb133-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="dv">0</span>, </span>
<span id="cb133-10"><a href="#cb133-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">pt.size =</span> <span class="dv">1</span>,</span>
<span id="cb133-11"><a href="#cb133-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb133-12"><a href="#cb133-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">label.size =</span> <span class="dv">7</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-rinrk" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-rinrk-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;26: UMAP plot for the expression of gene RINRK1 in the data</figcaption>
</figure>
</div>
</div>
</div>
<p>From biological knowledge, we expect the gene mostly expressed in the cortex and trichoblasts upon inoculation with rhizobia, and that is what happens in our data as well. We can see it in the code and violin plot of <a href="#fig-vln">Figure&nbsp;27</a></p>
<div class="cell" data-tags="[]" data-execution_count="133">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Cells in inoculated L.J. expressing"</span>, RINRK1.id, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( <span class="fu">sum</span>( <span class="fu">as.numeric</span>(<span class="fu">GetAssayData</span>(seurat.clustered[RINRK1.id,]))<span class="sc">&gt;</span><span class="dv">0</span> <span class="sc">&amp;</span> </span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a>     seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Condition<span class="sc">==</span><span class="st">"R7A"</span> ) )</span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-6"><a href="#cb134-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Cells in control L.J. expressing"</span>, RINRK1.id, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb134-7"><a href="#cb134-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-8"><a href="#cb134-8" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>( <span class="fu">sum</span>( <span class="fu">as.numeric</span>(<span class="fu">GetAssayData</span>(seurat.clustered[RINRK1.id,]))<span class="sc">&gt;</span><span class="dv">0</span> <span class="sc">&amp;</span> </span>
<span id="cb134-9"><a href="#cb134-9" aria-hidden="true" tabindex="-1"></a>    seurat.clustered<span class="sc">@</span>meta.data<span class="sc">$</span>Condition<span class="sc">==</span><span class="st">"Control"</span> ) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Cells in inoculated L.J. expressing LotjaGi1g1v0182900 
56
Cells in control L.J. expressing LotjaGi1g1v0182900 
0</code></pre>
</div>
</div>
<div class="cell" data-execution_count="134">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(seurat.clustered, </span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> RINRK1.id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-vln" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-vln-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;27: Violin plot for the expression of gene RINRK1 in various clusters</figcaption>
</figure>
</div>
</div>
</div>
<p>The code below uses <code>FindMarkers</code> to compare <code>R7A</code> condition against <code>Control</code> for each cluster, with a filter to remove non-singnificant genes (keeping p-value below 0.001 and log fold change &gt; 1 and &lt; -1). We keep also genes expressed 30% more than in the condition where they are underexpressed. We use only the 2000 most variable genes, since we are interested in very variable gene expressions across data.</p>
<div class="cell" data-execution_count="135">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(seurat.clustered, <span class="at">nfeatures =</span> <span class="dv">2000</span>, <span class="at">assay =</span> <span class="st">"integrated"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Not all features provided are in this Assay object, removing the following feature(s): LotjaGi2g1v0240800, LotjaGi2g1v0095600, LotjaGi1g1v0324300, LotjaGi2g1v0394200, LotjaGi5g1v0341600, LotjaGi2g1v0130600-LC, LotjaGi6g1v0167000-LC, LotjaGi1g1v0358100, LotjaGi3g1v0487600, LotjaGi1g1v0766200, LotjaGi1g1v0781500, LotjaGi1g1v0578800, LotjaGi3g1v0274700, LotjaGi2g1v0228100, LotjaGi4g1v0338800, LotjaGi2g1v0341400, LotjaGi1g1v0315100, LotjaGi1g1v0544500, LotjaGi5g1v0019000, LotjaGi5g1v0202500, LotjaGi4g1v0107800, LotjaGi3g1v0150400, LotjaGi1g1v0098500, LotjaGi2g1v0000800, LotjaGi2g1v0177400, LotjaGi5g1v0178500, LotjaGi6g1v0097400, LotjaGi2g1v0005400, LotjaGi1g1v0667100, LotjaGi4g1v0006600, LotjaGi3g1v0347800, LotjaGi2g1v0430200, LotjaGi2g1v0143500, LotjaGi4g1v0074400, LotjaGi4g1v0346500, LotjaGi4g1v0068000, LotjaGi5g1v0226200, LotjaGi3g1v0464000, LotjaGi6g1v0005900, LotjaGi3g1v0468000, LotjaGi3g1v0280300, LotjaGi5g1v0323100, LotjaGi4g1v0193600-LC, LotjaGi6g1v0155700, LotjaGi6g1v0253000, LotjaGi6g1v0184900, LotjaGi4g1v0299200, LotjaGi4g1v0463700, LotjaGi4g1v0235000, LotjaGi4g1v0227700, LotjaGi4g1v0114900-LC, LotjaGi5g1v0339200, LotjaGi6g1v0081800, LotjaGi1g1v0054800, LotjaGi3g1v0223400, LotjaGi4g1v0010100, LotjaGi5g1v0085800, LotjaGi4g1v0402200, LotjaGi3g1v0178100, LotjaGi4g1v0400500, LotjaGi1g1v0453100, LotjaGi2g1v0326500, LotjaGi6g1v0010000, LotjaGi3g1v0413600, LotjaGi4g1v0399100, LotjaGi2g1v0176600, LotjaGi4g1v0348500, LotjaGi5g1v0298500, LotjaGi2g1v0081400, LotjaGi6g1v0271600, LotjaGi3g1v0393500, LotjaGi1g1v0261100, LotjaGi1g1v0152000, LotjaGi5g1v0314700, LotjaGi6g1v0012300-LC, LotjaGi2g1v0137900, LotjaGi2g1v0106500, LotjaGi1g1v0361600, LotjaGi4g1v0266600, LotjaGi4g1v0185900, LotjaGi3g1v0426100, LotjaGi3g1v0533100, LotjaGi3g1v0002600, LotjaGi6g1v0128400-LC, LotjaGi3g1v0241200, LotjaGi1g1v0783200-LC, LotjaGi3g1v0460400, LotjaGi2g1v0357000, LotjaGi1g1v0005100, LotjaGi4g1v0278500-LC, LotjaGi2g1v0369100, LotjaGi1g1v0191800, LotjaGi3g1v0191800, LotjaGi2g1v0142600-LC, LotjaGi2g1v0199400, LotjaGi3g1v0222800, LotjaGi2g1v0129700, LotjaGi1g1v0799200, LotjaGi4g1v0191000, LotjaGi3g1v0083400, LotjaGi6g1v0098400, LotjaGi6g1v0233800, LotjaGi3g1v0324800, LotjaGi2g1v0310400, LotjaGi1g1v0471800, LotjaGi4g1v0005400, LotjaGi5g1v0111200-LC, LotjaGi1g1v0390100, LotjaGi1g1v0661600, LotjaGi3g1v0526400-LC, LotjaGi3g1v0107400, LotjaGi4g1v0261600, LotjaGi2g1v0278500, LotjaGi3g1v0494100, LotjaGi1g1v0515700, LotjaGi5g1v0047700, LotjaGi5g1v0167000, LotjaGi6g1v0217400, LotjaGi5g1v0285900, LotjaGi4g1v0039800, LotjaGi6g1v0327300, LotjaGi1g1v0688900, LotjaGi2g1v0431800, LotjaGi5g1v0034900, LotjaGi5g1v0355500, LotjaGi1g1v0691200, LotjaGi4g1v0236000, LotjaGi5g1v0344500, LotjaGi3g1v0457600-LC, LotjaGi4g1v0313800, LotjaGi4g1v0152900, LotjaGi5g1v0082200, LotjaGi4g1v0437000, LotjaGi3g1v0481200, LotjaGi1g1v0079200, LotjaGi2g1v0275000, LotjaGi2g1v0103900, LotjaGi6g1v0263000, LotjaGi6g1v0209300, LotjaGi2g1v0323700, LotjaGi1g1v0546000, LotjaGi4g1v0358100, LotjaGi5g1v0118700, LotjaGi4g1v0278300-LC, LotjaGi2g1v0018700, LotjaGi5g1v0242200, LotjaGi6g1v0049600, LotjaGi4g1v0199100, LotjaGi6g1v0197000, LotjaGi5g1v0174700, LotjaGi6g1v0016800, LotjaGi2g1v0291500, LotjaGi1g1v0692100, LotjaGi5g1v0297900, LotjaGi3g1v0414000, LotjaGi2g1v0401800, LotjaGi2g1v0424100, LotjaGi4g1v0246100, LotjaGi6g1v0045500, LotjaGi3g1v0177500, LotjaGi4g1v0406700, LotjaGi6g1v0181400, LotjaGi2g1v0308500, LotjaGi1g1v0636100, LotjaGi3g1v0147700-LC, LotjaGi3g1v0300200, LotjaGi6g1v0138400, LotjaGi1g1v0545900, LotjaGi1g1v0567100, LotjaGi1g1v0758000-LC, LotjaGi6g1v0294700, LotjaGi4g1v0430000, LotjaGi3g1v0081700, LotjaGi6g1v0050600, LotjaGi6g1v0228700, LotjaGi6g1v0165600, LotjaGi5g1v0345000, LotjaGi4g1v0141700, LotjaGi3g1v0074600, LotjaGi3g1v0055500, LotjaGi5g1v0301600-LC, LotjaGi5g1v0168500, LotjaGi3g1v0451700, LotjaGi1g1v0094700, LotjaGi1g1v0759800, LotjaGi6g1v0166800, LotjaGi1g1v0504700, LotjaGi4g1v0045000, LotjaGi1g1v0776200, LotjaGi5g1v0247100, LotjaGi6g1v0086100, LotjaGi5g1v0263900, LotjaGi3g1v0152600, LotjaGi3g1v0381800, LotjaGi3g1v0069500, LotjaGi1g1v0525500, LotjaGi1g1v0626900, LotjaGi3g1v0208100, LotjaGi1g1v0307300, LotjaGi5g1v0224700, LotjaGi6g1v0107500, LotjaGi1g1v0679300, LotjaGi2g1v0121500, LotjaGi4g1v0091000, LotjaGi2g1v0098800, LotjaGi3g1v0433100, LotjaGi4g1v0096200, LotjaGi2g1v0093200, LotjaGi4g1v0194400, LotjaGi6g1v0065000-LC, LotjaGi1g1v0081100-LC, LotjaGi3g1v0034500, LotjaGi3g1v0078750, LotjaGi3g1v0041900, LotjaGi4g1v0390600, LotjaGi1g1v0768500-LC, LotjaGi4g1v0250800, LotjaGi5g1v0036100, LotjaGi3g1v0451900, LotjaGi1g1v0216700, LotjaGi1g1v0261000, LotjaGi5g1v0028600, LotjaGi2g1v0126400, LotjaGi6g1v0310100, LotjaGi2g1v0400400, LotjaGi2g1v0193900, LotjaGi4g1v0461300, LotjaGi3g1v0387900, LotjaGi4g1v0332500, LotjaGi2g1v0019100, LotjaGi4g1v0036300, LotjaGi1g1v0736100, LotjaGi2g1v0245100, LotjaGi6g1v0266700, LotjaGi6g1v0008100, LotjaGi5g1v0222800, LotjaGi4g1v0231900, LotjaGi1g1v0744000, LotjaGi4g1v0019600, LotjaGi6g1v0304600, LotjaGi1g1v0071200, LotjaGi1g1v0473700, LotjaGi6g1v0086500, LotjaGi4g1v0330300, LotjaGi2g1v0309500, LotjaGi1g1v0636000, LotjaGi1g1v0271500, LotjaGi4g1v0299700-LC, LotjaGi1g1v0566000, LotjaGi4g1v0061600, LotjaGi4g1v0424500, LotjaGi5g1v0352800-LC, LotjaGi6g1v0070200, LotjaGi1g1v0673700, LotjaGi4g1v0253000, LotjaGi2g1v0168900, LotjaGi2g1v0362400, LotjaGi1g1v0684400-LC, LotjaGi5g1v0324400, LotjaGi3g1v0035700, LotjaGi3g1v0361600, LotjaGi3g1v0231900, LotjaGi2g1v0176900, LotjaGi2g1v0277100, LotjaGi1g1v0709700, LotjaGi1g1v0682100, LotjaGi1g1v0720300, LotjaGi6g1v0001700, LotjaGi2g1v0272700, LotjaGi1g1v0624800, LotjaGi5g1v0010700, LotjaGi6g1v0177900, LotjaGi6g1v0353500, LotjaGi3g1v0442400, LotjaGi3g1v0511600, LotjaGi4g1v0076100, LotjaGi6g1v0095700, LotjaGi2g1v0445600, LotjaGi3g1v0005000, LotjaGi2g1v0251400, LotjaGi1g1v0274100, LotjaGi6g1v0183300, LotjaGi3g1v0446100, LotjaGi3g1v0448900, LotjaGi3g1v0249200, LotjaGi3g1v0183200, LotjaGi4g1v0338600, LotjaGi3g1v0225600, LotjaGi2g1v0074700, LotjaGi3g1v0038500, LotjaGi2g1v0285700, LotjaGi1g1v0609600, LotjaGi2g1v0081200, LotjaGi2g1v0377700, LotjaGi3g1v0432500, LotjaGi2g1v0400000, LotjaGi5g1v0017600, LotjaGi2g1v0105800, LotjaGi6g1v0322500, LotjaGi4g1v0008300, LotjaGi1g1v0091300, LotjaGi1g1v0500100, LotjaGi4g1v0165100, LotjaGi6g1v0035000, LotjaGi3g1v0404000-LC, LotjaGi1g1v0240600-LC, LotjaGi3g1v0551400, LotjaGi4g1v0130900, LotjaGi6g1v0119900, LotjaGi4g1v0103300-LC, LotjaGi1g1v0562800, LotjaGi6g1v0236600, LotjaGi5g1v0236500, LotjaGi6g1v0092000, LotjaGi4g1v0395300, LotjaGi6g1v0334000”</code></pre>
</div>
</div>
<div class="cell" data-execution_count="136">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>DEG_table <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat.clustered, </span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">assay=</span><span class="st">'integrated'</span>,</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">ident.1 =</span> <span class="st">"R7A"</span>,</span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">ident.2 =</span> <span class="st">"Control"</span>, </span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>                             <span class="at">group.by =</span> <span class="st">"Condition"</span>,</span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>                             <span class="at">subset.ident=</span><span class="st">"Cortex"</span>,</span>
<span id="cb139-7"><a href="#cb139-7" aria-hidden="true" tabindex="-1"></a>                             <span class="at">min.diff.pct =</span> <span class="fl">0.3</span>,</span>
<span id="cb139-8"><a href="#cb139-8" aria-hidden="true" tabindex="-1"></a>                             <span class="at">verbose =</span> <span class="cn">TRUE</span>, </span>
<span id="cb139-9"><a href="#cb139-9" aria-hidden="true" tabindex="-1"></a>                             <span class="at">features =</span> seurat.clustered<span class="sc">@</span>assays<span class="sc">$</span>integrated<span class="sc">@</span>var.features,</span>
<span id="cb139-10"><a href="#cb139-10" aria-hidden="true" tabindex="-1"></a>                             <span class="at">test.use =</span> <span class="st">"wilcox"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb139-11"><a href="#cb139-11" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">filter</span>(p_val_adj <span class="sc">&lt;=</span> <span class="fl">0.001</span> <span class="sc">&amp;</span> <span class="fu">abs</span>(avg_log2FC)<span class="sc">&gt;</span><span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb139-12"><a href="#cb139-12" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">select</span>(<span class="sc">-</span>p_val)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="137">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat.clustered) <span class="ot">&lt;-</span> <span class="st">"integrated"</span> <span class="co">#return to the integrated data</span></span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a>DEG <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a>cluster.names <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">Idents</span>(seurat.clustered))</span>
<span id="cb140-4"><a href="#cb140-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-5"><a href="#cb140-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(CLUSTER <span class="cf">in</span> cluster.names){</span>
<span id="cb140-6"><a href="#cb140-6" aria-hidden="true" tabindex="-1"></a>    DEG_table <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat.clustered,</span>
<span id="cb140-7"><a href="#cb140-7" aria-hidden="true" tabindex="-1"></a>                             <span class="at">assay=</span><span class="st">'integrated'</span>,</span>
<span id="cb140-8"><a href="#cb140-8" aria-hidden="true" tabindex="-1"></a>                             <span class="at">ident.1 =</span> <span class="st">"R7A"</span>,</span>
<span id="cb140-9"><a href="#cb140-9" aria-hidden="true" tabindex="-1"></a>                             <span class="at">ident.2 =</span> <span class="st">"Control"</span>, </span>
<span id="cb140-10"><a href="#cb140-10" aria-hidden="true" tabindex="-1"></a>                             <span class="at">group.by =</span> <span class="st">"Condition"</span>,</span>
<span id="cb140-11"><a href="#cb140-11" aria-hidden="true" tabindex="-1"></a>                             <span class="at">subset.ident=</span>CLUSTER,</span>
<span id="cb140-12"><a href="#cb140-12" aria-hidden="true" tabindex="-1"></a>                             <span class="at">verbose =</span> <span class="cn">TRUE</span>, </span>
<span id="cb140-13"><a href="#cb140-13" aria-hidden="true" tabindex="-1"></a>                             <span class="at">min.diff.pct =</span> <span class="fl">0.3</span>,</span>
<span id="cb140-14"><a href="#cb140-14" aria-hidden="true" tabindex="-1"></a>                             <span class="at">features =</span> seurat.clustered<span class="sc">@</span>assays<span class="sc">$</span>integrated<span class="sc">@</span>var.features,</span>
<span id="cb140-15"><a href="#cb140-15" aria-hidden="true" tabindex="-1"></a>                             <span class="at">test.use =</span> <span class="st">"wilcox"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb140-16"><a href="#cb140-16" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">filter</span>(p_val_adj <span class="sc">&lt;=</span> <span class="fl">0.001</span> <span class="sc">&amp;</span> <span class="fu">abs</span>(avg_log2FC)<span class="sc">&gt;</span><span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb140-17"><a href="#cb140-17" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">select</span>(<span class="sc">-</span>p_val)</span>
<span id="cb140-18"><a href="#cb140-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">dim</span>(DEG_table)[<span class="dv">1</span>]<span class="sc">&gt;</span><span class="dv">0</span>){</span>
<span id="cb140-19"><a href="#cb140-19" aria-hidden="true" tabindex="-1"></a>        DEG_table<span class="sc">$</span><span class="st">'cluster'</span> <span class="ot">&lt;-</span> CLUSTER</span>
<span id="cb140-20"><a href="#cb140-20" aria-hidden="true" tabindex="-1"></a>        DEG <span class="ot">&lt;-</span> <span class="fu">rbind</span>(DEG, DEG_table)}</span>
<span id="cb140-21"><a href="#cb140-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>{</span>
<span id="cb140-22"><a href="#cb140-22" aria-hidden="true" tabindex="-1"></a>        <span class="fu">message</span>(<span class="fu">paste0</span>(<span class="st">"---&gt; Warning: No DE genes in cluster "</span>, CLUSTER), <span class="at">appendLF=</span><span class="cn">FALSE</span>)}</span>
<span id="cb140-23"><a href="#cb140-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="fu">paste0</span>(<span class="st">"Done with cluster "</span>,CLUSTER), <span class="at">appendLF=</span><span class="cn">FALSE</span>)</span>
<span id="cb140-24"><a href="#cb140-24" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb140-25"><a href="#cb140-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-26"><a href="#cb140-26" aria-hidden="true" tabindex="-1"></a>DEG <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(DEG)</span>
<span id="cb140-27"><a href="#cb140-27" aria-hidden="true" tabindex="-1"></a>DEG<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">rownames</span>(DEG)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Done with cluster Cortex
Done with cluster Trichoblasts
Done with cluster Root tip
Done with cluster Meristem
Done with cluster Phloem
Done with cluster Pericycle
Done with cluster Stele
Done with cluster Atrichoblasts
Done with cluster Xylem
Done with cluster Endodermis
Done with cluster Quiescent center</code></pre>
</div>
</div>
<p>The table looks like this. Columns represent:</p>
<ul>
<li>average logfoldchange between R7A and Control</li>
<li>percentage of cells in R7A expressing the gene</li>
<li>percentage of cells in Control expressing the gene</li>
<li>adjusted p-value</li>
<li>cluster</li>
<li>gene</li>
</ul>
<p>There are some genes in the table</p>
<div class="cell" data-execution_count="138">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(DEG)[<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
712
</div>
</div>
<p>We now integrate GO terms using a table of predefined GO terms. The function doing that is <code>addGOterms</code>:</p>
<div class="cell" data-tags="[]" data-execution_count="139">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a>go_table <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../Data/LJ_GO_terms.gaf"</span>, <span class="at">skip=</span><span class="dv">6</span>, <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">quote =</span> <span class="st">"</span><span class="sc">\"</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="140">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a>DEG <span class="ot">&lt;-</span> <span class="fu">addGOterms</span>(DEG, go_table, <span class="at">n.cores =</span> <span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>So we can see which GO terms are relevant in each cluster for the infected samples against the control:</p>
<div class="cell" data-scrolled="true" data-execution_count="141">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a>DEG <span class="sc">%&gt;%</span> <span class="fu">filter</span>(cluster<span class="sc">==</span><span class="st">"Cortex"</span> <span class="sc">&amp;</span> GO<span class="sc">!=</span><span class="st">"Undefined"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 3 × 7</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">avg_log2FC</th>
<th data-quarto-table-cell-role="th" scope="col">pct.1</th>
<th data-quarto-table-cell-role="th" scope="col">pct.2</th>
<th data-quarto-table-cell-role="th" scope="col">p_val_adj</th>
<th data-quarto-table-cell-role="th" scope="col">cluster</th>
<th data-quarto-table-cell-role="th" scope="col">gene</th>
<th data-quarto-table-cell-role="th" scope="col">GO</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">LotjaGi2g1v0246800</td>
<td>3.061302</td>
<td>0.414</td>
<td>0.072</td>
<td>1.208504e-294</td>
<td>Cortex</td>
<td>LotjaGi2g1v0246800</td>
<td>L-threonine 3-dehydrogenase; TAIR: AT1G23740.1 Oxidoreductase, zinc-binding dehydrogenase family protein; Swiss-Prot: sp</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">LotjaGi1g1v0066900</td>
<td>1.438645</td>
<td>0.403</td>
<td>0.072</td>
<td>5.606684e-140</td>
<td>Cortex</td>
<td>LotjaGi1g1v0066900</td>
<td>1-aminocyclopropane-1-carboxylate oxidase; TAIR: AT1G05010.1 ethylene-forming enzyme; Swiss-Prot: sp</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">LotjaGi1g1v0154100</td>
<td>4.261879</td>
<td>0.452</td>
<td>0.119</td>
<td>8.705013e-128</td>
<td>Cortex</td>
<td>LotjaGi1g1v0154100</td>
<td>Flavonol synthase; TAIR: AT5G08640.1 flavonol synthase 1; Swiss-Prot: sp</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Finally, we do not expect to find the RINRK1 gene as differentially expressed, because its expression is on average too low.</p>
<div class="cell" data-execution_count="142">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a>DEG <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gene<span class="sc">==</span>RINRK1.id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 0 × 7</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th" scope="col">avg_log2FC</th>
<th data-quarto-table-cell-role="th" scope="col">pct.1</th>
<th data-quarto-table-cell-role="th" scope="col">pct.2</th>
<th data-quarto-table-cell-role="th" scope="col">p_val_adj</th>
<th data-quarto-table-cell-role="th" scope="col">cluster</th>
<th data-quarto-table-cell-role="th" scope="col">gene</th>
<th data-quarto-table-cell-role="th" scope="col">GO</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>
</div>
<p>We can save the table for future use. Such a table can be downloaded and open in any table processor (such as Excel).</p>
<div class="cell" data-tags="[]" data-execution_count="143">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(DEG_table, <span class="st">"DEG_table.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sec-networks" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="sec-networks"><span class="header-section-number">4.2</span> Coexpression analysis</h2>
<p>We use the package hdWGCNA to detect groups of cells expressed simultaneously, and we find which modules are differentially expressed in specific clusters. We look at the GO terms to gain biological insight in the data.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Before running hdWGCNA, we first have to set up the Seurat object. Most of the information computed by hdWGCNA will be stored in the Seurat object’s <code>@misc</code> slot.</p>
</div>
</div>
<p>Here we will set up the Seurat object using the <code>SetupForWGCNA</code> function, specifying the name of the <code>hdWGNCA</code> experiment. This function also selects the genes that will be used for WGCNA. The user can select genes using three different approaches using the posse gene_select parameter:</p>
<ul>
<li>variable: use the genes stored in the Seurat object’s VariableFeatures.</li>
<li>fraction: use genes that are expressed in a certain fraction of cells for in the whole dataset or in each group of cells, specified by <code>group.by</code>.</li>
<li>custom: use genes that are specified in a custom list.</li>
</ul>
<p>In this example, we will select genes that are expressed in at least 5% of cells in this dataset, and we will name our <code>hdWGCNA</code> experiment “tutorial”.</p>
<div class="cell" data-execution_count="144">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">"seurat.clustered.h5Seurat"</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Validating h5Seurat file

Warning message:
“Adding a command log without an assay associated with it”</code></pre>
</div>
</div>
<div class="cell" data-execution_count="145">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">SetupForWGCNA</span>(</span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>  seurat.clustered,</span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_select =</span> <span class="st">"variable"</span>,</span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">fraction =</span> <span class="fl">0.05</span>, <span class="co"># fraction of cells that a gene needs to be expressed in order to be included</span></span>
<span id="cb150-5"><a href="#cb150-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">wgcna_name =</span> <span class="st">"tutorial"</span> <span class="co"># the name of the hdWGCNA experiment</span></span>
<span id="cb150-6"><a href="#cb150-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="construct-metacells" class="level3" data-number="4.2.1">
<h3 data-number="4.2.1" class="anchored" data-anchor-id="construct-metacells"><span class="header-section-number">4.2.1</span> Construct metacells</h3>
<p>After we have set up our Seurat object, the first step in running the hdWGCNA pipeline is to construct metacells from the single-cell dataset. Briefly, metacells are <strong>aggregates of small groups of similar cells originating from the same biological sample of origin</strong>.</p>
<p><img src="./images/hdwgcna.jpg" class="img-fluid" alt="Workflow of hdWGCNA. Similar cells are summed or averaged together to form metacells. The resulting matrix is then used as the basis for inferring modules of coexpressed genes. The metacells expression matrix is the base for all the downstream analysis, such as differential expression of genes modules in specific clusters"> {#fig-hdwgcna, width=800}</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Something more about hdWGCNA">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Something more about hdWGCNA
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The k-Nearest Neighbors (KNN) algorithm is used to identify groups of similar cells to aggregate, and then the average or summed expression of these cells is computed, thus yielding a metacell gene expression matrix (<strong>as if you had many bulk samples</strong>). The <strong>sparsity of the metacell expression matrix is considerably reduced</strong> when compared to the original expression matrix, and therefore it is preferable to use. We were originally motivated to use metacells in place of the original single cells because correlation <strong>network approaches such as WGCNA are sensitive to data sparsity</strong>.</p>
<p><code>hdWGCNA</code> includes a function <code>MetacellsByGroups</code> to construct metacell expression matrices given a single-cell dataset. This function constructs a new Seurat object for the metacell dataset which is stored internally in the <code>hdWGCNA</code> experiment. The <code>group.by</code> parameter determines which groups metacells will be constructed in. We only want to construct metacells from cells that came from the same biological sample of origin, so it is critical to pass that information to <code>hdWGCNA</code> via the <code>group.by</code> parameter. Additionally, we usually construct metacells for each cell type separately. Thus, in this example, we are <strong>grouping by Sample and cell type</strong> to achieve the desired result.</p>
<p><strong>The number of cells to be aggregated <code>k</code> should be tuned based on the size of the input dataset</strong>, in general a lower number for <code>k</code> can be used for small datasets. We generally use <code>k</code> values between 20 and 75. The dataset used for this tutorial has 21,369 cells, and here we use <code>k=30</code>. The amount of allowable overlap between metacells can be tuned using the <code>max_shared</code> argument. There should be a range of <code>k</code> values that are suitable for reducing the sparsity while retaining cellular heterogeneity for a given dataset, rather than a single optimal value.</p>
<p>Note: the authors of <code>hdWGCNA</code> have found that the metacell aggregation approach does not yield good results for extremely underrepresented cell types. For example, in this dataset, the <code>Meristem</code> and <code>Xylem</code> are the least represented, and will be excluded them from this analysis. <code>MetacellsByGroups</code> has a parameter <code>min_cells</code> to exclude groups that are smaller than a specified number of cells. <strong>Errors are likely to arise if the selected value for <code>min_cells</code> is too low</strong>.</p>
</div>
</div>
</div>
<p>Here <strong>we construct metacells and normalize the resulting expression matrix</strong> using the following code (read the additional text in the box above to understand the parameters). Note how the clusters Xylem and Quiescent center are removed, simply because they contain too few cells and cannot be used. So from now on we do not consider them for the coexpression analysis. Note that we use the scaled integrated data matrix (assay <code>integrated</code> and slot <code>scale.data</code>), so that we take advantage of the integration procedure we have been running before to unbias the data from technical features.</p>
<div class="cell" data-scrolled="true" data-execution_count="147">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="co"># construct metacells  in each group</span></span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">MetacellsByGroups</span>(</span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">seurat_obj =</span> seurat.clustered,</span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"predicted.id"</span>, <span class="st">"Condition"</span>), <span class="co"># specify metadata to split by cluster and condition</span></span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">'pca'</span>, <span class="co"># select the dimensionality reduction to perform KNN on</span></span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">k =</span> <span class="dv">30</span>, <span class="co"># nearest-neighbors parameter</span></span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">max_shared =</span> <span class="dv">10</span>, <span class="co"># maximum number of shared cells between two metacells</span></span>
<span id="cb151-8"><a href="#cb151-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ident.group =</span> <span class="st">'predicted.id'</span>, <span class="co"># set the Idents of the metacell seurat object</span></span>
<span id="cb151-9"><a href="#cb151-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">"integrated"</span>,</span>
<span id="cb151-10"><a href="#cb151-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">slot =</span> <span class="st">"scale.data"</span>,</span>
<span id="cb151-11"><a href="#cb151-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_cells =</span> <span class="dv">100</span>,</span>
<span id="cb151-12"><a href="#cb151-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">wgcna_name =</span> <span class="st">"tutorial"</span>,</span>
<span id="cb151-13"><a href="#cb151-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose=</span><span class="cn">TRUE</span></span>
<span id="cb151-14"><a href="#cb151-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message in MetacellsByGroups(seurat_obj = seurat.clustered, group.by = c("predicted.id", :
“Removing the following groups that did not meet min_cells: Quiescent center#Control, Quiescent center#R7A, Xylem#Control, Xylem#R7A”
Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 1.64052287581699
Median shared cells bin-bin: 0

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 1.26674432149097
Median shared cells bin-bin: 1

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 0.346860782529572
Median shared cells bin-bin: 0

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 0.888643725646683
Median shared cells bin-bin: 1

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 1.5741935483871
Median shared cells bin-bin: 0

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 1.83823529411765
Median shared cells bin-bin: 1

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 3.31111111111111
Median shared cells bin-bin: 2

Warning message in (function (seurat_obj, name = "agg", ident.group = "seurat_clusters", :
“On average, more than 10% of cells are shared between paired bins.”
Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 6.7
Median shared cells bin-bin: 8

Warning message in (function (seurat_obj, name = "agg", ident.group = "seurat_clusters", :
“On average, more than 10% of cells are shared between paired bins.”
Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 2.85789473684211
Median shared cells bin-bin: 1

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 1.70824524312896
Median shared cells bin-bin: 1

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 9
Mean shared cells bin-bin: 3
Median shared cells bin-bin: 1.5

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 4.97222222222222
Median shared cells bin-bin: 5

Warning message in (function (seurat_obj, name = "agg", ident.group = "seurat_clusters", :
“On average, more than 10% of cells are shared between paired bins.”
Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 0.676916171024576
Median shared cells bin-bin: 0

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 1.53609831029186
Median shared cells bin-bin: 1

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 1.27272727272727
Median shared cells bin-bin: 0

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 1.43179880647911
Median shared cells bin-bin: 1

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 1.51052048726467
Median shared cells bin-bin: 0

Overlap QC metrics:
Cells per bin: 30
Maximum shared cells bin-bin: 10
Mean shared cells bin-bin: 1.55862775217614
Median shared cells bin-bin: 1
</code></pre>
</div>
</div>
<p>The metacell matrix always needs to be normalized, which is immediately done below</p>
<div class="cell" data-execution_count="148">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="co">#normalize metacell expression matrix:</span></span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">NormalizeMetacells</span>(seurat.clustered)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Cannot find a parent environment called Seurat”</code></pre>
</div>
</div>
<p>Here we specify the expression matrix that we will use for network analysis. It is natural to choose the integrated data matrix (which is chosen with the assay <code>integrated</code> and the matrix slot <code>counts</code>). We still need the data matrix after building the gene coexpression network when we want to evaluate the expression of each cell and gene in the original data.</p>
<div class="cell" data-scrolled="true" data-execution_count="150">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">SetDatExpr</span>(</span>
<span id="cb155-2"><a href="#cb155-2" aria-hidden="true" tabindex="-1"></a>  seurat.clustered,</span>
<span id="cb155-3"><a href="#cb155-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">group_name =</span> <span class="fu">c</span>(<span class="st">'Pericycle'</span>,<span class="st">'Cortex'</span>,<span class="st">'Trichoblasts'</span>,<span class="st">'Root tip'</span>,</span>
<span id="cb155-4"><a href="#cb155-4" aria-hidden="true" tabindex="-1"></a>                 <span class="st">'Phloem'</span>,<span class="st">'Stele'</span>,<span class="st">'Endodermis'</span>,<span class="st">'Atrichoblasts'</span>,</span>
<span id="cb155-5"><a href="#cb155-5" aria-hidden="true" tabindex="-1"></a>                 <span class="st">'Meristem'</span>), <span class="co">#all groups of interest in the data (we use all clusters)</span></span>
<span id="cb155-6"><a href="#cb155-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by=</span><span class="st">'predicted.id'</span>, <span class="co"># the metadata column containing the cell type info. This same column should have also been used in MetacellsByGroups</span></span>
<span id="cb155-7"><a href="#cb155-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">'integrated'</span>, <span class="co"># using integrated assay</span></span>
<span id="cb155-8"><a href="#cb155-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">slot =</span> <span class="st">'counts'</span> <span class="co"># using count data</span></span>
<span id="cb155-9"><a href="#cb155-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="select-soft-power-threshold" class="level3" data-number="4.2.2">
<h3 data-number="4.2.2" class="anchored" data-anchor-id="select-soft-power-threshold"><span class="header-section-number">4.2.2</span> Select soft-power threshold</h3>
<p>Next we will select the soft power threshold. This is an extremely important step in the pipeline. <code>hdWGCNA</code> constructs a gene-gene correlation adjacency matrix to infer co-expression relationships between genes. The correlations are <strong>raised to a power to reduce the amount of noise present in the correlation matrix</strong>, thereby <strong>retaining the strong connections and removing the weak connections</strong>. Therefore, it is critical to determine a proper value for the soft power threshold.</p>
<p>We use the function <code>TestSoftPowers</code> to perform a parameter sweep for different soft power thresholds. This function helps us to guide our choice in a soft power threshold for constructing the co-expression network by inspecting the resulting network topology for different power values. The following code performs the parameter sweep and outputs a summary figure.</p>
<div class="cell" data-scrolled="true" data-execution_count="151">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Test different soft powers:</span></span>
<span id="cb156-2"><a href="#cb156-2" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">TestSoftPowers</span>(<span class="at">powers =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>,</span>
<span id="cb156-3"><a href="#cb156-3" aria-hidden="true" tabindex="-1"></a>  seurat.clustered,</span>
<span id="cb156-4"><a href="#cb156-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">networkType =</span> <span class="st">'signed'</span></span>
<span id="cb156-5"><a href="#cb156-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb156-6"><a href="#cb156-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb156-7"><a href="#cb156-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the results:</span></span>
<span id="cb156-8"><a href="#cb156-8" aria-hidden="true" tabindex="-1"></a>plot_list <span class="ot">&lt;-</span> <span class="fu">PlotSoftPowers</span>(seurat.clustered)</span>
<span id="cb156-9"><a href="#cb156-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb156-10"><a href="#cb156-10" aria-hidden="true" tabindex="-1"></a><span class="co"># assemble with patchwork</span></span>
<span id="cb156-11"><a href="#cb156-11" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(plot_list, <span class="at">ncol=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>pickSoftThreshold: will use block size 2000.
 pickSoftThreshold: calculating connectivity for given powers...
   ..working on genes 1 through 2000 of 2000
   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.  max.k.
1      1   0.7040  5.810          0.864 1.19e+03  1.21e+03 1400.00
2      2   0.3190  1.640          0.800 7.29e+02  7.42e+02 1020.00
3      3   0.0118  0.208          0.806 4.59e+02  4.63e+02  762.00
4      4   0.1290 -0.619          0.869 2.96e+02  2.93e+02  585.00
5      5   0.3790 -1.050          0.929 1.95e+02  1.88e+02  459.00
6      6   0.5690 -1.340          0.961 1.31e+02  1.22e+02  367.00
7      7   0.6620 -1.500          0.965 9.02e+01  8.02e+01  298.00
8      8   0.7310 -1.630          0.977 6.31e+01  5.31e+01  246.00
9      9   0.7900 -1.710          0.979 4.50e+01  3.54e+01  205.00
10    10   0.8280 -1.760          0.983 3.26e+01  2.37e+01  173.00
11    11   0.8590 -1.800          0.988 2.40e+01  1.60e+01  147.00
12    12   0.8690 -1.840          0.986 1.79e+01  1.08e+01  127.00
13    13   0.8730 -1.880          0.984 1.36e+01  7.50e+00  110.00
14    14   0.8930 -1.860          0.988 1.04e+01  5.15e+00   95.70
15    15   0.9030 -1.850          0.986 8.08e+00  3.56e+00   84.00
16    16   0.9120 -1.850          0.994 6.34e+00  2.48e+00   74.20
17    17   0.9150 -1.840          0.989 5.03e+00  1.72e+00   65.80
18    18   0.9200 -1.810          0.992 4.02e+00  1.21e+00   58.70
19    19   0.9090 -1.800          0.973 3.25e+00  8.88e-01   52.50
20    20   0.8780 -1.840          0.940 2.64e+00  6.34e-01   47.20
21    21   0.9180 -1.790          0.987 2.16e+00  4.64e-01   42.60
22    22   0.9280 -1.770          0.994 1.78e+00  3.37e-01   38.50
23    23   0.9380 -1.750          0.995 1.48e+00  2.50e-01   35.00
24    24   0.9390 -1.750          0.992 1.24e+00  1.83e-01   31.90
25    25   0.9410 -1.720          0.993 1.04e+00  1.33e-01   29.10
26    26   0.9440 -1.690          0.990 8.78e-01  9.76e-02   26.60
27    27   0.9200 -1.690          0.955 7.45e-01  7.20e-02   24.40
28    28   0.9280 -1.680          0.957 6.36e-01  5.31e-02   22.50
29    29   0.9340 -1.670          0.956 5.44e-01  4.00e-02   20.70
30    30   0.9320 -1.660          0.951 4.68e-01  2.99e-02   19.10
31    31   0.9540 -1.630          0.978 4.04e-01  2.24e-02   17.70
32    32   0.9550 -1.620          0.976 3.50e-01  1.66e-02   16.40
33    33   0.9550 -1.620          0.976 3.04e-01  1.26e-02   15.30
34    34   0.9550 -1.610          0.970 2.66e-01  9.86e-03   14.20
35    35   0.9630 -1.590          0.980 2.33e-01  7.41e-03   13.20
36    36   0.9510 -1.590          0.961 2.04e-01  5.64e-03   12.40
37    37   0.9520 -1.570          0.958 1.80e-01  4.27e-03   11.60
38    38   0.9580 -1.560          0.960 1.59e-01  3.22e-03   10.80
39    39   0.3870 -2.160          0.317 1.41e-01  2.50e-03   10.20
40    40   0.3890 -2.150          0.317 1.25e-01  1.93e-03    9.54
41    41   0.3900 -2.120          0.323 1.11e-01  1.49e-03    8.97
42    42   0.3920 -2.110          0.323 9.94e-02  1.15e-03    8.44
43    43   0.3920 -2.090          0.323 8.90e-02  8.93e-04    7.96
44    44   0.3910 -2.050          0.315 7.99e-02  6.95e-04    7.51
45    45   0.3940 -2.030          0.317 7.18e-02  5.47e-04    7.10
46    46   0.3950 -2.020          0.319 6.47e-02  4.28e-04    6.71
47    47   0.3960 -2.010          0.318 5.85e-02  3.33e-04    6.36
48    48   0.3990 -1.990          0.319 5.29e-02  2.55e-04    6.02
49    49   0.4000 -1.980          0.318 4.80e-02  1.97e-04    5.72
50    50   0.4000 -1.960          0.315 4.36e-02  1.56e-04    5.43
  Power  SFT.R.sq      slope truncated.R.sq   mean.k. median.k.    max.k.
1     1 0.7044481  5.8105577      0.8637387 1189.8481 1207.0679 1400.9927
2     2 0.3189464  1.6432278      0.7995616  728.9757  741.5904 1017.7691
3     3 0.0118177  0.2081181      0.8058857  458.5484  462.9321  761.7961
4     4 0.1291123 -0.6185606      0.8688956  295.5520  293.2026  584.7981
5     5 0.3787799 -1.0460651      0.9288564  194.8670  188.1986  458.7552
6     6 0.5691444 -1.3404624      0.9608961  131.2447  122.2770  366.6863</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“executing %dopar% sequentially: no parallel backend registered”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-sweep" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-sweep-output-3.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;28: Parameter sweep to choose the soft power threshold</figcaption>
</figure>
</div>
</div>
</div>
<p>The general guidance for <code>hdWGCNA</code> is to pick the lowest soft power threshold that has Model Fit greater than or equal to 0.8. We will do it automatically, so you do not need to do it manually following the illustration fo <a href="#fig-sweep">Figure&nbsp;28</a>.</p>
</section>
</section>
<section id="construction-of-co-expression-network" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="construction-of-co-expression-network"><span class="header-section-number">4.3</span> Construction of co-expression network</h2>
<p>We now have everything that we need to construct our co-expression network. Here we use the <code>hdWGCNA</code> function <code>ConstructNetwork</code>. This function has quite a few parameters to play with if you are an advanced user (read <a href="https://rdrr.io/github/smorabit/hdWGCNA/man/ConstructNetwork.html">this manual</a> and the function <code>ConstructNetwork</code> is based on <a href="https://www.rdocumentation.org/packages/WGCNA/versions/1.72-5/topics/blockwiseConsensusModules">here</a>), but we use default parameters that work well with many single-cell datasets.</p>
<p>The following code selects the power threshold from the values plotted in <a href="#fig-sweep">Figure&nbsp;28</a> and constructs the co-expression network.</p>
<div class="cell" data-execution_count="154">
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="co"># automatic choice of soft power value</span></span>
<span id="cb159-2"><a href="#cb159-2" aria-hidden="true" tabindex="-1"></a>softPowerValue <span class="ot">&lt;-</span> <span class="fu">which</span>(seurat.clustered<span class="sc">@</span>misc<span class="sc">$</span>tutorial<span class="sc">$</span>wgcna_powerTable<span class="sc">$</span>SFT.R.sq <span class="sc">&gt;</span> .<span class="dv">8</span>)[<span class="dv">1</span>]</span>
<span id="cb159-3"><a href="#cb159-3" aria-hidden="true" tabindex="-1"></a><span class="fu">message</span>(<span class="fu">paste0</span>(<span class="st">"Chosen Soft Power Value: "</span>,softPowerValue))</span>
<span id="cb159-4"><a href="#cb159-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb159-5"><a href="#cb159-5" aria-hidden="true" tabindex="-1"></a><span class="co">#multithreading for faster calculations</span></span>
<span id="cb159-6"><a href="#cb159-6" aria-hidden="true" tabindex="-1"></a><span class="fu">enableWGCNAThreads</span>(<span class="at">nThreads =</span> <span class="dv">8</span>)</span>
<span id="cb159-7"><a href="#cb159-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb159-8"><a href="#cb159-8" aria-hidden="true" tabindex="-1"></a><span class="co"># construct co-expression network</span></span>
<span id="cb159-9"><a href="#cb159-9" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">ConstructNetwork</span>(<span class="at">na.rm=</span><span class="cn">TRUE</span>,</span>
<span id="cb159-10"><a href="#cb159-10" aria-hidden="true" tabindex="-1"></a>  seurat.clustered,</span>
<span id="cb159-11"><a href="#cb159-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">soft_power=</span>softPowerValue,                                   </span>
<span id="cb159-12"><a href="#cb159-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">setDatExpr=</span><span class="cn">FALSE</span>,</span>
<span id="cb159-13"><a href="#cb159-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">tom_name =</span> <span class="st">'Network'</span>, <span class="co"># name of the topoligical overlap matrix written to disk</span></span>
<span id="cb159-14"><a href="#cb159-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">overwrite_tom =</span> <span class="cn">TRUE</span>,</span>
<span id="cb159-15"><a href="#cb159-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">randomSeed =</span> <span class="dv">123</span></span>
<span id="cb159-16"><a href="#cb159-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Chosen Soft Power Value: 10

Warning message in ConstructNetwork(na.rm = TRUE, seurat.clustered, soft_power = softPowerValue, :
“Overwriting TOM TOM/Network_TOM.rda”</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Allowing parallel execution with up to 8 working processes.
 Calculating consensus modules and module eigengenes block-wise from all genes
 Calculating topological overlaps block-wise from all genes
   Flagging genes and samples with too many missing values...
    ..step 1
    TOM calculation: adjacency..
    ..will use 8 parallel threads.
     Fraction of slow calculations: 0.000000
    ..connectivity..
    ..matrix multiplication (system BLAS)..
    ..normalization..
    ..done.
 ..Working on block 1 .
 ..Working on block 1 .
 ..merging consensus modules that are too close..</code></pre>
</div>
</div>
<p>We can plot the network dendrogram in <a href="#fig-dendro">Figure&nbsp;29</a>. Each leaf on the dendrogram represents a single gene, and the color at the bottom indicates the co-expression module assignment. Clusters with similar brancing heights contain genes that are more similar to each other in terms of expression patterns when compared to genes in clusters with higher variation in merging heights.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>the “gray” module consists of genes that were not grouped into any co-expression module. The gray module is to be ignored for all downstream analysis and interpretation.**</p>
</div>
</div>
<div class="cell" data-execution_count="155">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="fu">PlotDendrogram</span>(seurat.clustered, <span class="at">main=</span><span class="st">'Phloem hdWGCNA Dendrogram'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-dendro" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-dendro-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;29: Dendrogram of the genes coexpression network. Leaves at the bottom are genes, and the colours represent a module of coexpression for the corresponding genes. Gray areas are for genes not included in any module.</figcaption>
</figure>
</div>
</div>
</div>
<section id="module-eigengenes-mes" class="level3" data-number="4.3.1">
<h3 data-number="4.3.1" class="anchored" data-anchor-id="module-eigengenes-mes"><span class="header-section-number">4.3.1</span> Module Eigengenes (MEs)</h3>
<p>We calculate harmonized module eigengenes. This is a way of doing PCA of the expression matrix of metacells including the genes of one coexpression module at a time. Thus we have a PCA of the data for each module. The first principal component of each PCA is called module eigengene: it is enough to distinguish one module from all the others and characterize the expression pattern of the module (<span class="citation" data-cites="langfelder_eigengene_2007">Langfelder and Horvath (<a href="#ref-langfelder_eigengene_2007" role="doc-biblioref">2007</a>)</span>)!</p>
<p>Dimensionality reduction techniques are a very hot topic in single-cell genomics (<span class="citation" data-cites="xiang_comparison_2021">Xiang et al. (<a href="#ref-xiang_comparison_2021" role="doc-biblioref">2021</a>)</span>). It is well known that technical artifacts can muddy the analysis of single-cell datasets, and over the years there have been many methods that aim to reduce the effects of these artifacts. Therefore it stands to reason that MEs would be subject to these technical artifacts as well, and <code>hdWGCNA</code> seeks to alleviate these effects by using the integration software <code>Harmony</code> (<span class="citation" data-cites="korsunsky_fast_2019">Korsunsky et al. (<a href="#ref-korsunsky_fast_2019" role="doc-biblioref">2019</a>)</span>).</p>
<p>The following code performs the module eigengene computation harmonizing by the Sample of origin using the <code>group.by.vars</code> parameter. We need to run <code>ScaleData</code> to avoid errors from the <code>harmony</code> software.</p>
<div class="cell" data-scrolled="true" data-execution_count="156">
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="co"># need to run ScaleData first or else harmony throws an error:</span></span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(seurat.clustered, </span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">features=</span><span class="fu">VariableFeatures</span>(seurat.clustered), </span>
<span id="cb163-4"><a href="#cb163-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We still want to use the integrated data matrix to calculate the modules eigengenes, so we assign the scaled matrix to the <code>counts</code> slot, which is where <code>hdWGCNA</code> looks by default for computing module eigengenes.</p>
<div class="cell" data-execution_count="157">
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>seurat.clustered<span class="sc">@</span>assays<span class="sc">$</span>integrated<span class="sc">@</span>counts <span class="ot">&lt;-</span> seurat.clustered<span class="sc">@</span>assays<span class="sc">$</span>integrated<span class="sc">@</span>scale.data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-scrolled="true" data-execution_count="158">
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute all MEs in the full single-cell dataset</span></span>
<span id="cb165-2"><a href="#cb165-2" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">ModuleEigengenes</span>( </span>
<span id="cb165-3"><a href="#cb165-3" aria-hidden="true" tabindex="-1"></a> seurat.clustered,</span>
<span id="cb165-4"><a href="#cb165-4" aria-hidden="true" tabindex="-1"></a> <span class="at">group.by.vars=</span><span class="st">"orig.ident"</span>,</span>
<span id="cb165-5"><a href="#cb165-5" aria-hidden="true" tabindex="-1"></a> <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb165-6"><a href="#cb165-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "yellow"
[1] "turquoise"
[1] "tan"
[1] "brown"
[1] "red"
[1] "lightcyan"
[1] "cyan"
[1] "pink"
[1] "salmon"
[1] "grey"
[1] "purple"
[1] "greenyellow"
[1] "green"
[1] "midnightblue"
[1] "blue"
[1] "black"
[1] "magenta"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcayellow to pcayellow_”
Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcaturquoise to pcaturquoise_”
Warning message:
“Key ‘harmony_’ taken, using ‘xlofa_’ instead”
Centering and scaling data matrix

Warning message in irlba(A = t(x = object), nv = npcs, ...):
“You're computing too large a percentage of total singular values, use a standard svd instead.”
Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcatan to pcatan_”
Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcabrown to pcabrown_”
Warning message:
“Key ‘harmony_’ taken, using ‘ujakm_’ instead”
Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcared to pcared_”
Centering and scaling data matrix

Warning message in irlba(A = t(x = object), nv = npcs, ...):
“You're computing too large a percentage of total singular values, use a standard svd instead.”
Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcalightcyan to pcalightcyan_”
Warning message:
“Key ‘harmony_’ taken, using ‘uyrtn_’ instead”
Centering and scaling data matrix

Warning message in irlba(A = t(x = object), nv = npcs, ...):
“You're computing too large a percentage of total singular values, use a standard svd instead.”
Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcacyan to pcacyan_”
Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcapink to pcapink_”
Warning message:
“Key ‘harmony_’ taken, using ‘sakpn_’ instead”
Centering and scaling data matrix

Warning message in irlba(A = t(x = object), nv = npcs, ...):
“You're computing too large a percentage of total singular values, use a standard svd instead.”
Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcasalmon to pcasalmon_”
Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcagrey to pcagrey_”
Warning message:
“Key ‘harmony_’ taken, using ‘esfhp_’ instead”
Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcapurple to pcapurple_”
Centering and scaling data matrix

Warning message in irlba(A = t(x = object), nv = npcs, ...):
“You're computing too large a percentage of total singular values, use a standard svd instead.”
Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcagreenyellow to pcagreenyellow_”
Warning message:
“Key ‘harmony_’ taken, using ‘wajvi_’ instead”
Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcagreen to pcagreen_”
Centering and scaling data matrix

Warning message in irlba(A = t(x = object), nv = npcs, ...):
“You're computing too large a percentage of total singular values, use a standard svd instead.”
Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcamidnightblue to pcamidnightblue_”
Warning message:
“Key ‘harmony_’ taken, using ‘rtnba_’ instead”
Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcablue to pcablue_”
Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcablack to pcablack_”
Warning message:
“Key ‘harmony_’ taken, using ‘logay_’ instead”
Centering and scaling data matrix

Warning message:
“Keys should be one or more alphanumeric characters followed by an underscore, setting key from pcamagenta to pcamagenta_”</code></pre>
</div>
</div>
</section>
<section id="compute-module-connectivity" class="level3" data-number="4.3.2">
<h3 data-number="4.3.2" class="anchored" data-anchor-id="compute-module-connectivity"><span class="header-section-number">4.3.2</span> Compute module connectivity</h3>
<p>In co-expression network analysis, we often want to focus on the <strong>hub genes, those which are highly connected within each module</strong>. Therefore we wish to determine the eigengene-based connectivity, also known as kME, of each gene. The <code>ModuleConnectivity</code> function computes the kMEs in the single-cell dataset (rather than the metacell dataset). This function essentially computes pairwise correlations between genes and module eigengenes. kME can be computed for all cells in the dataset, but it is recommended computing kMEs in the cell type(s) or group(s) previously used to run <code>ConstructNetwork</code>.</p>
<div class="cell" data-execution_count="159">
<div class="sourceCode cell-code" id="cb168"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute eigengene-based connectivity (kME):</span></span>
<span id="cb168-2"><a href="#cb168-2" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">ModuleConnectivity</span>(</span>
<span id="cb168-3"><a href="#cb168-3" aria-hidden="true" tabindex="-1"></a>  seurat.clustered,</span>
<span id="cb168-4"><a href="#cb168-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">'predicted.id'</span>, </span>
<span id="cb168-5"><a href="#cb168-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">group_name =</span> <span class="fu">c</span>(<span class="st">'Pericycle'</span>,<span class="st">'Cortex'</span>,<span class="st">'Trichoblasts'</span>,<span class="st">'Root tip'</span>,</span>
<span id="cb168-6"><a href="#cb168-6" aria-hidden="true" tabindex="-1"></a>                 <span class="st">'Phloem'</span>,<span class="st">'Stele'</span>,<span class="st">'Endodermis'</span>,<span class="st">'Atrichoblasts'</span>,</span>
<span id="cb168-7"><a href="#cb168-7" aria-hidden="true" tabindex="-1"></a>                 <span class="st">'Meristem'</span>),</span>
<span id="cb168-8"><a href="#cb168-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For convenience, we re-name the hdWGCNA modules to indicate such that they are not called <code>cluster##</code>, but <code>Lotus-Mod##</code> instead, which is a more meaningful name.</p>
<div class="cell" data-execution_count="160">
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rename the modules</span></span>
<span id="cb169-2"><a href="#cb169-2" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">ResetModuleNames</span>(</span>
<span id="cb169-3"><a href="#cb169-3" aria-hidden="true" tabindex="-1"></a>  seurat.clustered,</span>
<span id="cb169-4"><a href="#cb169-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">new_name =</span> <span class="st">"Lotus-Mod"</span></span>
<span id="cb169-5"><a href="#cb169-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can visualize the genes in each module ranked by kME using the <code>PlotKMEs</code> function. Each plot shows the kME-score (which is at most 1) for a module. On the left side of the y axis you see the kME scores (how much a gene is well-correlated with others in the module). On the right side of the y axis you have a bar showing how many genes have the corresponding kME score. The x axis should show the number of genes corresponding to the bar (but no number is shown by this plotting function). What is important to observe is that only few genes have high kME score in each module, and those are the ones you want to focus on. The top ten names are plotted for each module as well.</p>
<div class="cell" data-execution_count="161">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">20</span>, <span class="at">repr.plot.height=</span><span class="dv">25</span>)</span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">capture.output</span>({<span class="fu">PlotKMEs</span>(seurat.clustered, <span class="at">ncol=</span><span class="dv">3</span>, <span class="at">text_size =</span> <span class="dv">4</span>);})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-kme" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-kme-output-1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;30: KMEs for each module, where the genes on the right of each plot gives the top genes based on the module’s KMEs. Each plot is a histogram which has been rotated, so that bars starting on the left side of the y axis represent the counts of how many genes have the connectivity written on the right side of the y axis. The number of counts should be written on the x axis, but is not plotted by the function.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="getting-the-module-assignment-table" class="level3" data-number="4.3.3">
<h3 data-number="4.3.3" class="anchored" data-anchor-id="getting-the-module-assignment-table"><span class="header-section-number">4.3.3</span> Getting the module assignment table</h3>
<p>The plots of <a href="#fig-kme">Figure&nbsp;30</a> are a bit hard to read, though fancy to plot. <code>hdWGCNA</code> allows for easy access of the module assignment table using the <code>GetModules</code> function. This table consists of three columns: <code>gene_name</code> stores the gene’s symbol or ID, module stores the gene’s module assignment, and color stores a color mapping for each module (which is used in plotting steps). If <code>ModuleConnectivity</code> has been used on this hdWGCNA experiment, as it is our case, this table has additional columns with the connectivities plotted in <a href="#fig-kme">Figure&nbsp;30</a></p>
<div class="cell" data-execution_count="162">
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the module assignment table:</span></span>
<span id="cb171-2"><a href="#cb171-2" aria-hidden="true" tabindex="-1"></a>modules <span class="ot">&lt;-</span> <span class="fu">GetModules</span>(seurat.clustered)</span>
<span id="cb171-3"><a href="#cb171-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-4"><a href="#cb171-4" aria-hidden="true" tabindex="-1"></a><span class="co"># show the first 6 columns:</span></span>
<span id="cb171-5"><a href="#cb171-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(modules[,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 6</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">gene_name</th>
<th data-quarto-table-cell-role="th" scope="col">module</th>
<th data-quarto-table-cell-role="th" scope="col">color</th>
<th data-quarto-table-cell-role="th" scope="col">kME_Lotus-Mod1</th>
<th data-quarto-table-cell-role="th" scope="col">kME_Lotus-Mod2</th>
<th data-quarto-table-cell-role="th" scope="col">kME_Lotus-Mod3</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">LotjaGi3g1v0222100</td>
<td>LotjaGi3g1v0222100</td>
<td>Lotus-Mod1</td>
<td>yellow</td>
<td>0.66927700</td>
<td>-0.16607049</td>
<td>-0.20407421</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">LotjaGi2g1v0360900</td>
<td>LotjaGi2g1v0360900</td>
<td>Lotus-Mod2</td>
<td>turquoise</td>
<td>-0.14382340</td>
<td>0.90236572</td>
<td>-0.24851350</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">LotjaGi3g1v0445300</td>
<td>LotjaGi3g1v0445300</td>
<td>Lotus-Mod3</td>
<td>tan</td>
<td>-0.18127709</td>
<td>-0.21469999</td>
<td>0.86945694</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">LotjaGi3g1v0506700</td>
<td>LotjaGi3g1v0506700</td>
<td>Lotus-Mod2</td>
<td>turquoise</td>
<td>-0.13031149</td>
<td>0.72715888</td>
<td>-0.20028857</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">LotjaGi2g1v0160200</td>
<td>LotjaGi2g1v0160200</td>
<td>Lotus-Mod4</td>
<td>brown</td>
<td>-0.12981105</td>
<td>-0.02298910</td>
<td>-0.08982969</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">LotjaGi5g1v0119900</td>
<td>LotjaGi5g1v0119900</td>
<td>Lotus-Mod5</td>
<td>red</td>
<td>0.02807917</td>
<td>-0.09403294</td>
<td>-0.17121588</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>A table of the top N hub genes sorted by kME can be extracted using the <code>GetHubGenes</code> function. Here we choose the top 10 genes, but you can change the value if you want. The table contains all the top 10 genes for each module, the name of the module they come from and the kME score.</p>
<div class="cell" data-scrolled="true" data-execution_count="163">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get hub genes</span></span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a>hub_df <span class="ot">&lt;-</span> <span class="fu">GetHubGenes</span>(seurat.clustered, <span class="at">n_hubs =</span> <span class="dv">10</span>)</span>
<span id="cb172-3"><a href="#cb172-3" aria-hidden="true" tabindex="-1"></a><span class="co"># print the first lines of the table to see its structure</span></span>
<span id="cb172-4"><a href="#cb172-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>( hub_df )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 3</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">gene_name</th>
<th data-quarto-table-cell-role="th" scope="col">module</th>
<th data-quarto-table-cell-role="th" scope="col">kME</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">1</td>
<td>LotjaGi1g1v0160700</td>
<td>Lotus-Mod1</td>
<td>0.8112420</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">2</td>
<td>LotjaGi6g1v0003600</td>
<td>Lotus-Mod1</td>
<td>0.7821258</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">3</td>
<td>LotjaGi1g1v0783700-LC</td>
<td>Lotus-Mod1</td>
<td>0.7813964</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">4</td>
<td>LotjaGi6g1v0273300-LC</td>
<td>Lotus-Mod1</td>
<td>0.7544668</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">5</td>
<td>LotjaGi5g1v0047400</td>
<td>Lotus-Mod1</td>
<td>0.7442475</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">6</td>
<td>LotjaGi5g1v0188800</td>
<td>Lotus-Mod1</td>
<td>0.7403723</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Again, we can assign GO terms to better check associated functions</p>
<div class="cell" data-tags="[]" data-execution_count="164">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>go_table <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"../Data/LJ_GO_terms.gaf"</span>, <span class="at">skip=</span><span class="dv">6</span>, <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">fill=</span><span class="cn">TRUE</span>, <span class="at">quote =</span> <span class="st">"</span><span class="sc">\"</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-execution_count="165">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a>hub_df <span class="ot">&lt;-</span> <span class="fu">addGOterms</span>(<span class="at">input_table =</span> hub_df, </span>
<span id="cb174-2"><a href="#cb174-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">go_table =</span> go_table, </span>
<span id="cb174-3"><a href="#cb174-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">gene_column =</span> <span class="st">'gene_name'</span>,</span>
<span id="cb174-4"><a href="#cb174-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">n.cores =</span> <span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now it is easy to look at each module and relevant GO terms.</p>
<div class="cell" data-execution_count="166">
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a>hub_filtered <span class="ot">&lt;-</span> hub_df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(module <span class="sc">==</span> <span class="st">"Lotus-Mod3"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="167">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a>hub_filtered</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 10 × 4</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th" scope="col">gene_name</th>
<th data-quarto-table-cell-role="th" scope="col">module</th>
<th data-quarto-table-cell-role="th" scope="col">kME</th>
<th data-quarto-table-cell-role="th" scope="col">GO</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;fct&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>LotjaGi3g1v0445300</td>
<td>Lotus-Mod3</td>
<td>0.8694569</td>
<td>Undefined</td>
</tr>
<tr class="even">
<td>LotjaGi4g1v0275500</td>
<td>Lotus-Mod3</td>
<td>0.8430255</td>
<td>Undefined</td>
</tr>
<tr class="odd">
<td>LotjaGi4g1v0208100</td>
<td>Lotus-Mod3</td>
<td>0.7431116</td>
<td>Undefined</td>
</tr>
<tr class="even">
<td>LotjaGi1g1v0594900</td>
<td>Lotus-Mod3</td>
<td>0.7429926</td>
<td>Undefined</td>
</tr>
<tr class="odd">
<td>LotjaGi3g1v0505900</td>
<td>Lotus-Mod3</td>
<td>0.7373019</td>
<td>Undefined</td>
</tr>
<tr class="even">
<td>LotjaGi2g1v0406200</td>
<td>Lotus-Mod3</td>
<td>0.7185567</td>
<td>Undefined</td>
</tr>
<tr class="odd">
<td>LotjaGi6g1v0151500</td>
<td>Lotus-Mod3</td>
<td>0.7072212</td>
<td>Undefined</td>
</tr>
<tr class="even">
<td>LotjaGi6g1v0028000-LC</td>
<td>Lotus-Mod3</td>
<td>0.7023895</td>
<td>Undefined</td>
</tr>
<tr class="odd">
<td>LotjaGi1g1v0502700</td>
<td>Lotus-Mod3</td>
<td>0.7008408</td>
<td>Undefined</td>
</tr>
<tr class="even">
<td>LotjaGi2g1v0402200</td>
<td>Lotus-Mod3</td>
<td>0.6794909</td>
<td>Undefined</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We save the table for later use</p>
<div class="cell" data-execution_count="168">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(hub_df, <span class="st">"top_genes_networks.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also plot the modules to see where they are most relevant on the UMAP plot. First we calculate the scores of each module and renames each module from <code>Cluster##</code> to <code>Lotus-Mod##</code>. The scores are the same for the markers plots for clustering, but calculated instead on the genes of each module.</p>
<div class="cell" data-execution_count="169">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a>features_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb178-2"><a href="#cb178-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-3"><a href="#cb178-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(MOD <span class="cf">in</span> hub_df<span class="sc">$</span>module){</span>
<span id="cb178-4"><a href="#cb178-4" aria-hidden="true" tabindex="-1"></a>    genes <span class="ot">&lt;-</span> hub_df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(module <span class="sc">==</span> MOD)</span>
<span id="cb178-5"><a href="#cb178-5" aria-hidden="true" tabindex="-1"></a>    genes_mod <span class="ot">&lt;-</span> genes<span class="sc">$</span>gene_name</span>
<span id="cb178-6"><a href="#cb178-6" aria-hidden="true" tabindex="-1"></a>    features_list[[MOD]] <span class="ot">&lt;-</span> genes_mod</span>
<span id="cb178-7"><a href="#cb178-7" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb178-8"><a href="#cb178-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-9"><a href="#cb178-9" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">AddModuleScore</span>(</span>
<span id="cb178-10"><a href="#cb178-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> seurat.clustered,</span>
<span id="cb178-11"><a href="#cb178-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> features_list,</span>
<span id="cb178-12"><a href="#cb178-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">ctrl =</span> <span class="dv">5</span>,</span>
<span id="cb178-13"><a href="#cb178-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb178-14"><a href="#cb178-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-15"><a href="#cb178-15" aria-hidden="true" tabindex="-1"></a>seurat.clustered <span class="ot">&lt;-</span> <span class="fu">renameScores</span>(<span class="at">markers_list =</span> features_list, <span class="at">seurat_data =</span> seurat.clustered)</span>
<span id="cb178-16"><a href="#cb178-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plotScoresUMAP</span>(<span class="at">markers_list =</span> features_list, <span class="at">seurat_data =</span> seurat.clustered)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scores renamed FROM




TO



</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Cluster1
Cluster2
Cluster3
Cluster4
Cluster5
Cluster6
Cluster7
Cluster8
Cluster9
Cluster10
Cluster11
Cluster12
Cluster13
Cluster14
Cluster15
Cluster16
Lotus-Mod1
Lotus-Mod2
Lotus-Mod3
Lotus-Mod4
Lotus-Mod5
Lotus-Mod6
Lotus-Mod7
Lotus-Mod8
Lotus-Mod9
Lotus-Mod10
Lotus-Mod11
Lotus-Mod12
Lotus-Mod13
Lotus-Mod14
Lotus-Mod15
Lotus-Mod16</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="notebook.source_files/figure-html/cell-118-output-3.png" class="img-fluid"></p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Plots can be hard to see when they are small. You can plot any module singularly. For example Mod1. You simply need to choose one of the elements from <code>feature_list</code> (below we chose <code>feature_list[1]</code> for <code>Lotus-Mod1</code>)</p>
</div>
</div>
<div class="cell" data-execution_count="170">
<div class="sourceCode cell-code" id="cb181"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotScoresUMAP</span>(<span class="at">markers_list =</span> features_list[<span class="dv">1</span>], <span class="at">seurat_data =</span> seurat.clustered)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="notebook.source_files/figure-html/cell-119-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>We save the seurat object and the network separately (because for some inscrutable error, <code>hdWGCNA</code> cannot load together with the Seurat object when it is needed again).</p>
<div class="cell" data-execution_count="171">
<div class="sourceCode cell-code" id="cb182"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb182-1"><a href="#cb182-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SaveH5Seurat</span>(seurat.clustered, <span class="st">'seurat.network.h5Seurat'</span>, <span class="at">overwrite =</span> <span class="cn">TRUE</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Overwriting previous file seurat.network.h5Seurat”
Creating h5Seurat file for version 3.1.5.9900
</code></pre>
</div>
</div>
<div class="cell" data-execution_count="172">
<div class="sourceCode cell-code" id="cb184"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb184-1"><a href="#cb184-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(seurat.clustered<span class="sc">@</span>misc, <span class="st">"network_lotus.RDS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="differential-module-expression-dme-analysis" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="differential-module-expression-dme-analysis"><span class="header-section-number">4.4</span> Differential module Expression (DME) analysis</h2>
<p>Lastly, we can see which modules are most expressed in a specific cluster against the others, in a similar way to differential gene expression.</p>
<div class="cell" data-scrolled="true" data-execution_count="173">
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a>seurat.network <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">'seurat.network.h5Seurat'</span>, <span class="at">misc=</span><span class="cn">FALSE</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Validating h5Seurat file

Warning message:
“Adding a command log without an assay associated with it”</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Couldn't delete        144115188075857558"</code></pre>
</div>
</div>
<div class="cell" data-execution_count="174">
<div class="sourceCode cell-code" id="cb188"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb188-1"><a href="#cb188-1" aria-hidden="true" tabindex="-1"></a>seurat.network<span class="sc">@</span>misc <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"network_lotus.RDS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we discuss how to perform <strong>DME testing between two different groups</strong>. We use the <code>hdWGCNA</code> function <code>FindDMEs</code>, which is similar to the Seurat function FindMarkers. We use the Mann-Whitney U test, also known as the Wilcoxon test, to compare two groups, but other tests can be used at the user’s discretion with the <code>test.use</code> parameter.</p>
<p>We are interested in defining our two groups both by condition. We filter DMEs by p-value and fold-change, so to remove non-significant ones from the list.</p>
<div class="cell" data-execution_count="175">
<div class="sourceCode cell-code" id="cb189"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb189-1"><a href="#cb189-1" aria-hidden="true" tabindex="-1"></a>R7Agroup <span class="ot">&lt;-</span> seurat.network<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">subset</span>(Condition <span class="sc">==</span> <span class="st">'R7A'</span>) <span class="sc">%&gt;%</span> rownames</span>
<span id="cb189-2"><a href="#cb189-2" aria-hidden="true" tabindex="-1"></a>WTgroup <span class="ot">&lt;-</span> seurat.network<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">subset</span>(Condition <span class="sc">==</span> <span class="st">'Control'</span>) <span class="sc">%&gt;%</span> rownames</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="176">
<div class="sourceCode cell-code" id="cb190"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb190-1"><a href="#cb190-1" aria-hidden="true" tabindex="-1"></a>DMEs <span class="ot">&lt;-</span> <span class="fu">FindDMEs</span>(</span>
<span id="cb190-2"><a href="#cb190-2" aria-hidden="true" tabindex="-1"></a>  seurat.network,</span>
<span id="cb190-3"><a href="#cb190-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">barcodes1 =</span> R7Agroup,</span>
<span id="cb190-4"><a href="#cb190-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">barcodes2 =</span> WTgroup,</span>
<span id="cb190-5"><a href="#cb190-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">test.use=</span><span class="st">'wilcox'</span>,</span>
<span id="cb190-6"><a href="#cb190-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">wgcna_name=</span><span class="st">'tutorial'</span></span>
<span id="cb190-7"><a href="#cb190-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(p_val_adj <span class="sc">&lt;</span> <span class="fl">0.001</span> <span class="sc">&amp;</span> <span class="fu">abs</span>(avg_log2FC)<span class="sc">&gt;</span><span class="dv">1</span>) <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="sc">-</span>p_val)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The resulting table below is very similar to the one for differential gene expression. Now the p-values and fold changes are referred to the differential module expression in the inoculated VS the control cells. We can see there are some differences and one could study those modules in depth by looking at the top genes and their GO terms.</p>
<div class="cell" data-execution_count="177">
<div class="sourceCode cell-code" id="cb191"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1" aria-hidden="true" tabindex="-1"></a>DMEs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 3 × 5</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">avg_log2FC</th>
<th data-quarto-table-cell-role="th" scope="col">pct.1</th>
<th data-quarto-table-cell-role="th" scope="col">pct.2</th>
<th data-quarto-table-cell-role="th" scope="col">p_val_adj</th>
<th data-quarto-table-cell-role="th" scope="col">module</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Lotus-Mod2</td>
<td>-1.078807</td>
<td>0.130</td>
<td>0.239</td>
<td>5.577347e-93</td>
<td>Lotus-Mod2</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Lotus-Mod6</td>
<td>1.120521</td>
<td>0.276</td>
<td>0.193</td>
<td>8.258406e-49</td>
<td>Lotus-Mod6</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Lotus-Mod15</td>
<td>2.009786</td>
<td>0.215</td>
<td>0.165</td>
<td>5.693663e-27</td>
<td>Lotus-Mod15</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Graphically, you can always produce a lollipop plot as in <a href="#fig-lol">Figure&nbsp;31</a>, where the size of each circle is the p-value, and the x-axis is the log-fold change. A cross on a circle means the p-value is not significant. In our case we filtered out p-values that were too high, so we have only significant modules in the plot.</p>
<div class="cell" data-execution_count="178">
<div class="sourceCode cell-code" id="cb192"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb192-1"><a href="#cb192-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">6</span>, <span class="at">repr.plot.height=</span><span class="dv">6</span>)</span>
<span id="cb192-2"><a href="#cb192-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb192-3"><a href="#cb192-3" aria-hidden="true" tabindex="-1"></a><span class="fu">PlotDMEsLollipop</span>(</span>
<span id="cb192-4"><a href="#cb192-4" aria-hidden="true" tabindex="-1"></a>  seurat.network, </span>
<span id="cb192-5"><a href="#cb192-5" aria-hidden="true" tabindex="-1"></a>  DMEs, </span>
<span id="cb192-6"><a href="#cb192-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">wgcna_name=</span><span class="st">'tutorial'</span>, </span>
<span id="cb192-7"><a href="#cb192-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">pvalue =</span> <span class="st">"p_val_adj"</span>,</span>
<span id="cb192-8"><a href="#cb192-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ggforestplot
</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Please be aware comparison group/groups are not provided, which may casue an ERROR. PlotDMEsLollipop function will automatically assume all values are within the same group."</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-lol" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-lol-output-3.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;31: Lollipop plot that shovs the average log-fold change of each module and the p-value (size of each circle). Crossed circles, when present, have a non-significant p-value</figcaption>
</figure>
</div>
</div>
</div>
<section id="one-versus-all-dme-analysis" class="level3" data-number="4.4.1">
<h3 data-number="4.4.1" class="anchored" data-anchor-id="one-versus-all-dme-analysis"><span class="header-section-number">4.4.1</span> One-versus-all DME analysis</h3>
<p>This is another case of DME analysis, where each cluster is tested against the rest of the data to see which modules are differentially expressed and are specific of each cell type. We can test by cell type (<code>predicted.id</code>) or by condition (<code>Condition</code>), or by any other label a dataset might have and that is contained in the meta data.</p>
<div class="cell" data-execution_count="179">
<div class="sourceCode cell-code" id="cb195"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb195-1"><a href="#cb195-1" aria-hidden="true" tabindex="-1"></a>DMEs_all <span class="ot">&lt;-</span> <span class="fu">FindAllDMEs</span>(</span>
<span id="cb195-2"><a href="#cb195-2" aria-hidden="true" tabindex="-1"></a>  seurat.network,</span>
<span id="cb195-3"><a href="#cb195-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">'predicted.id'</span>,</span>
<span id="cb195-4"><a href="#cb195-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">wgcna_name =</span> <span class="st">'tutorial'</span></span>
<span id="cb195-5"><a href="#cb195-5" aria-hidden="true" tabindex="-1"></a>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Cortex"
[1] "Trichoblasts"
[1] "Root tip"
[1] "Meristem"
[1] "Phloem"
[1] "Pericycle"
[1] "Stele"
[1] "Atrichoblasts"
[1] "Xylem"
[1] "Endodermis"
[1] "Quiescent center"</code></pre>
</div>
</div>
<p>The table is usually big, but you can choose to filter by various parameters as below, and to look only at one cluster of interest</p>
<div class="cell" data-execution_count="180">
<div class="sourceCode cell-code" id="cb197"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb197-1"><a href="#cb197-1" aria-hidden="true" tabindex="-1"></a>DMEs_cortex <span class="ot">&lt;-</span> DMEs_all <span class="sc">%&gt;%</span> <span class="fu">filter</span>(p_val_adj <span class="sc">&lt;</span> .<span class="dv">001</span> <span class="sc">&amp;</span> <span class="fu">abs</span>(avg_log2FC)<span class="sc">&gt;</span><span class="dv">1</span> </span>
<span id="cb197-2"><a href="#cb197-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="sc">&amp;</span> <span class="fu">is.finite</span>(avg_log2FC)</span>
<span id="cb197-3"><a href="#cb197-3" aria-hidden="true" tabindex="-1"></a>                                       <span class="sc">&amp;</span> group<span class="sc">==</span><span class="st">'Cortex'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The DME analysis results in a lot of significant results for the cortex. The most significant modules for a cluster are the ones where <code>pct.1</code> and <code>pct.2</code> are very different, and at the same time have a large fold-change and a small p-value. The column <code>group</code> tells you in which cluster the module is differentially overexpressed (positive fold-change <code>avg_log2FC</code>) or underexpressed (negative fold-change).</p>
<div class="cell" data-execution_count="181">
<div class="sourceCode cell-code" id="cb198"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb198-1"><a href="#cb198-1" aria-hidden="true" tabindex="-1"></a>DMEs_cortex</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 14 × 7</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">p_val</th>
<th data-quarto-table-cell-role="th" scope="col">avg_log2FC</th>
<th data-quarto-table-cell-role="th" scope="col">pct.1</th>
<th data-quarto-table-cell-role="th" scope="col">pct.2</th>
<th data-quarto-table-cell-role="th" scope="col">p_val_adj</th>
<th data-quarto-table-cell-role="th" scope="col">module</th>
<th data-quarto-table-cell-role="th" scope="col">group</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod3</td>
<td>0.000000e+00</td>
<td>2.371340</td>
<td>0.480</td>
<td>0.152</td>
<td>0.000000e+00</td>
<td>Lotus-Mod3</td>
<td>Cortex</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod6</td>
<td>0.000000e+00</td>
<td>2.666149</td>
<td>0.408</td>
<td>0.125</td>
<td>0.000000e+00</td>
<td>Lotus-Mod6</td>
<td>Cortex</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod14</td>
<td>0.000000e+00</td>
<td>2.453980</td>
<td>0.523</td>
<td>0.225</td>
<td>0.000000e+00</td>
<td>Lotus-Mod14</td>
<td>Cortex</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod1</td>
<td>0.000000e+00</td>
<td>-4.708455</td>
<td>0.061</td>
<td>0.277</td>
<td>0.000000e+00</td>
<td>Lotus-Mod1</td>
<td>Cortex</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod7</td>
<td>0.000000e+00</td>
<td>-4.395113</td>
<td>0.082</td>
<td>0.370</td>
<td>0.000000e+00</td>
<td>Lotus-Mod7</td>
<td>Cortex</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod11</td>
<td>0.000000e+00</td>
<td>-4.699137</td>
<td>0.034</td>
<td>0.296</td>
<td>0.000000e+00</td>
<td>Lotus-Mod11</td>
<td>Cortex</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod8</td>
<td>0.000000e+00</td>
<td>-3.563797</td>
<td>0.108</td>
<td>0.368</td>
<td>0.000000e+00</td>
<td>Lotus-Mod8</td>
<td>Cortex</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod15</td>
<td>0.000000e+00</td>
<td>-2.962244</td>
<td>0.071</td>
<td>0.289</td>
<td>0.000000e+00</td>
<td>Lotus-Mod15</td>
<td>Cortex</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod2</td>
<td>0.000000e+00</td>
<td>-5.524643</td>
<td>0.033</td>
<td>0.273</td>
<td>0.000000e+00</td>
<td>Lotus-Mod2</td>
<td>Cortex</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod4</td>
<td>1.859292e-188</td>
<td>-3.855152</td>
<td>0.070</td>
<td>0.211</td>
<td>2.974867e-187</td>
<td>Lotus-Mod4</td>
<td>Cortex</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod13</td>
<td>3.339974e-141</td>
<td>-3.844141</td>
<td>0.073</td>
<td>0.194</td>
<td>5.343958e-140</td>
<td>Lotus-Mod13</td>
<td>Cortex</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod9</td>
<td>2.981487e-99</td>
<td>-2.893193</td>
<td>0.147</td>
<td>0.254</td>
<td>4.770380e-98</td>
<td>Lotus-Mod9</td>
<td>Cortex</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod10</td>
<td>3.126878e-65</td>
<td>-3.120111</td>
<td>0.108</td>
<td>0.188</td>
<td>5.003005e-64</td>
<td>Lotus-Mod10</td>
<td>Cortex</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Cortex.Lotus-Mod12</td>
<td>1.669763e-07</td>
<td>-1.290486</td>
<td>0.185</td>
<td>0.215</td>
<td>2.671620e-06</td>
<td>Lotus-Mod12</td>
<td>Cortex</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>You can look at any other cluster. For example trichoblasts. Here you can see how module 4 pops up as being basically entirely expressed only in Trichoblasts.</p>
<div class="cell" data-execution_count="182">
<div class="sourceCode cell-code" id="cb199"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb199-1"><a href="#cb199-1" aria-hidden="true" tabindex="-1"></a>DMEs_tricho <span class="ot">&lt;-</span> DMEs_all <span class="sc">%&gt;%</span> <span class="fu">filter</span>(p_val_adj <span class="sc">&lt;</span> .<span class="dv">001</span> <span class="sc">&amp;</span> <span class="fu">abs</span>(avg_log2FC)<span class="sc">&gt;</span><span class="dv">1</span> </span>
<span id="cb199-2"><a href="#cb199-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="sc">&amp;</span> <span class="fu">is.finite</span>(avg_log2FC)</span>
<span id="cb199-3"><a href="#cb199-3" aria-hidden="true" tabindex="-1"></a>                                       <span class="sc">&amp;</span> group<span class="sc">==</span><span class="st">'Trichoblasts'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="183">
<div class="sourceCode cell-code" id="cb200"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1" aria-hidden="true" tabindex="-1"></a>DMEs_tricho</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 15 × 7</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">p_val</th>
<th data-quarto-table-cell-role="th" scope="col">avg_log2FC</th>
<th data-quarto-table-cell-role="th" scope="col">pct.1</th>
<th data-quarto-table-cell-role="th" scope="col">pct.2</th>
<th data-quarto-table-cell-role="th" scope="col">p_val_adj</th>
<th data-quarto-table-cell-role="th" scope="col">module</th>
<th data-quarto-table-cell-role="th" scope="col">group</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod4</td>
<td>0.000000e+00</td>
<td>6.750784</td>
<td>0.996</td>
<td>0.076</td>
<td>0.000000e+00</td>
<td>Lotus-Mod4</td>
<td>Trichoblasts</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod16</td>
<td>5.097196e-185</td>
<td>-6.481375</td>
<td>0.014</td>
<td>0.371</td>
<td>8.155514e-184</td>
<td>Lotus-Mod16</td>
<td>Trichoblasts</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod14</td>
<td>3.025439e-146</td>
<td>-4.349619</td>
<td>0.069</td>
<td>0.376</td>
<td>4.840702e-145</td>
<td>Lotus-Mod14</td>
<td>Trichoblasts</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod8</td>
<td>1.126480e-121</td>
<td>-4.872991</td>
<td>0.014</td>
<td>0.279</td>
<td>1.802368e-120</td>
<td>Lotus-Mod8</td>
<td>Trichoblasts</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod7</td>
<td>6.713357e-119</td>
<td>-6.178666</td>
<td>0.012</td>
<td>0.269</td>
<td>1.074137e-117</td>
<td>Lotus-Mod7</td>
<td>Trichoblasts</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod6</td>
<td>1.863436e-115</td>
<td>-5.773750</td>
<td>0.013</td>
<td>0.266</td>
<td>2.981498e-114</td>
<td>Lotus-Mod6</td>
<td>Trichoblasts</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod5</td>
<td>2.017255e-92</td>
<td>-3.586865</td>
<td>0.072</td>
<td>0.299</td>
<td>3.227608e-91</td>
<td>Lotus-Mod5</td>
<td>Trichoblasts</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod1</td>
<td>7.273801e-83</td>
<td>-6.444630</td>
<td>0.010</td>
<td>0.201</td>
<td>1.163808e-81</td>
<td>Lotus-Mod1</td>
<td>Trichoblasts</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod11</td>
<td>3.116022e-75</td>
<td>-5.254196</td>
<td>0.018</td>
<td>0.199</td>
<td>4.985635e-74</td>
<td>Lotus-Mod11</td>
<td>Trichoblasts</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod3</td>
<td>7.598832e-73</td>
<td>-3.568617</td>
<td>0.117</td>
<td>0.306</td>
<td>1.215813e-71</td>
<td>Lotus-Mod3</td>
<td>Trichoblasts</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod12</td>
<td>8.732984e-73</td>
<td>-3.106583</td>
<td>0.033</td>
<td>0.217</td>
<td>1.397277e-71</td>
<td>Lotus-Mod12</td>
<td>Trichoblasts</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod15</td>
<td>3.656164e-65</td>
<td>-3.844803</td>
<td>0.039</td>
<td>0.210</td>
<td>5.849862e-64</td>
<td>Lotus-Mod15</td>
<td>Trichoblasts</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod9</td>
<td>5.426530e-44</td>
<td>-3.051174</td>
<td>0.081</td>
<td>0.220</td>
<td>8.682448e-43</td>
<td>Lotus-Mod9</td>
<td>Trichoblasts</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod10</td>
<td>1.031038e-29</td>
<td>-3.738713</td>
<td>0.061</td>
<td>0.162</td>
<td>1.649662e-28</td>
<td>Lotus-Mod10</td>
<td>Trichoblasts</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">Trichoblasts.Lotus-Mod13</td>
<td>4.773544e-17</td>
<td>-2.852839</td>
<td>0.075</td>
<td>0.149</td>
<td>7.637670e-16</td>
<td>Lotus-Mod13</td>
<td>Trichoblasts</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>For better visualization, you can always use the lolliplot plots using the correct table as an argument, so that you can plot multiple instances for the various clusters. In <a href="#fig-lol2">Figure&nbsp;32</a> we plot boyh for cortex and trichoblasts from the tables determined in the code above.</p>
<div class="cell" data-execution_count="184">
<div class="sourceCode cell-code" id="cb201"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb201-1"><a href="#cb201-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">12</span>, <span class="at">repr.plot.height=</span><span class="dv">6</span>)</span>
<span id="cb201-2"><a href="#cb201-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb201-3"><a href="#cb201-3" aria-hidden="true" tabindex="-1"></a><span class="co">#lollipop with cortex table</span></span>
<span id="cb201-4"><a href="#cb201-4" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">PlotDMEsLollipop</span>(</span>
<span id="cb201-5"><a href="#cb201-5" aria-hidden="true" tabindex="-1"></a>  seurat.network, </span>
<span id="cb201-6"><a href="#cb201-6" aria-hidden="true" tabindex="-1"></a>  DMEs_cortex, </span>
<span id="cb201-7"><a href="#cb201-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">wgcna_name=</span><span class="st">'tutorial'</span>, </span>
<span id="cb201-8"><a href="#cb201-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">pvalue =</span> <span class="st">"p_val_adj"</span>,</span>
<span id="cb201-9"><a href="#cb201-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb201-10"><a href="#cb201-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb201-11"><a href="#cb201-11" aria-hidden="true" tabindex="-1"></a><span class="co">#lollipop with trichoblasts table</span></span>
<span id="cb201-12"><a href="#cb201-12" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">PlotDMEsLollipop</span>(</span>
<span id="cb201-13"><a href="#cb201-13" aria-hidden="true" tabindex="-1"></a>  seurat.network, </span>
<span id="cb201-14"><a href="#cb201-14" aria-hidden="true" tabindex="-1"></a>  DMEs_tricho, </span>
<span id="cb201-15"><a href="#cb201-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">wgcna_name=</span><span class="st">'tutorial'</span>, </span>
<span id="cb201-16"><a href="#cb201-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">pvalue =</span> <span class="st">"p_val_adj"</span>,</span>
<span id="cb201-17"><a href="#cb201-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb201-18"><a href="#cb201-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb201-19"><a href="#cb201-19" aria-hidden="true" tabindex="-1"></a><span class="co">#plot an add a title</span></span>
<span id="cb201-20"><a href="#cb201-20" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"DMEs of Cortex VS data"</span>) <span class="sc">+</span> </span>
<span id="cb201-21"><a href="#cb201-21" aria-hidden="true" tabindex="-1"></a>p2 <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"DMEs of Trichoblasts VS data"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Please be aware comparison group/groups are not provided, which may casue an ERROR. PlotDMEsLollipop function will automatically assume all values are within the same group."
[1] "Please be aware comparison group/groups are not provided, which may casue an ERROR. PlotDMEsLollipop function will automatically assume all values are within the same group."</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-lol2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="notebook.source_files/figure-html/fig-lol2-output-2.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;32: Lollipop plot that shovs the average log-fold change of each module and the p-value (size of each circle). Crossed circles, when present, have a non-significant p-value. Plotting lolliplot plots for various clusters can be useful to detect modules being simultaneously significant.</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Wrapping Up">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Wrapping Up
</div>
</div>
<div class="callout-body-container callout-body">
<p>This is the end of the tutorial. We have not shown much biological conclusion from the analysis - this is something that comes out by studying for example the GO terms of the various genes and modules identified. The scope of the tutorial was mainly to give all the means to perform your own analysis, from which you can then gain biological insight.</p>
<p>If you are interested in more resources to learn single cell analysis, you can find them at some of our courses. We have other single-cell analysis tutorials including different tools at</p>
<ul>
<li><strong>Introduction to NGS data analysis</strong> (found in the <code>Genomics Sandbox</code> <strong><a href="https://cloud.sdu.dk/app/jobs/create?app=genomics&amp;version=2023.03.01">at this link</a></strong> - note that this is in the <code>python</code> language )</li>
<li><strong>Introduction to scRNAseq analysis in R</strong> (found in the <code>Transcriptomics Sandbox</code> <strong><a href="https://cloud.sdu.dk/app/jobs/create?app=transcriptomics&amp;version=2023.11">at this link</a></strong> )</li>
</ul>
<p>You can also find a lot of material in the <strong><a href="https://satijalab.org/seurat/articles/get_started.html"><code>Seurat</code> webpage</a></strong>. If you create a new notebook in this environment, you are likely to have all the packages needed to try the <code>Seurat</code> tutorials.</p>
</div>
</div>



</section>
</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-adossa_computational_2021" class="csl-entry" role="listitem">
Adossa, Nigatu, Sofia Khan, Kalle T. Rytkönen, and Laura L. Elo. 2021. <span>“Computational Strategies for Single-Cell Multi-Omics Integration.”</span> <em>Computational and Structural Biotechnology Journal</em> 19: 2588–96. <a href="https://doi.org/10.1016/j.csbj.2021.04.060">https://doi.org/10.1016/j.csbj.2021.04.060</a>.
</div>
<div id="ref-becht_dimensionality_2019" class="csl-entry" role="listitem">
Becht, Etienne, Leland McInnes, John Healy, Charles-Antoine Dutertre, Immanuel W H Kwok, Lai Guan Ng, Florent Ginhoux, and Evan W Newell. 2019. <span>“Dimensionality Reduction for Visualizing Single-Cell Data Using <span>UMAP</span>.”</span> <em>Nature Biotechnology</em> 37 (1): 38–44. <a href="https://doi.org/10.1038/nbt.4314">https://doi.org/10.1038/nbt.4314</a>.
</div>
<div id="ref-cheng_review_2023" class="csl-entry" role="listitem">
Cheng, Changde, Wenan Chen, Hongjian Jin, and Xiang Chen. 2023. <span>“A <span>Review</span> of <span>Single</span>-<span>Cell</span> <span>RNA</span>-<span>Seq</span> <span>Annotation</span>, <span>Integration</span>, and <span>Cell</span>–<span>Cell</span> <span>Communication</span>.”</span> <em>Cells</em> 12 (15): 1970. <a href="https://doi.org/10.3390/cells12151970">https://doi.org/10.3390/cells12151970</a>.
</div>
<div id="ref-fleming_unsupervised_2023" class="csl-entry" role="listitem">
Fleming, Stephen J., Mark D. Chaffin, Alessandro Arduini, Amer-Denis Akkad, Eric Banks, John C. Marioni, Anthony A. Philippakis, Patrick T. Ellinor, and Mehrtash Babadi. 2023. <span>“Unsupervised Removal of Systematic Background Noise from Droplet-Based Single-Cell Experiments Using <span>CellBender</span>.”</span> <em>Nature Methods</em> 20 (9): 1323–35. <a href="https://doi.org/10.1038/s41592-023-01943-7">https://doi.org/10.1038/s41592-023-01943-7</a>.
</div>
<div id="ref-frank_single-cell_2023" class="csl-entry" role="listitem">
Frank, Manuel, Lavinia Ioana Fechete, Francesca Tedeschi, Marcin Nadzieja, Malita Malou Malekzadeh Nørgaard, Jesus Montiel, Kasper Røjkjær Andersen, Mikkel H. Schierup, Dugald Reid, and Stig Uggerhøj Andersen. 2023. <span>“Single-Cell Analysis Identifies Genes Facilitating Rhizobium Infection in <span>Lotus</span> Japonicus.”</span> <em>Nature Communications</em> 14 (November): 7171. <a href="https://doi.org/10.1038/s41467-023-42911-1">https://doi.org/10.1038/s41467-023-42911-1</a>.
</div>
<div id="ref-hafemeister_normalization_2019" class="csl-entry" role="listitem">
Hafemeister, Christoph, and Rahul Satija. 2019. <span>“Normalization and Variance Stabilization of Single-Cell <span>RNA</span>-Seq Data Using Regularized Negative Binomial Regression.”</span> <em>Genome Biology</em> 20 (1): 296. <a href="https://doi.org/10.1186/s13059-019-1874-1">https://doi.org/10.1186/s13059-019-1874-1</a>.
</div>
<div id="ref-heumos_best_2023" class="csl-entry" role="listitem">
Heumos, Lukas, Anna C. Schaar, Christopher Lance, Anastasia Litinetskaya, Felix Drost, Luke Zappia, Malte D. Lücken, et al. 2023. <span>“Best Practices for Single-Cell Analysis Across Modalities.”</span> <em>Nature Reviews Genetics</em> 24 (8): 550–72. <a href="https://doi.org/10.1038/s41576-023-00586-w">https://doi.org/10.1038/s41576-023-00586-w</a>.
</div>
<div id="ref-korsunsky_fast_2019" class="csl-entry" role="listitem">
Korsunsky, Ilya, Nghia Millard, Jean Fan, Kamil Slowikowski, Fan Zhang, Kevin Wei, Yuriy Baglaenko, Michael Brenner, Po-ru Loh, and Soumya Raychaudhuri. 2019. <span>“Fast, Sensitive and Accurate Integration of Single-Cell Data with <span>Harmony</span>.”</span> <em>Nature Methods</em> 16 (12): 1289–96. <a href="https://doi.org/10.1038/s41592-019-0619-0">https://doi.org/10.1038/s41592-019-0619-0</a>.
</div>
<div id="ref-langfelder_eigengene_2007" class="csl-entry" role="listitem">
Langfelder, Peter, and Steve Horvath. 2007. <span>“Eigengene Networks for Studying the Relationships Between Co-Expression Modules.”</span> <em>BMC Systems Biology</em> 1 (1): 54. <a href="https://doi.org/10.1186/1752-0509-1-54">https://doi.org/10.1186/1752-0509-1-54</a>.
</div>
<div id="ref-li_atypical_2019" class="csl-entry" role="listitem">
Li, Xiaolin, Zhiqiong Zheng, Xiangxiao Kong, Ji Xu, Liping Qiu, Jongho Sun, Dugald Reid, et al. 2019. <span>“Atypical Receptor Kinase <span>RINRK</span>1 Required for Rhizobial Infection but Not Nodule Development in <em>Lotus Japonicus</em>.”</span> <em>Plant Physiology</em> 181 (2): 804–16. <a href="https://doi.org/10.1104/pp.19.00509">https://doi.org/10.1104/pp.19.00509</a>.
</div>
<div id="ref-luecken_current_2019" class="csl-entry" role="listitem">
Luecken, Malte D, and Fabian J Theis. 2019. <span>“Current Best Practices in Single‐cell <span>RNA</span>‐seq Analysis: A Tutorial.”</span> <em>Molecular Systems Biology</em> 15 (6): e8746. <a href="https://doi.org/10.15252/msb.20188746">https://doi.org/10.15252/msb.20188746</a>.
</div>
<div id="ref-mcginnis_doubletfinder_2019" class="csl-entry" role="listitem">
McGinnis, Christopher S., Lyndsay M. Murrow, and Zev J. Gartner. 2019. <span>“<span>DoubletFinder</span>: <span>Doublet</span> <span>Detection</span> in <span>Single</span>-<span>Cell</span> <span>RNA</span> <span>Sequencing</span> <span>Data</span> <span>Using</span> <span>Artificial</span> <span>Nearest</span> <span>Neighbors</span>.”</span> <em>Cell Systems</em> 8 (4): 329–337.e4. <a href="https://doi.org/10.1016/j.cels.2019.03.003">https://doi.org/10.1016/j.cels.2019.03.003</a>.
</div>
<div id="ref-mcinnes_umap_2018" class="csl-entry" role="listitem">
McInnes, Leland, John Healy, Nathaniel Saul, and Lukas Großberger. 2018. <span>“<span>UMAP</span>: <span>Uniform</span> <span>Manifold</span> <span>Approximation</span> and <span>Projection</span>.”</span> <em>Journal of Open Source Software</em> 3 (29): 861. <a href="https://doi.org/10.21105/joss.00861">https://doi.org/10.21105/joss.00861</a>.
</div>
<div id="ref-morabito_hdwgcna_2023" class="csl-entry" role="listitem">
Morabito, Samuel, Fairlie Reese, Negin Rahimzadeh, Emily Miyoshi, and Vivek Swarup. 2023. <span>“<span class="nocase">hdWGCNA</span> Identifies Co-Expression Networks in High-Dimensional Transcriptomics Data.”</span> <em>Cell Reports Methods</em> 3 (6): 100498. <a href="https://doi.org/10.1016/j.crmeth.2023.100498">https://doi.org/10.1016/j.crmeth.2023.100498</a>.
</div>
<div id="ref-okamoto_shoot_2015" class="csl-entry" role="listitem">
Okamoto, Satoru, and Masayoshi Kawaguchi. 2015. <span>“Shoot <span>HAR</span>1 Mediates Nitrate Inhibition of Nodulation in <em>Lotus Japonicus</em>.”</span> <em>Plant Signaling &amp; Behavior</em> 10 (5): e1000138. <a href="https://doi.org/10.1080/15592324.2014.1000138">https://doi.org/10.1080/15592324.2014.1000138</a>.
</div>
<div id="ref-stuart_comprehensive_2019" class="csl-entry" role="listitem">
Stuart, Tim, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M. Mauck, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. 2019. <span>“Comprehensive <span>Integration</span> of <span>Single</span>-<span>Cell</span> <span>Data</span>.”</span> <em>Cell</em> 177 (7): 1888–1902.e21. <a href="https://doi.org/10.1016/j.cell.2019.05.031">https://doi.org/10.1016/j.cell.2019.05.031</a>.
</div>
<div id="ref-szczyglowski_nodule_1998" class="csl-entry" role="listitem">
Szczyglowski, Krzysztof, Robert S. Shaw, Judith Wopereis, Sue Copeland, Dirk Hamburger, Beth Kasiborski, Frank B. Dazzo, and Frans J. De Bruijn. 1998. <span>“Nodule Organogenesis and Symbiotic Mutants of the Model Legume <em>Lotus Japonicus</em>.”</span> <em>Molecular Plant-Microbe Interactions®</em> 11 (7): 684–97. <a href="https://doi.org/10.1094/MPMI.1998.11.7.684">https://doi.org/10.1094/MPMI.1998.11.7.684</a>.
</div>
<div id="ref-traag_louvain_2019" class="csl-entry" role="listitem">
Traag, V. A., L. Waltman, and N. J. Van Eck. 2019. <span>“From <span>Louvain</span> to <span>Leiden</span>: Guaranteeing Well-Connected Communities.”</span> <em>Scientific Reports</em> 9 (1): 5233. <a href="https://doi.org/10.1038/s41598-019-41695-z">https://doi.org/10.1038/s41598-019-41695-z</a>.
</div>
<div id="ref-wang_genetic_2018" class="csl-entry" role="listitem">
Wang, Qi, Jinge Liu, and Hongyan Zhu. 2018. <span>“Genetic and Molecular Mechanisms Underlying Symbiotic Specificity in Legume-Rhizobium Interactions.”</span> <em>Frontiers in Plant Science</em> 9. <a href="https://www.frontiersin.org/journals/plant-science/articles/10.3389/fpls.2018.00313">https://www.frontiersin.org/journals/plant-science/articles/10.3389/fpls.2018.00313</a>.
</div>
<div id="ref-xiang_comparison_2021" class="csl-entry" role="listitem">
Xiang, Ruizhi, Wencan Wang, Lei Yang, Shiyuan Wang, Chaohan Xu, and Xiaowen Chen. 2021. <span>“A <span>Comparison</span> for <span>Dimensionality</span> <span>Reduction</span> <span>Methods</span> of <span>Single</span>-<span>Cell</span> <span>RNA</span>-Seq <span>Data</span>.”</span> <em>Frontiers in Genetics</em> 12 (March): 646936. <a href="https://doi.org/10.3389/fgene.2021.646936">https://doi.org/10.3389/fgene.2021.646936</a>.
</div>
<div id="ref-xinming_seurat_2022" class="csl-entry" role="listitem">
Xinming, Tu. 2022. <span>“Seurat <span>CCA</span>? <span>It</span>’s Just a Simple Extension of <span>PCA</span>!”</span> <a href="https://xinmingtu.cn/blog/2022/CCA_dual_PCA/">https://xinmingtu.cn/blog/2022/CCA_dual_PCA/</a>.
</div>
<div id="ref-young_soupx_2020" class="csl-entry" role="listitem">
Young, Matthew D, and Sam Behjati. 2020. <span>“<span>SoupX</span> Removes Ambient <span>RNA</span> Contamination from Droplet-Based Single-Cell <span>RNA</span> Sequencing Data.”</span> <em>GigaScience</em> 9 (12): giaa151. <a href="https://doi.org/10.1093/gigascience/giaa151">https://doi.org/10.1093/gigascience/giaa151</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>